// @instantdb/react/dist/standalone/index.mjs
import Wn, { useRef as Fe, useCallback as Ce, useSyncExternalStore as at, useEffect as ne, useState as Vn, useMemo as Bn, createElement as Xr } from "react";
let tt = !1;
typeof window < "u" && typeof window.localStorage < "u" && (tt = !!window.localStorage.getItem("devBackend") || !!window.localStorage.getItem("__instantLogging"));
const N = {
  info: tt ? console.info.bind(console) : () => {
  },
  debug: tt ? console.debug.bind(console) : () => {
  },
  error: tt ? console.error.bind(console) : () => {
  }
};
function V(t) {
  if (typeof t == "number")
    return (Math.abs(t * 2654435761) >>> 0).toString(16);
  if (typeof t == "boolean")
    return t ? "1" : "0";
  if (t === null)
    return "null";
  if (t === void 0)
    return "undefined";
  if (typeof t == "string") {
    let e = 2166136261;
    for (let n = 0; n < t.length; n++)
      e ^= t.charCodeAt(n), e += (e << 1) + (e << 4) + (e << 7) + (e << 8) + (e << 24), e = e >>> 0;
    return e.toString(16);
  }
  if (Array.isArray(t)) {
    let e = 2166136261;
    for (let n = 0; n < t.length; n++) {
      e ^= (n + 1) * 2654435761;
      const r = V(t[n]);
      for (let i = 0; i < r.length; i++)
        e ^= r.charCodeAt(i), e *= 16777619, e = e >>> 0;
    }
    return e.toString(16);
  }
  if (typeof t == "object") {
    let e = 2166136261;
    const n = Object.keys(t).sort();
    for (let r = 0; r < n.length; r++) {
      const i = n[r], s = V(i);
      e ^= parseInt(s, 16), e *= 16777619, e = e >>> 0;
      const o = V(t[i]);
      e ^= parseInt(o, 16), e *= 16777619, e = e >>> 0;
    }
    return e.toString(16);
  }
  return V(String(t));
}
const M = {
  Remove: "remove",
  Replace: "replace",
  Add: "add"
}, Qn = Symbol.for("__MUTATIVE_PROXY_DRAFT__"), Zr = Symbol("__MUTATIVE_RAW_RETURN_SYMBOL__"), nt = Symbol.iterator, H = {
  mutable: "mutable",
  immutable: "immutable"
}, Qt = {};
function Y(t) {
  var e;
  return (e = t.copy) !== null && e !== void 0 ? e : t.original;
}
function me(t) {
  return !!j(t);
}
function j(t) {
  return typeof t != "object" ? null : t == null ? void 0 : t[Qn];
}
function Jt(t) {
  var e;
  const n = j(t);
  return n ? (e = n.copy) !== null && e !== void 0 ? e : n.original : t;
}
function re(t, e) {
  if (!t || typeof t != "object")
    return !1;
  let n;
  return Object.getPrototypeOf(t) === Object.prototype || Array.isArray(t) || t instanceof Map || t instanceof Set || !!(e != null && e.mark) && ((n = e.mark(t, H)) === H.immutable || typeof n == "function");
}
function Jn(t, e = []) {
  if (Object.hasOwnProperty.call(t, "key")) {
    const n = j(fe(t.parent.copy, t.key));
    if (n !== null && (n == null ? void 0 : n.original) !== t.original)
      return null;
    e.push(t.parent.type === 3 ? Array.from(t.parent.setMap.keys()).indexOf(t.key) : t.key);
  }
  if (t.parent)
    return Jn(t.parent, e);
  e.reverse();
  try {
    ei(t.copy, e);
  } catch {
    return null;
  }
  return e;
}
function Oe(t) {
  return Array.isArray(t) ? 1 : t instanceof Map ? 2 : t instanceof Set ? 3 : 0;
}
function fe(t, e) {
  return Oe(t) === 2 ? t.get(e) : t[e];
}
function ze(t, e, n) {
  Oe(t) === 2 ? t.set(e, n) : t[e] = n;
}
function Et(t, e) {
  const n = j(t);
  return (n ? Y(n) : t)[e];
}
function de(t, e) {
  return t === e ? t !== 0 || 1 / t === 1 / e : t !== t && e !== e;
}
function Ct(t) {
  if (t)
    for (; t.finalities.revoke.length > 0; )
      t.finalities.revoke.pop()();
}
function ge(t, e) {
  return e ? t : [""].concat(t).map((n) => {
    const r = `${n}`;
    return r.indexOf("/") === -1 && r.indexOf("~") === -1 ? r : r.replace(/~/g, "~0").replace(/\//g, "~1");
  }).join("/");
}
function ei(t, e) {
  for (let n = 0; n < e.length - 1; n += 1) {
    const r = e[n];
    if (t = fe(Oe(t) === 3 ? Array.from(t) : t, r), typeof t != "object")
      throw new Error(`Cannot resolve patch at '${e.join("/")}'.`);
  }
  return t;
}
function ti(t) {
  const e = Object.create(Object.getPrototypeOf(t));
  return Reflect.ownKeys(t).forEach((n) => {
    let r = Reflect.getOwnPropertyDescriptor(t, n);
    if (r.enumerable && r.configurable && r.writable) {
      e[n] = t[n];
      return;
    }
    r.writable || (r.writable = !0, r.configurable = !0), (r.get || r.set) && (r = {
      configurable: !0,
      writable: !0,
      enumerable: r.enumerable,
      value: t[n]
    }), Reflect.defineProperty(e, n, r);
  }), e;
}
const ni = Object.prototype.propertyIsEnumerable;
function Kn(t, e) {
  let n;
  if (Array.isArray(t))
    return Array.prototype.concat.call(t);
  if (t instanceof Set)
    return new Set(t.values());
  if (t instanceof Map)
    return new Map(t);
  if (e != null && e.mark && (n = e.mark(t, H), n !== void 0) && n !== H.mutable) {
    if (n === H.immutable)
      return ti(t);
    if (typeof n == "function") {
      if (e.enablePatches || e.enableAutoFreeze)
        throw new Error("You can't use mark and patches or auto freeze together.");
      return n();
    }
    throw new Error(`Unsupported mark result: ${n}`);
  } else if (typeof t == "object" && Object.getPrototypeOf(t) === Object.prototype) {
    const r = {};
    return Object.keys(t).forEach((i) => {
      r[i] = t[i];
    }), Object.getOwnPropertySymbols(t).forEach((i) => {
      ni.call(t, i) && (r[i] = t[i]);
    }), r;
  } else
    throw new Error("Please check mark() to ensure that it is a stable marker draftable function.");
}
function F(t) {
  t.copy || (t.copy = Kn(t.original, t.options));
}
function Le(t) {
  if (!re(t))
    return Jt(t);
  if (Array.isArray(t))
    return t.map(Le);
  if (t instanceof Map)
    return new Map(Array.from(t.entries()).map(([n, r]) => [n, Le(r)]));
  if (t instanceof Set)
    return new Set(Array.from(t).map(Le));
  const e = Object.create(Object.getPrototypeOf(t));
  for (const n in t)
    e[n] = Le(t[n]);
  return e;
}
function rt(t) {
  return me(t) ? Le(t) : t;
}
function oe(t) {
  var e;
  t.assignedMap = (e = t.assignedMap) !== null && e !== void 0 ? e : /* @__PURE__ */ new Map(), t.operated || (t.operated = !0, t.parent && oe(t.parent));
}
function yn() {
  throw new Error("Cannot modify frozen object");
}
function Ee(t, e, n, r, i) {
  {
    n = n ?? /* @__PURE__ */ new WeakMap(), r = r ?? [], i = i ?? [];
    const o = n.has(t) ? n.get(t) : t;
    if (r.length > 0) {
      const a = r.indexOf(o);
      if (o && typeof o == "object" && a !== -1)
        throw r[0] === o ? new Error("Forbids circular reference") : new Error(`Forbids circular reference: ~/${i.slice(0, a).map((c, u) => {
          if (typeof c == "symbol")
            return `[${c.toString()}]`;
          const l = r[u];
          return typeof c == "object" && (l instanceof Map || l instanceof Set) ? Array.from(l.keys()).indexOf(c) : c;
        }).join("/")}`);
      r.push(o), i.push(e);
    } else
      r.push(o);
  }
  if (Object.isFrozen(t) || me(t)) {
    r.pop(), i.pop();
    return;
  }
  switch (Oe(t)) {
    case 2:
      for (const [a, c] of t)
        Ee(a, a, n, r, i), Ee(c, a, n, r, i);
      t.set = t.clear = t.delete = yn;
      break;
    case 3:
      for (const a of t)
        Ee(a, a, n, r, i);
      t.add = t.clear = t.delete = yn;
      break;
    case 1:
      Object.freeze(t);
      let o = 0;
      for (const a of t)
        Ee(a, o, n, r, i), o += 1;
      break;
    default:
      Object.freeze(t), Object.keys(t).forEach((a) => {
        const c = t[a];
        Ee(c, a, n, r, i);
      });
  }
  r.pop(), i.pop();
}
function it(t, e) {
  return t instanceof Map ? t.has(e) : Object.prototype.hasOwnProperty.call(t, e);
}
function gn(t, e) {
  if (e in t) {
    let n = Reflect.getPrototypeOf(t);
    for (; n; ) {
      const r = Reflect.getOwnPropertyDescriptor(n, e);
      if (r)
        return r;
      n = Reflect.getPrototypeOf(n);
    }
  }
}
function Kt(t, e) {
  const n = Oe(t);
  if (n === 0)
    Reflect.ownKeys(t).forEach((r) => {
      e(r, t[r], t);
    });
  else if (n === 1) {
    let r = 0;
    for (const i of t)
      e(r, i, t), r += 1;
  } else
    t.forEach((r, i) => e(i, r, t));
}
function Yn(t, e, n) {
  if (me(t) || !re(t, n) || e.has(t) || Object.isFrozen(t))
    return;
  const r = t instanceof Set, i = r ? /* @__PURE__ */ new Map() : void 0;
  if (e.add(t), Kt(t, (s, o) => {
    var a;
    if (me(o)) {
      const c = j(o);
      F(c);
      const u = !((a = c.assignedMap) === null || a === void 0) && a.size || c.operated ? c.copy : c.original;
      ze(r ? i : t, s, u);
    } else
      Yn(o, e, n);
  }), i) {
    const s = t, o = Array.from(s);
    s.clear(), o.forEach((a) => {
      s.add(i.has(a) ? i.get(a) : a);
    });
  }
}
function ri(t, e) {
  const n = t.type === 3 ? t.setMap : t.copy;
  t.finalities.revoke.length > 1 && t.assignedMap.get(e) && n && Yn(fe(n, e), t.finalities.handledSet, t.options);
}
function xt(t) {
  t.type === 3 && t.copy && (t.copy.clear(), t.setMap.forEach((e) => {
    t.copy.add(Jt(e));
  }));
}
function At(t, e, n, r) {
  if (t.operated && t.assignedMap && t.assignedMap.size > 0 && !t.finalized) {
    if (n && r) {
      const s = Jn(t);
      s && e(t, s, n, r);
    }
    t.finalized = !0;
  }
}
function Yt(t, e, n, r) {
  const i = j(n);
  i && (i.callbacks || (i.callbacks = []), i.callbacks.push((s, o) => {
    var a;
    const c = t.type === 3 ? t.setMap : t.copy;
    if (de(fe(c, e), n)) {
      let u = i.original;
      i.copy && (u = i.copy), xt(t), At(t, r, s, o), t.options.enableAutoFreeze && (t.options.updatedValues = (a = t.options.updatedValues) !== null && a !== void 0 ? a : /* @__PURE__ */ new WeakMap(), t.options.updatedValues.set(u, i.original)), ze(c, e, u);
    }
  }), t.options.enableAutoFreeze && i.finalities !== t.finalities && (t.options.enableAutoFreeze = !1)), re(n, t.options) && t.finalities.draft.push(() => {
    const s = t.type === 3 ? t.setMap : t.copy;
    de(fe(s, e), n) && ri(t, e);
  });
}
function ii(t, e, n, r, i) {
  let { original: s, assignedMap: o, options: a } = t, c = t.copy;
  c.length < s.length && ([s, c] = [c, s], [n, r] = [r, n]);
  for (let u = 0; u < s.length; u += 1)
    if (o.get(u.toString()) && c[u] !== s[u]) {
      const l = e.concat([u]), f = ge(l, i);
      n.push({
        op: M.Replace,
        path: f,
        // If it is a draft, it needs to be deep cloned, and it may also be non-draft.
        value: rt(c[u])
      }), r.push({
        op: M.Replace,
        path: f,
        // If it is a draft, it needs to be deep cloned, and it may also be non-draft.
        value: rt(s[u])
      });
    }
  for (let u = s.length; u < c.length; u += 1) {
    const l = e.concat([u]), f = ge(l, i);
    n.push({
      op: M.Add,
      path: f,
      // If it is a draft, it needs to be deep cloned, and it may also be non-draft.
      value: rt(c[u])
    });
  }
  if (s.length < c.length) {
    const { arrayLengthAssignment: u = !0 } = a.enablePatches;
    if (u) {
      const l = e.concat(["length"]), f = ge(l, i);
      r.push({
        op: M.Replace,
        path: f,
        value: s.length
      });
    } else
      for (let l = c.length; s.length < l; l -= 1) {
        const f = e.concat([l - 1]), p = ge(f, i);
        r.push({
          op: M.Remove,
          path: p
        });
      }
  }
}
function si({ original: t, copy: e, assignedMap: n }, r, i, s, o) {
  n.forEach((a, c) => {
    const u = fe(t, c), l = rt(fe(e, c)), f = a ? it(t, c) ? M.Replace : M.Add : M.Remove;
    if (de(u, l) && f === M.Replace)
      return;
    const p = r.concat(c), _ = ge(p, o);
    i.push(f === M.Remove ? { op: f, path: _ } : { op: f, path: _, value: l }), s.push(f === M.Add ? { op: M.Remove, path: _ } : f === M.Remove ? { op: M.Add, path: _, value: u } : { op: M.Replace, path: _, value: u });
  });
}
function oi({ original: t, copy: e }, n, r, i, s) {
  let o = 0;
  t.forEach((a) => {
    if (!e.has(a)) {
      const c = n.concat([o]), u = ge(c, s);
      r.push({
        op: M.Remove,
        path: u,
        value: a
      }), i.unshift({
        op: M.Add,
        path: u,
        value: a
      });
    }
    o += 1;
  }), o = 0, e.forEach((a) => {
    if (!t.has(a)) {
      const c = n.concat([o]), u = ge(c, s);
      r.push({
        op: M.Add,
        path: u,
        value: a
      }), i.unshift({
        op: M.Remove,
        path: u,
        value: a
      });
    }
    o += 1;
  });
}
function Ne(t, e, n, r) {
  const { pathAsArray: i = !0 } = t.options.enablePatches;
  switch (t.type) {
    case 0:
    case 2:
      return si(t, e, n, r, i);
    case 1:
      return ii(t, e, n, r, i);
    case 3:
      return oi(t, e, n, r, i);
  }
}
let ai = !1;
const ot = (t, e, n = !1) => {
  if (typeof t == "object" && t !== null && (!re(t, e) || n) && !ai)
    throw new Error("Strict mode: Mutable data cannot be accessed directly, please use 'unsafe(callback)' wrap.");
}, Mt = {
  get size() {
    return Y(j(this)).size;
  },
  has(t) {
    return Y(j(this)).has(t);
  },
  set(t, e) {
    const n = j(this), r = Y(n);
    return (!r.has(t) || !de(r.get(t), e)) && (F(n), oe(n), n.assignedMap.set(t, !0), n.copy.set(t, e), Yt(n, t, e, Ne)), this;
  },
  delete(t) {
    if (!this.has(t))
      return !1;
    const e = j(this);
    return F(e), oe(e), e.original.has(t) ? e.assignedMap.set(t, !1) : e.assignedMap.delete(t), e.copy.delete(t), !0;
  },
  clear() {
    const t = j(this);
    if (this.size) {
      F(t), oe(t), t.assignedMap = /* @__PURE__ */ new Map();
      for (const [e] of t.original)
        t.assignedMap.set(e, !1);
      t.copy.clear();
    }
  },
  forEach(t, e) {
    const n = j(this);
    Y(n).forEach((r, i) => {
      t.call(e, this.get(i), i, this);
    });
  },
  get(t) {
    var e, n;
    const r = j(this), i = Y(r).get(t), s = ((n = (e = r.options).mark) === null || n === void 0 ? void 0 : n.call(e, i, H)) === H.mutable;
    if (r.options.strict && ot(i, r.options, s), s || r.finalized || !re(i, r.options) || i !== r.original.get(t))
      return i;
    const o = Qt.createDraft({
      original: i,
      parentDraft: r,
      key: t,
      finalities: r.finalities,
      options: r.options
    });
    return F(r), r.copy.set(t, o), o;
  },
  keys() {
    return Y(j(this)).keys();
  },
  values() {
    const t = this.keys();
    return {
      [nt]: () => this.values(),
      next: () => {
        const e = t.next();
        return e.done ? e : {
          done: !1,
          value: this.get(e.value)
        };
      }
    };
  },
  entries() {
    const t = this.keys();
    return {
      [nt]: () => this.entries(),
      next: () => {
        const e = t.next();
        if (e.done)
          return e;
        const n = this.get(e.value);
        return {
          done: !1,
          value: [e.value, n]
        };
      }
    };
  },
  [nt]() {
    return this.entries();
  }
}, ci = Reflect.ownKeys(Mt), wn = (t, e, { isValuesIterator: n }) => () => {
  var r, i;
  const s = e.next();
  if (s.done)
    return s;
  const o = s.value;
  let a = t.setMap.get(o);
  const c = j(a), u = ((i = (r = t.options).mark) === null || i === void 0 ? void 0 : i.call(r, a, H)) === H.mutable;
  if (t.options.strict && ot(o, t.options, u), !u && !c && re(o, t.options) && !t.finalized && t.original.has(o)) {
    const l = Qt.createDraft({
      original: o,
      parentDraft: t,
      key: o,
      finalities: t.finalities,
      options: t.options
    });
    t.setMap.set(o, l), a = l;
  } else c && (a = c.proxy);
  return {
    done: !1,
    value: n ? a : [a, a]
  };
}, It = {
  get size() {
    return j(this).setMap.size;
  },
  has(t) {
    const e = j(this);
    if (e.setMap.has(t))
      return !0;
    F(e);
    const n = j(t);
    return !!(n && e.setMap.has(n.original));
  },
  add(t) {
    const e = j(this);
    return this.has(t) || (F(e), oe(e), e.assignedMap.set(t, !0), e.setMap.set(t, t), Yt(e, t, t, Ne)), this;
  },
  delete(t) {
    if (!this.has(t))
      return !1;
    const e = j(this);
    F(e), oe(e);
    const n = j(t);
    return n && e.setMap.has(n.original) ? (e.assignedMap.set(n.original, !1), e.setMap.delete(n.original)) : (!n && e.setMap.has(t) ? e.assignedMap.set(t, !1) : e.assignedMap.delete(t), e.setMap.delete(t));
  },
  clear() {
    if (!this.size)
      return;
    const t = j(this);
    F(t), oe(t);
    for (const e of t.original)
      t.assignedMap.set(e, !1);
    t.setMap.clear();
  },
  values() {
    const t = j(this);
    F(t);
    const e = t.setMap.keys();
    return {
      [Symbol.iterator]: () => this.values(),
      next: wn(t, e, { isValuesIterator: !0 })
    };
  },
  entries() {
    const t = j(this);
    F(t);
    const e = t.setMap.keys();
    return {
      [Symbol.iterator]: () => this.entries(),
      next: wn(t, e, {
        isValuesIterator: !1
      })
    };
  },
  keys() {
    return this.values();
  },
  [nt]() {
    return this.values();
  },
  forEach(t, e) {
    const n = this.values();
    let r = n.next();
    for (; !r.done; )
      t.call(e, r.value, r.value, this), r = n.next();
  }
}, ui = Reflect.ownKeys(It), Hn = /* @__PURE__ */ new WeakSet(), Gn = {
  get(t, e, n) {
    var r, i;
    const s = (r = t.copy) === null || r === void 0 ? void 0 : r[e];
    if (s && Hn.has(s))
      return s;
    if (e === Qn)
      return t;
    let o;
    if (t.options.mark) {
      const u = e === "size" && (t.original instanceof Map || t.original instanceof Set) ? Reflect.get(t.original, e) : Reflect.get(t.original, e, n);
      if (o = t.options.mark(u, H), o === H.mutable)
        return t.options.strict && ot(u, t.options, !0), u;
    }
    const a = Y(t);
    if (a instanceof Map && ci.includes(e)) {
      if (e === "size")
        return Object.getOwnPropertyDescriptor(Mt, "size").get.call(t.proxy);
      const u = Mt[e];
      if (u)
        return u.bind(t.proxy);
    }
    if (a instanceof Set && ui.includes(e)) {
      if (e === "size")
        return Object.getOwnPropertyDescriptor(It, "size").get.call(t.proxy);
      const u = It[e];
      if (u)
        return u.bind(t.proxy);
    }
    if (!it(a, e)) {
      const u = gn(a, e);
      return u ? "value" in u ? u.value : (
        // !case: support for getter
        (i = u.get) === null || i === void 0 ? void 0 : i.call(t.proxy)
      ) : void 0;
    }
    const c = a[e];
    if (t.options.strict && ot(c, t.options), t.finalized || !re(c, t.options))
      return c;
    if (c === Et(t.original, e)) {
      if (F(t), t.copy[e] = Ht({
        original: t.original[e],
        parentDraft: t,
        key: t.type === 1 ? Number(e) : e,
        finalities: t.finalities,
        options: t.options
      }), typeof o == "function") {
        const u = j(t.copy[e]);
        return F(u), oe(u), u.copy;
      }
      return t.copy[e];
    }
    return c;
  },
  set(t, e, n) {
    var r;
    if (t.type === 3 || t.type === 2)
      throw new Error("Map/Set draft does not support any property assignment.");
    let i;
    if (t.type === 1 && e !== "length" && !(Number.isInteger(i = Number(e)) && i >= 0 && (e === 0 || i === 0 || String(i) === String(e))))
      throw new Error("Only supports setting array indices and the 'length' property.");
    const s = gn(Y(t), e);
    if (s != null && s.set)
      return s.set.call(t.proxy, n), !0;
    const o = Et(Y(t), e), a = j(o);
    return a && de(a.original, n) ? (t.copy[e] = n, t.assignedMap = (r = t.assignedMap) !== null && r !== void 0 ? r : /* @__PURE__ */ new Map(), t.assignedMap.set(e, !1), !0) : (de(n, o) && (n !== void 0 || it(t.original, e)) || (F(t), oe(t), it(t.original, e) && de(n, t.original[e]) ? t.assignedMap.delete(e) : t.assignedMap.set(e, !0), t.copy[e] = n, Yt(t, e, n, Ne)), !0);
  },
  has(t, e) {
    return e in Y(t);
  },
  ownKeys(t) {
    return Reflect.ownKeys(Y(t));
  },
  getOwnPropertyDescriptor(t, e) {
    const n = Y(t), r = Reflect.getOwnPropertyDescriptor(n, e);
    return r && {
      writable: !0,
      configurable: t.type !== 1 || e !== "length",
      enumerable: r.enumerable,
      value: n[e]
    };
  },
  getPrototypeOf(t) {
    return Reflect.getPrototypeOf(t.original);
  },
  setPrototypeOf() {
    throw new Error("Cannot call 'setPrototypeOf()' on drafts");
  },
  defineProperty() {
    throw new Error("Cannot call 'defineProperty()' on drafts");
  },
  deleteProperty(t, e) {
    var n;
    return t.type === 1 ? Gn.set.call(this, t, e, void 0, t.proxy) : (Et(t.original, e) !== void 0 || e in t.original ? (F(t), oe(t), t.assignedMap.set(e, !1)) : (t.assignedMap = (n = t.assignedMap) !== null && n !== void 0 ? n : /* @__PURE__ */ new Map(), t.assignedMap.delete(e)), t.copy && delete t.copy[e], !0);
  }
};
function Ht(t) {
  const { original: e, parentDraft: n, key: r, finalities: i, options: s } = t, o = Oe(e), a = {
    type: o,
    finalized: !1,
    parent: n,
    original: e,
    copy: null,
    proxy: null,
    finalities: i,
    options: s,
    // Mapping of draft Set items to their corresponding draft values.
    setMap: o === 3 ? new Map(e.entries()) : void 0
  };
  (r || "key" in t) && (a.key = r);
  const { proxy: c, revoke: u } = Proxy.revocable(o === 1 ? Object.assign([], a) : a, Gn);
  if (i.revoke.push(u), Hn.add(c), a.proxy = c, n) {
    const l = n;
    l.finalities.draft.push((f, p) => {
      var _, g;
      const m = j(c);
      let O = l.type === 3 ? l.setMap : l.copy;
      const S = fe(O, r), b = j(S);
      if (b) {
        let y = b.original;
        b.operated && (y = Jt(S)), xt(b), At(b, Ne, f, p), l.options.enableAutoFreeze && (l.options.updatedValues = (_ = l.options.updatedValues) !== null && _ !== void 0 ? _ : /* @__PURE__ */ new WeakMap(), l.options.updatedValues.set(y, b.original)), ze(O, r, y);
      }
      (g = m.callbacks) === null || g === void 0 || g.forEach((y) => {
        y(f, p);
      });
    });
  } else {
    const l = j(c);
    l.finalities.draft.push((f, p) => {
      xt(l), At(l, Ne, f, p);
    });
  }
  return c;
}
Qt.createDraft = Ht;
function li(t, e, n, r, i) {
  var s;
  const o = j(t), a = (s = o == null ? void 0 : o.original) !== null && s !== void 0 ? s : t, c = !!e.length;
  if (o != null && o.operated)
    for (; o.finalities.draft.length > 0; )
      o.finalities.draft.pop()(n, r);
  const u = c ? e[0] : o ? o.operated ? o.copy : o.original : t;
  return o && Ct(o), i && Ee(u, u, o == null ? void 0 : o.options.updatedValues), [
    u,
    n && c ? [{ op: M.Replace, path: [], value: e[0] }] : n,
    r && c ? [{ op: M.Replace, path: [], value: a }] : r
  ];
}
function di(t, e) {
  var n;
  const r = {
    draft: [],
    revoke: [],
    handledSet: /* @__PURE__ */ new WeakSet()
  };
  let i, s;
  e.enablePatches && (i = [], s = []);
  const a = ((n = e.mark) === null || n === void 0 ? void 0 : n.call(e, t, H)) === H.mutable || !re(t, e) ? t : Ht({
    original: t,
    parentDraft: null,
    finalities: r,
    options: e
  });
  return [
    a,
    (c = []) => {
      const [u, l, f] = li(a, c, i, s, e.enableAutoFreeze);
      return e.enablePatches ? [u, l, f] : u;
    }
  ];
}
function kt(t) {
  const { rootDraft: e, value: n, useRawReturn: r = !1, isRoot: i = !0 } = t;
  Kt(n, (s, o, a) => {
    const c = j(o);
    if (c && e && c.finalities === e.finalities) {
      t.isContainDraft = !0;
      const u = c.original;
      if (a instanceof Set) {
        const l = Array.from(a);
        a.clear(), l.forEach((f) => a.add(s === f ? u : f));
      } else
        ze(a, s, u);
    } else typeof o == "object" && o !== null && (t.value = o, t.isRoot = !1, kt(t));
  }), i && (t.isContainDraft || console.warn("The return value does not contain any draft, please use 'rawReturn()' to wrap the return value to improve performance."), r && console.warn("The return value contains drafts, please don't use 'rawReturn()' to wrap the return value."));
}
function Xn(t) {
  const e = j(t);
  if (!re(t, e == null ? void 0 : e.options))
    return t;
  const n = Oe(t);
  if (e && !e.operated)
    return e.original;
  let r;
  function i() {
    r = n === 2 ? new Map(t) : n === 3 ? Array.from(e.setMap.values()) : Kn(t, e == null ? void 0 : e.options);
  }
  if (e) {
    e.finalized = !0;
    try {
      i();
    } finally {
      e.finalized = !1;
    }
  } else
    r = t;
  return Kt(r, (s, o) => {
    if (e && de(fe(e.original, s), o))
      return;
    const a = Xn(o);
    a !== o && (r === t && i(), ze(r, s, a));
  }), n === 3 ? new Set(r) : r;
}
function mn(t) {
  if (!me(t))
    throw new Error(`current() is only used for Draft, parameter: ${t}`);
  return Xn(t);
}
const fi = (t) => function e(n, r, i) {
  var s, o, a;
  if (typeof n == "function" && typeof r != "function")
    return function(R, ...z) {
      return e(R, ($) => n.call(this, $, ...z), r);
    };
  const c = n, u = r;
  let l = i;
  if (typeof r != "function" && (l = r), l !== void 0 && Object.prototype.toString.call(l) !== "[object Object]")
    throw new Error(`Invalid options: ${l}, 'options' should be an object.`);
  l = Object.assign(Object.assign({}, t), l);
  const f = me(c) ? mn(c) : c, p = Array.isArray(l.mark) ? (R, z) => {
    for (const $ of l.mark) {
      if (typeof $ != "function")
        throw new Error(`Invalid mark: ${$}, 'mark' should be a function.`);
      const ce = $(R, z);
      if (ce)
        return ce;
    }
  } : l.mark, _ = (s = l.enablePatches) !== null && s !== void 0 ? s : !1, g = (o = l.strict) !== null && o !== void 0 ? o : !1, O = {
    enableAutoFreeze: (a = l.enableAutoFreeze) !== null && a !== void 0 ? a : !1,
    mark: p,
    strict: g,
    enablePatches: _
  };
  if (!re(f, O) && typeof f == "object" && f !== null)
    throw new Error("Invalid base state: create() only supports plain objects, arrays, Set, Map or using mark() to mark the state as immutable.");
  const [S, b] = di(f, O);
  if (typeof r != "function") {
    if (!re(f, O))
      throw new Error("Invalid base state: create() only supports plain objects, arrays, Set, Map or using mark() to mark the state as immutable.");
    return [S, b];
  }
  let y;
  try {
    y = u(S);
  } catch (R) {
    throw Ct(j(S)), R;
  }
  const x = (R) => {
    const z = j(S);
    if (!me(R)) {
      if (R !== void 0 && !de(R, S) && (z != null && z.operated))
        throw new Error("Either the value is returned as a new non-draft value, or only the draft is modified without returning any value.");
      const ce = R == null ? void 0 : R[Zr];
      if (ce) {
        const xe = ce[0];
        return O.strict && typeof R == "object" && R !== null && kt({
          rootDraft: z,
          value: R,
          useRawReturn: !0
        }), b([xe]);
      }
      if (R !== void 0)
        return typeof R == "object" && R !== null && kt({ rootDraft: z, value: R }), b([R]);
    }
    if (R === S || R === void 0)
      return b([]);
    const $ = j(R);
    if (O === $.options) {
      if ($.operated)
        throw new Error("Cannot return a modified child draft.");
      return b([mn(R)]);
    }
    return b([R]);
  };
  return y instanceof Promise ? y.then(x, (R) => {
    throw Ct(j(S)), R;
  }) : x(y);
}, hi = fi();
Object.prototype.constructor.toString();
function Zn(t, e) {
  const n = Object.keys(t), r = Object.keys(e);
  return n.length === r.length && Object.keys(t).every((i) => e.hasOwnProperty(i));
}
function On(t, e) {
  return Object.keys(t).length === Object.keys(e).length && Object.keys(t).every((n) => e.hasOwnProperty(n) && t[n] === e[n]);
}
function st(t, e) {
  return typeof t != "object" || typeof e != "object" || t === null || e === null ? t === e : Zn(t, e) ? Object.keys(t).every((n) => st(t[n], e[n])) : !1;
}
function er(t, e) {
  if (!je(t) || !je(e))
    return e;
  const n = {};
  for (const r of Object.keys(t))
    e[r] !== null && (n[r] = t[r]);
  for (const r of Object.keys(e)) {
    if (e[r] === null)
      continue;
    const i = je(t[r]) && je(e[r]);
    n[r] = i ? er(t[r], e[r]) : e[r];
  }
  return n;
}
function tr(t, e, n) {
  if (!je(t))
    return t;
  const r = {};
  for (const [i, s] of Object.entries(t))
    r[i] = je(s) ? tr(s, e, n) : s === e ? n : s;
  return r;
}
function je(t) {
  return typeof t == "object" && t !== null && !Array.isArray(t);
}
function pi(t) {
  return t.cardinality === "one";
}
function Gt(t) {
  return t["value-type"] === "ref";
}
function nr(t) {
  return t["value-type"] === "blob";
}
function ct(t, e) {
  return t[e];
}
function Xt(t, e) {
  return e.reduce((n, r) => n && n.get(r), t);
}
function te(t, e) {
  if (e.length === 0)
    throw new Error("path must have at least one element");
  if (e.length === 1) {
    t.delete(e[0]);
    return;
  }
  const [n, ...r] = e;
  t.has(n) && te(t.get(n), r);
}
function B(t, e, n) {
  if (e.length === 0)
    throw new Error("path must have at least one element");
  if (e.length === 1) {
    t.set(e[0], n);
    return;
  }
  const [r, ...i] = e;
  let s = t.get(r);
  s || (s = /* @__PURE__ */ new Map(), t.set(r, s)), B(s, i, n);
}
function rr(t, e) {
  const n = /* @__PURE__ */ new Map(), r = /* @__PURE__ */ new Map(), i = /* @__PURE__ */ new Map();
  for (const s of e) {
    const [o, a, c, u] = s, l = ct(t, a);
    if (!l) {
      console.warn("no such attr", o, t);
      continue;
    }
    Gt(l) && B(i, [c, a, o], s), B(n, [o, a, c], s), B(r, [a, o, c], s);
  }
  return { eav: n, aev: r, vae: i };
}
function ir(t) {
  const e = /* @__PURE__ */ new Map(), n = /* @__PURE__ */ new Map(), r = /* @__PURE__ */ new Map(), i = /* @__PURE__ */ new Map();
  for (const s of Object.values(t)) {
    const o = s["forward-identity"], [a, c, u] = o, l = s["reverse-identity"];
    if (B(r, [c, u], s), nr(s) && B(e, [c, u], s), s["primary?"] && B(n, [c], s), l) {
      const [f, p, _] = l;
      B(i, [p, _], s);
    }
  }
  return { blobAttrs: e, primaryKeys: n, forwardIdents: r, revIdents: i };
}
function vi(t) {
  return {
    __type: t.__type,
    attrs: t.attrs,
    triples: X(t.eav, 3),
    cardinalityInference: t.cardinalityInference,
    linkIndex: t.linkIndex
  };
}
function _i(t) {
  return Ut(t.attrs, t.triples, t.cardinalityInference, t.linkIndex);
}
function Zt(t) {
  t.attrIndexes = ir(t.attrs);
}
function Ut(t, e, n, r) {
  const i = rr(t, e);
  return i.attrs = t, i.attrIndexes = ir(t), i.cardinalityInference = n, i.linkIndex = r, i.__type = "store", i;
}
function ut(t, e) {
  var n, r;
  let i;
  if (Array.isArray(e[0])) {
    const [o, a] = e[0], c = t.aev.get(o);
    if (!c)
      return null;
    i = (n = X(c, 2).find((l) => l[2] === a)) === null || n === void 0 ? void 0 : n[0];
  } else
    i = e[0];
  if (!i)
    return null;
  const s = e[2];
  if (Array.isArray(s) && s.length === 2 && t.aev.get(s[0])) {
    const [o, a] = s, c = t.aev.get(o);
    if (!c)
      return null;
    const l = (r = X(c, 2).find((m) => m[2] === a)) === null || r === void 0 ? void 0 : r[0];
    if (!l)
      return null;
    const [f, p, _, ...g] = e;
    return [i, p, l, ...g];
  } else {
    const [o, ...a] = e;
    return [i, ...a];
  }
}
function bi(t, e) {
  const n = ut(t, e);
  if (!n)
    return;
  const [r, i, s] = n, o = ct(t.attrs, i);
  o && (te(t.eav, [r, i, s]), te(t.aev, [i, r, s]), Gt(o) && te(t.vae, [s, i, r]));
}
let yi = 0;
function sr(t, e, n) {
  const [r, i, s] = n;
  let o;
  const a = Xt(t.ea, [r, i, s]);
  return a && (o = a[3]), o || Date.now() * 10 + yi++;
}
function gi(t, e) {
  var n;
  const r = ut(t, e);
  if (!r)
    return;
  const [i, s, o] = r, a = ct(t.attrs, s);
  if (!a)
    return;
  const c = Xt(t.eav, [i, s, o]), u = (n = c == null ? void 0 : c[3]) !== null && n !== void 0 ? n : sr(t, a, r), l = [i, s, o, u];
  pi(a) ? (B(t.eav, [i, s], /* @__PURE__ */ new Map([[o, l]])), B(t.aev, [s, i], /* @__PURE__ */ new Map([[o, l]]))) : (B(t.eav, [i, s, o], l), B(t.aev, [s, i, o], l)), Gt(a) && B(t.vae, [o, s, i], l);
}
function wi(t, e) {
  var n;
  const r = ut(t, e);
  if (!r)
    return;
  const [i, s, o] = r, a = ct(t.attrs, s);
  if (!a)
    return;
  if (!nr(a))
    throw new Error("merge operation is not supported for links");
  const c = Xt(t.eav, [i, s]);
  if (!c)
    return;
  const u = (n = c.values().next()) === null || n === void 0 ? void 0 : n.value;
  if (!u)
    return;
  const l = u[2], f = er(l, o), p = [
    i,
    s,
    f,
    sr(t, a, u)
  ];
  B(t.eav, [i, s], /* @__PURE__ */ new Map([[f, p]]));
}
function mi(t, e) {
  var n, r;
  const [i, s] = e, o = ut(t, [i]);
  if (!o)
    return;
  const [a] = o, c = t.eav.get(a);
  if (c) {
    for (const l of c.keys()) {
      const f = t.attrs[l];
      // Fall back to deleting everything if we've rehydrated tx-steps from
      // the store that didn't set `etype` in deleteEntity
      (!s || // If we don't know about the attr, let's just get rid of it
      !f || // Make sure it matches the etype
      ((n = f["forward-identity"]) === null || n === void 0 ? void 0 : n[1]) === s) && (te(t.aev, [l, a]), te(t.eav, [a, l]));
    }
    c.size === 0 && te(t.eav, [a]);
  }
  const u = t.vae.get(a) && X(t.vae.get(a), 2);
  u && u.forEach((l) => {
    var f;
    const [p, _, g] = l, m = t.attrs[_];
    (!s || !m || ((f = m["reverse-identity"]) === null || f === void 0 ? void 0 : f[1]) === s) && (te(t.eav, [p, _, g]), te(t.aev, [_, p, g]), te(t.vae, [g, _, p]));
  }), ((r = t.vae.get(a)) === null || r === void 0 ? void 0 : r.size) === 0 && te(t.vae, [a]);
}
function or(t, e) {
  const n = rr(t.attrs, e);
  Object.keys(n).forEach((r) => {
    t[r] = n[r];
  });
}
function Oi(t, [e]) {
  t.attrs[e.id] = e, Zt(t);
}
function ar(t) {
  return X(t.eav, 3);
}
function Si(t, [e]) {
  if (!t.attrs[e])
    return;
  const n = ar(t).filter(([r, i]) => i !== e);
  delete t.attrs[e], Zt(t), or(t, n);
}
function Ri(t, [e]) {
  const n = t.attrs[e.id];
  n && (t.attrs[e.id] = Object.assign(Object.assign({}, n), e), Zt(t), or(t, ar(t)));
}
function Ei(t, e) {
  const [n, ...r] = e;
  switch (n) {
    case "add-triple":
      gi(t, r);
      break;
    case "deep-merge-triple":
      wi(t, r);
      break;
    case "retract-triple":
      bi(t, r);
      break;
    case "delete-entity":
      mi(t, r);
      break;
    case "add-attr":
      Oi(t, r);
      break;
    case "delete-attr":
      Si(t, r);
      break;
    case "update-attr":
      Ri(t, r);
      break;
    default:
      throw new Error(`unhandled transaction action: ${n}`);
  }
}
function X(t, e, n = []) {
  if (!t || e === 0)
    return n;
  if (e === 1) {
    for (const r of t.values())
      n.push(r);
    return n;
  }
  for (const r of t.values())
    X(r, e - 1, n);
  return n;
}
function Ge(t, e, n) {
  var r, i;
  const s = [];
  if (n != null && n.hasOwnProperty("$not")) {
    for (const a of e.keys())
      n.$not !== a && s.push(e.get(a));
    return s;
  }
  if (n != null && n.hasOwnProperty("$isNull")) {
    const { attrId: a, isNull: c, reverse: u } = n.$isNull;
    if (u)
      for (const l of e.keys()) {
        const f = t.vae.get(l), p = !f || ((r = f.get(a)) === null || r === void 0 ? void 0 : r.get(null)) || !f.get(a);
        (c ? p : !p) && s.push(e.get(l));
      }
    else {
      const l = t.aev.get(a);
      for (const f of e.keys()) {
        const p = !l || ((i = l.get(f)) === null || i === void 0 ? void 0 : i.get(null)) || !l.get(f);
        (c ? p : !p) && s.push(e.get(f));
      }
    }
    return s;
  }
  if (n != null && n.$comparator)
    return X(e, 1).filter(n.$op);
  const o = n.in || n.$in || [n];
  for (const a of o) {
    const c = e.get(a);
    c && s.push(c);
  }
  return s;
}
function ji(t, e, n) {
  let r = "";
  return t !== void 0 && (r += "e"), e !== void 0 && (r += "a"), n !== void 0 && (r += "v"), r;
}
function Pi(t, [e, n, r]) {
  var i, s;
  switch (ji(e, n, r)) {
    case "e": {
      const a = t.eav.get(e);
      return X(a, 2);
    }
    case "ea": {
      const a = (i = t.eav.get(e)) === null || i === void 0 ? void 0 : i.get(n);
      return X(a, 1);
    }
    case "eav": {
      const a = (s = t.eav.get(e)) === null || s === void 0 ? void 0 : s.get(n);
      return a ? Ge(t, a, r) : [];
    }
    case "ev": {
      const a = t.eav.get(e);
      if (!a)
        return [];
      const c = [];
      for (const u of a.values())
        c.push(...Ge(t, u, r));
      return c;
    }
    case "a": {
      const a = t.aev.get(n);
      return X(a, 2);
    }
    case "av": {
      const a = t.aev.get(n);
      if (!a)
        return [];
      const c = [];
      for (const u of a.values())
        c.push(...Ge(t, u, r));
      return c;
    }
    case "v": {
      const a = [];
      for (const c of t.eav.values())
        for (const u of c.values())
          a.push(...Ge(t, u, r));
    }
    default:
      return X(t.eav, 3);
  }
}
function Ti(t, e, n) {
  var r;
  const i = t.attrIndexes.blobAttrs.get(e), s = {};
  for (const [o, a] of i.entries()) {
    const c = (r = t.eav.get(n)) === null || r === void 0 ? void 0 : r.get(a.id), u = X(c, 1);
    for (const l of u)
      s[o] = l[2];
  }
  return s;
}
function $t(t, e, n) {
  var r;
  return (r = t.attrIndexes.forwardIdents.get(e)) === null || r === void 0 ? void 0 : r.get(n);
}
function cr(t, e, n) {
  var r;
  return (r = t.attrIndexes.revIdents.get(e)) === null || r === void 0 ? void 0 : r.get(n);
}
function Ci(t, e) {
  var n;
  const r = t.attrIndexes.primaryKeys.get(e);
  return r || ((n = t.attrIndexes.forwardIdents.get(e)) === null || n === void 0 ? void 0 : n.get("id"));
}
function xi(t, e) {
  return hi(t, (n) => {
    e.forEach((r) => {
      Ei(n, r);
    });
  });
}
function Ai(t) {
  return typeof t == "string" && t.startsWith("?");
}
function Mi(t, e, n) {
  if (n.hasOwnProperty(t)) {
    const r = n[t];
    return ur(r, e, n);
  }
  return Object.assign(Object.assign({}, n), { [t]: e });
}
function Sn(t, e, n) {
  return t === e ? n : null;
}
function Ii(t) {
  switch (typeof t) {
    case "string":
      return t.startsWith("?") ? Mi : Sn;
    default:
      return Sn;
  }
}
const ki = [
  "in",
  "$in",
  "$not",
  "$isNull",
  "$comparator"
  // covers all of $gt, $lt, etc.
];
function Ui(t) {
  for (const e of ki)
    if (t.hasOwnProperty(e))
      return !0;
  return !1;
}
function ur(t, e, n) {
  return n ? typeof t == "object" ? Ui(t) ? n : null : Ii(t)(t, e, n) : null;
}
function $i(t, e, n) {
  return t.reduce((r, i, s) => {
    const o = e[s];
    return ur(i, o, r);
  }, n);
}
function Di(t, e, n) {
  return Ni(t, e, n).map((r) => $i(e, r, n)).filter((r) => r);
}
function Li(t, e, n) {
  return e.or ? e.or.patterns.flatMap((r) => Dt(t, r, n)) : e.and ? e.and.patterns.reduce((r, i) => Dt(t, i, r), n) : n.flatMap((r) => Di(t, e, r));
}
function Dt(t, e, n = [{}]) {
  return e.reduce((r, i) => Li(t, i, r), n);
}
function en(t, e) {
  return Array.isArray(e) ? e.map((n) => en(t, n)) : Ai(e) ? t[e] : e;
}
function qi(t, { find: e, where: n }) {
  return Dt(t, n).map((i) => en(i, e));
}
function Ni(t, e, n) {
  return Pi(t, en(n, e));
}
let Xe;
const Fi = new Uint8Array(16);
function zi() {
  if (!Xe && (Xe = typeof crypto < "u" && crypto.getRandomValues && crypto.getRandomValues.bind(crypto), !Xe))
    throw new Error("crypto.getRandomValues() not supported. See https://github.com/uuidjs/uuid#getrandomvalues-not-supported");
  return Xe(Fi);
}
const U = [];
for (let t = 0; t < 256; ++t)
  U.push((t + 256).toString(16).slice(1));
function Wi(t, e = 0) {
  return U[t[e + 0]] + U[t[e + 1]] + U[t[e + 2]] + U[t[e + 3]] + "-" + U[t[e + 4]] + U[t[e + 5]] + "-" + U[t[e + 6]] + U[t[e + 7]] + "-" + U[t[e + 8]] + U[t[e + 9]] + "-" + U[t[e + 10]] + U[t[e + 11]] + U[t[e + 12]] + U[t[e + 13]] + U[t[e + 14]] + U[t[e + 15]];
}
const Vi = typeof crypto < "u" && crypto.randomUUID && crypto.randomUUID.bind(crypto), Rn = {
  randomUUID: Vi
};
function Bi(t, e, n) {
  if (Rn.randomUUID && !e && !t)
    return Rn.randomUUID();
  t = t || {};
  const r = t.random || (t.rng || zi)();
  return r[6] = r[6] & 15 | 64, r[8] = r[8] & 63 | 128, Wi(r);
}
function En(t) {
  const e = t.replace(/-/g, ""), n = [];
  for (let r = 0; r < e.length; r += 2)
    n.push(parseInt(e.substring(r, r + 2), 16));
  return n;
}
function Qi(t, e) {
  for (let n = 0; n < t.length; n++) {
    if (t[n] < e[n])
      return -1;
    if (t[n] > e[n])
      return 1;
  }
  return 0;
}
function jn(t, e) {
  return Qi(En(t), En(e));
}
function L() {
  return Bi();
}
let Ji = 0;
function qe(t) {
  return lt(`_${t}`, Ji++);
}
function lt(t, e) {
  return `?${t}-${e}`;
}
class Pe extends Error {
  constructor(e) {
    super(e), this.name = "AttrNotFoundError";
  }
}
function lr(t, e) {
  const n = Ci(t, e);
  if (!n)
    throw new Pe(`Could not find id attr for ${e}`);
  return n;
}
function Pn(t, e, n, r) {
  return [Ki(t, e, n, r)];
}
function Ki(t, e, n, r) {
  return [
    t(n, r),
    lr(e, n).id,
    t(n, r),
    t("time", r)
  ];
}
function Yi(t, e, n) {
  return t.map((r) => r === e ? n : r);
}
function dr(t, e, n, r, i) {
  const s = $t(e, n, i), o = cr(e, n, i), a = s || o;
  if (!a)
    throw new Pe(`Could not find attr for ${[n, i]}`);
  if (a["value-type"] !== "ref")
    throw new Error(`Attr ${a.id} is not a ref`);
  const [c, u] = a["forward-identity"], [l, f] = a["reverse-identity"], p = r + 1, _ = s ? [
    t(u, r),
    a.id,
    t(f, p),
    qe("time")
  ] : [
    t(u, p),
    a.id,
    t(f, r),
    qe("time")
  ];
  return [s ? f : u, p, _, a, !!s];
}
function Hi(t, e) {
  if (typeof e != "object" || e.hasOwnProperty("$in") || e.hasOwnProperty("in"))
    return e;
  const n = t["checked-data-type"] === "date";
  return e.hasOwnProperty("$gt") ? {
    $comparator: !0,
    $op: n ? function(i) {
      return new Date(i[2]) > new Date(e.$gt);
    } : function(i) {
      return i[2] > e.$gt;
    }
  } : e.hasOwnProperty("$gte") ? {
    $comparator: !0,
    $op: n ? function(i) {
      return new Date(i[2]) >= new Date(e.$gte);
    } : function(i) {
      return i[2] >= e.$gte;
    }
  } : e.hasOwnProperty("$lt") ? {
    $comparator: !0,
    $op: n ? function(i) {
      return new Date(i[2]) < new Date(e.$lt);
    } : function(i) {
      return i[2] < e.$lt;
    }
  } : e.hasOwnProperty("$lte") ? {
    $comparator: !0,
    $op: n ? function(i) {
      return new Date(i[2]) <= new Date(e.$lte);
    } : function(i) {
      return i[2] <= e.$lte;
    }
  } : e;
}
function Gi(t, e, n, r, i, s) {
  const o = $t(e, n, i), a = cr(e, n, i), c = o || a;
  if (!c)
    throw new Pe(`No attr for etype = ${n} label = ${i}`);
  if (s != null && s.hasOwnProperty("$isNull")) {
    const u = $t(e, n, "id");
    if (!u)
      throw new Pe(`No attr for etype = ${n} label = id`);
    return [
      t(n, r),
      u.id,
      { $isNull: { attrId: c.id, isNull: s.$isNull, reverse: !o } },
      qe("time")
    ];
  }
  return o ? [
    t(n, r),
    c.id,
    Hi(c, s),
    qe("time")
  ] : [s, c.id, t(n, r), qe("time")];
}
function Xi(t, e, n, r, i) {
  const [s, o, a] = i.reduce((c, u) => {
    const [l, f, p] = c, [_, g, m] = dr(t, e, l, f, u);
    return [_, g, [...p, m]];
  }, [n, r, []]);
  return [s, o, a];
}
function Lt(t, e, n, r, i, s) {
  const o = i.slice(0, i.length - 1), a = i[i.length - 1], [c, u, l] = Xi(t, e, n, r, o), f = Gi(t, e, c, u, a, s);
  return l.concat([f]);
}
function Zi(t, e) {
  return e ? [e].concat(t) : t;
}
function es([t, e]) {
  return t === "or" && Array.isArray(e);
}
function ts([t, e]) {
  return t === "and" && Array.isArray(e);
}
function ns(t, e, n) {
  return (r, i) => r == e ? t(r, i) : `${t(r, i)}-${n}`;
}
function Tn(t, e, n, r, i, s) {
  const o = s.map((c, u) => {
    const l = ns(t, r, u);
    return fr(l, n, r, i, c);
  }), a = t(r, i);
  return { [e]: { patterns: o, joinSym: a } };
}
function rs(t) {
  const e = [];
  for (let n = 1; n <= t.length; n++)
    e.push(t.slice(0, n));
  return e;
}
function Cn(t, e, n, r, i) {
  return rs(i).map((s) => Lt(t, e, n, r, s, { $isNull: !0 }));
}
function fr(t, e, n, r, i) {
  return Object.entries(i).flatMap(([s, o]) => {
    if (es([s, o]))
      return Tn(t, "or", e, n, r, o);
    if (ts([s, o]))
      return Tn(t, "and", e, n, r, o);
    const a = s.split(".");
    if (o != null && o.hasOwnProperty("$not")) {
      const c = Lt(t, e, n, r, a, o), u = Cn(t, e, n, r, a);
      return [
        {
          or: {
            patterns: [c, ...u],
            joinSym: t(n, r)
          }
        }
      ];
    }
    return o != null && o.hasOwnProperty("$isNull") && o.$isNull === !0 && a.length > 1 ? [
      {
        or: {
          patterns: Cn(t, e, n, r, a),
          joinSym: t(n, r)
        }
      }
    ] : Lt(t, e, n, r, a, o);
  });
}
function is(t, e, n, r) {
  const i = lt;
  return r ? fr(i, t, e, n, r).concat(Pn(i, t, e, n)) : Pn(i, t, e, n);
}
function ss(t, e, n) {
  return [t(e, n), t("time", n)];
}
function os(t, e, n, r, i, s) {
  const [o, a, c, u, l] = dr(t, e, n, r, i), f = Yi(c, t(n, r), s);
  return [o, a, f, u, l];
}
function as(t, e, { etype: n, level: r, form: i }, s) {
  const o = Object.keys(i).filter((a) => a !== "$");
  return o.length ? Object.entries(s).map(function([c, u]) {
    return o.map(function(p) {
      var _, g, m;
      const O = !!(e.cardinalityInference && (!((m = (g = (_ = e.linkIndex) === null || _ === void 0 ? void 0 : _[n]) === null || g === void 0 ? void 0 : g[p]) === null || m === void 0) && m.isSingular));
      try {
        const [S, b, y] = os(t, e, n, r, p, c), x = hr(e, {
          etype: S,
          level: b,
          form: i[p],
          join: y
        }), R = O ? x[0] : x;
        return { [p]: R };
      } catch (S) {
        if (S instanceof Pe)
          return { [p]: O ? void 0 : [] };
        throw S;
      }
    }).reduce(function(p, _) {
      return Object.assign(Object.assign({}, p), _);
    }, u);
  }) : Object.values(s);
}
function xn(t, e) {
  switch (t) {
    case "asc":
      switch (e) {
        case "number":
          return (n, r) => n < r;
        case "uuid":
          return (n, r) => jn(n, r) === -1;
      }
    case "desc":
      switch (e) {
        case "number":
          return (n, r) => n > r;
        case "uuid":
          return (n, r) => jn(n, r) === 1;
      }
  }
}
function cs(t, e, [n, r, i, s]) {
  return xn(e, "number")(s, t[3]) || s === t[3] && xn(e, "uuid")(n, t[0]);
}
function us(t, e, n, r, i) {
  const s = lr(t, e).id, o = qi(t, i).sort(function([l, f], [p, _]) {
    return n === "desc" ? _ - f : f - _;
  });
  let a = {};
  const c = r == null ? void 0 : r["start-cursor"];
  for (const [u, l] of o) {
    if (c && s === c[1] && cs(c, n, [u, s, u, l]))
      continue;
    const f = Ti(t, e, u);
    f && (a[u] = f);
  }
  return a;
}
function ls(t) {
  var e;
  const n = (e = t.$) === null || e === void 0 ? void 0 : e.order;
  return n && n[Object.keys(n)[0]] || "asc";
}
function ds(t, { etype: e, level: n, form: r, join: i, pageInfo: s }) {
  var o, a, c, u, l, f, p;
  const _ = ((o = r.$) === null || o === void 0 ? void 0 : o.limit) || ((a = r.$) === null || a === void 0 ? void 0 : a.first) || ((c = r.$) === null || c === void 0 ? void 0 : c.last), g = (u = r.$) === null || u === void 0 ? void 0 : u.offset, m = (l = r.$) === null || l === void 0 ? void 0 : l.before, O = (f = r.$) === null || f === void 0 ? void 0 : f.after;
  if ((g || m || O) && (!s || !s["start-cursor"]))
    return [];
  const S = Zi(is(t, e, n, (p = r.$) === null || p === void 0 ? void 0 : p.where), i), b = ss(lt, e, n), y = us(t, e, ls(r), s, { where: S, find: b });
  if (_ != null) {
    const x = Object.entries(y);
    return x.length <= _ ? y : Object.fromEntries(x.slice(0, _));
  }
  return y;
}
function fs(t, e) {
  try {
    return ds(t, e);
  } catch (n) {
    if (n instanceof Pe)
      return {};
    throw n;
  }
}
function hr(t, e) {
  const n = fs(t, e);
  return as(lt, t, e, n);
}
function hs(t) {
  const e = {};
  for (const [n, r] of Object.entries(t))
    e[n] = {
      startCursor: r["start-cursor"],
      endCursor: r["end-cursor"],
      hasNextPage: r["has-next-page?"],
      hasPreviousPage: r["has-previous-page?"]
    };
  return e;
}
function ps({ store: t, pageInfo: e, aggregate: n }, r) {
  const s = { data: Object.keys(r).reduce(function(a, c) {
    return n != null && n[c] || (a[c] = hr(t, {
      etype: c,
      form: r[c],
      level: 0,
      pageInfo: e == null ? void 0 : e[c]
    })), a;
  }, {}) };
  return e && (s.pageInfo = hs(e)), n && (s.aggregate = n), s;
}
function qt(t, e, n) {
  return new Proxy({}, {
    get: (r, i) => i === "__ops" ? n : (s) => qt(t, e, [
      ...n,
      [i, t, e, s]
    ])
  });
}
function Jo(t, e) {
  return `lookup__${t}__${JSON.stringify(e)}`;
}
function Nt(t) {
  return t.startsWith("lookup__");
}
function pr(t) {
  const [e, n, ...r] = t.split("__");
  return [n, JSON.parse(r.join("__"))];
}
function vs(t) {
  return new Proxy({}, {
    get(e, n) {
      return Nt(n) ? qt(t, pr(n), []) : qt(t, n, []);
    }
  });
}
function We() {
  return new Proxy({}, {
    get(t, e) {
      return vs(e);
    }
  });
}
const Ko = We();
function _s(t) {
  return t.__ops;
}
function bs(t, e) {
  const { attrIdMap: n, refSwapAttrIds: r } = t, i = [];
  for (const o of e) {
    const a = n[o];
    if (a)
      i.push(a);
    else if (Array.isArray(o) && o.length == 2 && n[o[0]]) {
      const [c, u] = o;
      i.push([n[c], u]);
    } else
      i.push(o);
  }
  const [s] = e;
  if ((s === "add-triple" || s === "retract-triple") && r.has(e[2])) {
    const o = i[1];
    i[1] = i[3], i[3] = o;
  }
  return i;
}
function Q(t, e, n) {
  return Object.values(t).find((r) => {
    const [i, s, o] = r["forward-identity"];
    return s === e && o === n;
  });
}
function we(t, e, n) {
  return Object.values(t).find((r) => {
    const i = r["reverse-identity"];
    if (!i)
      return !1;
    const [s, o, a] = i;
    return o === e && a === n;
  });
}
function ys(t) {
  if (Array.isArray(t))
    return t;
  const e = Object.entries(t);
  if (e.length !== 1)
    throw new Error("lookup must be an object with a single unique attr and value.");
  return e[0];
}
function Ft(t, e, n) {
  return n.indexOf(".") !== -1 && // attr names can have `.` in them, so use the attr we find with a `.`
  // before assuming it's a ref lookup.
  !Q(t, e, n);
}
function zt(t) {
  const [e, n, ...r] = t.split(".");
  if (r.length > 0 || n !== "id")
    throw new Error(`${t} is not a valid lookup attribute.`);
  return e;
}
function gs(t, e, n) {
  if (!Ft(t, e, n))
    return Q(t, e, n);
  const r = zt(n), i = Q(t, e, r) || we(t, e, r);
  if (i && i["value-type"] !== "ref")
    throw new Error(`${n} does not reference a valid link attribute.`);
  return i;
}
function Wt(t) {
  return typeof t == "string" && !Nt(t) ? null : typeof t == "string" && Nt(t) ? pr(t) : ys(t);
}
function Z(t, e, n) {
  const r = Wt(n);
  if (r === null)
    return n;
  const [i, s] = r, o = gs(t, e, i);
  if (!o || !o["unique?"])
    throw new Error(`${i} is not a unique attribute.`);
  return [o.id, s];
}
function ws(t, [e, n, r]) {
  return Object.entries(r).flatMap(([s, o]) => {
    const a = Array.isArray(o) ? o : [o], c = Q(t, e, s), u = we(t, e, s);
    return a.map((l) => c ? [
      "add-triple",
      Z(t, e, n),
      c.id,
      Z(t, c["reverse-identity"][1], l)
    ] : [
      "add-triple",
      Z(t, u["forward-identity"][1], l),
      u.id,
      Z(t, e, n)
    ]);
  });
}
function ms(t, [e, n, r]) {
  return Object.entries(r).flatMap(([s, o]) => {
    const a = Array.isArray(o) ? o : [o], c = Q(t, e, s), u = we(t, e, s);
    return a.map((l) => c ? [
      "retract-triple",
      Z(t, e, n),
      c.id,
      Z(t, c["reverse-identity"][1], l)
    ] : [
      "retract-triple",
      Z(t, u["forward-identity"][1], l),
      u.id,
      Z(t, e, n)
    ]);
  });
}
function Os(t, [e, n, r]) {
  const i = Z(t, e, n);
  return [["id", Z(t, e, n)]].concat(Object.entries(r)).map(([o, a]) => {
    const c = Q(t, e, o);
    return ["add-triple", i, c.id, a];
  });
}
function Ss(t, [e, n]) {
  return [["delete-entity", Z(t, e, n), e]];
}
function Rs(t, [e, n, r]) {
  const i = Z(t, e, n), s = Object.entries(r).map(([a, c]) => {
    const u = Q(t, e, a), l = tr(c, void 0, null);
    return ["deep-merge-triple", i, u.id, l];
  });
  return [[
    "add-triple",
    i,
    Q(t, e, "id").id,
    i
  ]].concat(s);
}
function Es(t) {
  const [e, n, r, i] = t;
  if (!i)
    return t;
  const s = Object.assign({}, i);
  return delete s.id, [e, n, r, s];
}
function js(t, e) {
  const [n, ...r] = Es(e);
  switch (n) {
    case "merge":
      return Rs(t, r);
    case "update":
      return Os(t, r);
    case "link":
      return ws(t, r);
    case "unlink":
      return ms(t, r);
    case "delete":
      return Ss(t, r);
    default:
      throw new Error(`unsupported action ${n}`);
  }
}
function Ps(t) {
  switch (t) {
    case "string":
    case "date":
    case "boolean":
    case "number":
      return t;
    default:
      return;
  }
}
function Ts(t, e, n) {
  var r, i;
  const s = (i = (r = t.entities[e]) === null || r === void 0 ? void 0 : r.attrs) === null || i === void 0 ? void 0 : i[n];
  if (n === "id")
    return null;
  if (!s)
    throw new Error(`${e}.${n} does not exist in your schema`);
  const { unique: o, indexed: a } = s == null ? void 0 : s.config, c = Ps(s == null ? void 0 : s.valueType);
  return {
    "index?": a,
    "unique?": o,
    "checked-data-type": c
  };
}
function jt(t, e, n, r) {
  const i = t ? Ts(t, e, n) : null, s = L(), a = [L(), e, n];
  return Object.assign(Object.assign({ id: s, "forward-identity": a, "value-type": "blob", cardinality: "one", "unique?": !1, "index?": !1, isUnsynced: !0 }, i || {}), r || {});
}
function Cs(t, e, n) {
  return Object.values(t.links).find((i) => i.forward.on === e && i.forward.label === n || i.reverse.on === e && i.reverse.label === n);
}
function xs(t, e, n) {
  const r = Cs(t, e, n);
  if (!r)
    throw new Error(`Couldn't find the link ${e}.${n} in your schema`);
  const { forward: i, reverse: s } = r;
  return {
    "forward-identity": [L(), i.on, i.label],
    "reverse-identity": [L(), s.on, s.label],
    cardinality: i.has === "one" ? "one" : "many",
    "unique?": s.has === "one"
  };
}
function An(t, e, n, r) {
  const i = t ? xs(t, e, n) : null, s = L(), o = [L(), e, n], a = [L(), n, e];
  return Object.assign(Object.assign({ id: s, "forward-identity": o, "reverse-identity": a, "value-type": "ref", cardinality: "many", "unique?": !1, "index?": !1, isUnsynced: !0 }, i || {}), r || {});
}
const As = /* @__PURE__ */ new Set(["update", "merge", "link", "unlink"]), Ms = /* @__PURE__ */ new Set(["link", "unlink"]), Is = /* @__PURE__ */ new Set(["update", "merge"]), ks = /* @__PURE__ */ new Set([
  "link",
  "unlink",
  "update",
  "merge",
  "delete"
]), Vt = { "unique?": !0, "index?": !0 }, Us = Object.assign(Object.assign({}, Vt), { cardinality: "one" });
function $s(t) {
  const e = [], [n, r, i, s] = t;
  if (!ks.has(n))
    return e;
  const o = Wt(i);
  if (o && e.push({ etype: r, lookupPair: o }), n === "link")
    for (const [a, c] of Object.entries(s)) {
      const u = Array.isArray(c) ? c : [c];
      for (const l of u) {
        const f = Wt(l);
        f && e.push({
          etype: r,
          lookupPair: f,
          linkLabel: a
        });
      }
    }
  return e;
}
function Ds({ attrs: t, schema: e }, n) {
  var r, i;
  const [s, o, a] = [/* @__PURE__ */ new Set(), Object.assign({}, t), []];
  function c(f) {
    o[f.id] = f, a.push(["add-attr", f]), s.add(f.id);
  }
  function u(f) {
    f != null && f.isUnsynced && !s.has(f.id) && (a.push(["add-attr", f]), s.add(f.id));
  }
  function l(f, p) {
    const _ = Q(o, f, p), g = we(o, f, p);
    u(_), u(g), !_ && !g && c(An(e, f, p, Us));
  }
  for (const f of n)
    for (const { etype: p, lookupPair: _, linkLabel: g } of $s(f)) {
      const m = _[0];
      if (g) {
        l(p, g);
        const O = Q(o, p, g), S = we(o, p, g);
        u(O), u(S);
        const b = ((r = O == null ? void 0 : O["reverse-identity"]) === null || r === void 0 ? void 0 : r[1]) || ((i = S == null ? void 0 : S["forward-identity"]) === null || i === void 0 ? void 0 : i[1]) || g;
        if (Ft(o, b, m))
          l(b, zt(m));
        else {
          const y = Q(o, b, m);
          y || c(jt(e, b, m, Vt)), u(y);
        }
      } else if (Ft(o, p, m))
        l(p, zt(m));
      else {
        const O = Q(o, p, m);
        O || c(jt(e, p, m, Vt)), u(O);
      }
    }
  for (const f of n) {
    const [p, _, g, m] = f;
    if (As.has(p)) {
      const O = Object.keys(m);
      O.push("id");
      for (const S of O) {
        const b = Q(o, _, S);
        if (u(b), Is.has(p) && (b || c(jt(e, _, S, S === "id" ? { "unique?": !0 } : null))), Ms.has(p)) {
          const y = we(o, _, S);
          !b && !y && c(An(e, _, S)), u(y);
        }
      }
    }
  }
  return [o, a];
}
function Ls(t, e) {
  const r = (Array.isArray(e) ? e : [e]).flatMap((a) => _s(a)), [i, s] = Ds(t, r), o = r.flatMap((a) => js(i, a));
  return [...s, ...o];
}
var Mn = function(t, e, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function a(l) {
      try {
        u(r.next(l));
      } catch (f) {
        o(f);
      }
    }
    function c(l) {
      try {
        u(r.throw(l));
      } catch (f) {
        o(f);
      }
    }
    function u(l) {
      l.done ? s(l.value) : i(l.value).then(a, c);
    }
    u((r = r.apply(t, e || [])).next());
  });
};
class tn {
  constructor(e) {
    this.dbName = e, this._storeName = "kv", this._dbPromise = this._init();
  }
  _init() {
    return new Promise((e, n) => {
      const r = indexedDB.open(this.dbName, 1);
      r.onerror = (i) => {
        n(i);
      }, r.onsuccess = (i) => {
        e(i.target.result);
      }, r.onupgradeneeded = (i) => {
        i.target.result.createObjectStore(this._storeName);
      };
    });
  }
  getItem(e) {
    return Mn(this, void 0, void 0, function* () {
      const n = yield this._dbPromise;
      return new Promise((r, i) => {
        const a = n.transaction([this._storeName], "readonly").objectStore(this._storeName).get(e);
        a.onerror = (c) => {
          i(c);
        }, a.onsuccess = (c) => {
          a.result ? r(a.result) : r(null);
        };
      });
    });
  }
  setItem(e, n) {
    return Mn(this, void 0, void 0, function* () {
      const r = yield this._dbPromise;
      return new Promise((i, s) => {
        const c = r.transaction([this._storeName], "readwrite").objectStore(this._storeName).put(n, e);
        c.onerror = (u) => {
          s(u);
        }, c.onsuccess = (u) => {
          i();
        };
      });
    });
  }
}
var qs = function(t, e, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function a(l) {
      try {
        u(r.next(l));
      } catch (f) {
        o(f);
      }
    }
    function c(l) {
      try {
        u(r.throw(l));
      } catch (f) {
        o(f);
      }
    }
    function u(l) {
      l.done ? s(l.value) : i(l.value).then(a, c);
    }
    u((r = r.apply(t, e || [])).next());
  });
};
class nn {
  static getIsOnline() {
    return qs(this, void 0, void 0, function* () {
      return navigator.onLine;
    });
  }
  static listen(e) {
    const n = () => {
      e(!0);
    }, r = () => {
      e(!1);
    };
    return addEventListener("online", n), addEventListener("offline", r), () => {
      removeEventListener("online", n), removeEventListener("offline", r);
    };
  }
}
var Ns = function(t, e, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function a(l) {
      try {
        u(r.next(l));
      } catch (f) {
        o(f);
      }
    }
    function c(l) {
      try {
        u(r.throw(l));
      } catch (f) {
        o(f);
      }
    }
    function u(l) {
      l.done ? s(l.value) : i(l.value).then(a, c);
    }
    u((r = r.apply(t, e || [])).next());
  });
};
function he(t, e) {
  return Ns(this, void 0, void 0, function* () {
    const n = yield fetch(t, e), r = yield n.json();
    return n.status === 200 ? Promise.resolve(r) : Promise.reject({ status: n.status, body: r });
  });
}
var Ve = function(t, e, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function a(l) {
      try {
        u(r.next(l));
      } catch (f) {
        o(f);
      }
    }
    function c(l) {
      try {
        u(r.throw(l));
      } catch (f) {
        o(f);
      }
    }
    function u(l) {
      l.done ? s(l.value) : i(l.value).then(a, c);
    }
    u((r = r.apply(t, e || [])).next());
  });
};
function Fs({ apiURI: t, appId: e, email: n }) {
  return he(`${t}/runtime/auth/send_magic_code`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ "app-id": e, email: n })
  });
}
function zs(t) {
  return Ve(this, arguments, void 0, function* ({ apiURI: e, appId: n, email: r, code: i }) {
    return yield he(`${e}/runtime/auth/verify_magic_code`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ "app-id": n, email: r, code: i })
    });
  });
}
function Ws(t) {
  return Ve(this, arguments, void 0, function* ({ apiURI: e, appId: n, refreshToken: r }) {
    return yield he(`${e}/runtime/auth/verify_refresh_token`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        "app-id": n,
        "refresh-token": r
      })
    });
  });
}
function In(t) {
  return Ve(this, arguments, void 0, function* ({ apiURI: e, appId: n, code: r, codeVerifier: i }) {
    return yield he(`${e}/runtime/oauth/token`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        app_id: n,
        code: r,
        code_verifier: i
      })
    });
  });
}
function Vs(t) {
  return Ve(this, arguments, void 0, function* ({ apiURI: e, appId: n, nonce: r, idToken: i, clientName: s, refreshToken: o }) {
    return yield he(`${e}/runtime/oauth/id_token`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        app_id: n,
        nonce: r,
        id_token: i,
        client_name: s,
        refresh_token: o
      })
    });
  });
}
function Bs(t) {
  return Ve(this, arguments, void 0, function* ({ apiURI: e, appId: n, refreshToken: r }) {
    return yield he(`${e}/runtime/signout`, {
      method: "POST",
      headers: {
        "content-type": "application/json"
      },
      body: JSON.stringify({
        app_id: n,
        refresh_token: r
      })
    });
  });
}
var dt = function(t, e, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function a(l) {
      try {
        u(r.next(l));
      } catch (f) {
        o(f);
      }
    }
    function c(l) {
      try {
        u(r.throw(l));
      } catch (f) {
        o(f);
      }
    }
    function u(l) {
      l.done ? s(l.value) : i(l.value).then(a, c);
    }
    u((r = r.apply(t, e || [])).next());
  });
};
function Qs(t) {
  return dt(this, arguments, void 0, function* ({ apiURI: e, appId: n, fileName: r, refreshToken: i, metadata: s = {} }) {
    const { data: o } = yield he(`${e}/storage/signed-upload-url`, {
      method: "POST",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${i}`
      },
      body: JSON.stringify({
        app_id: n,
        filename: r
      })
    });
    return o;
  });
}
function Js(t, e) {
  return dt(this, void 0, void 0, function* () {
    return (yield fetch(t, {
      method: "PUT",
      body: e,
      headers: {
        "Content-Type": e.type
      }
    })).ok;
  });
}
function Ks(t) {
  return dt(this, arguments, void 0, function* ({ apiURI: e, appId: n, path: r, refreshToken: i }) {
    const { data: s } = yield he(`${e}/storage/signed-download-url?app_id=${n}&filename=${encodeURIComponent(r)}`, {
      method: "GET",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${i}`
      }
    });
    return s;
  });
}
function Ys(t) {
  return dt(this, arguments, void 0, function* ({ apiURI: e, appId: n, path: r, refreshToken: i }) {
    const { data: s } = yield he(`${e}/storage/files?app_id=${n}&filename=${encodeURIComponent(r)}`, {
      method: "DELETE",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${i}`
      }
    });
    return s;
  });
}
function kn(t, e) {
  if (!e)
    return t;
  const n = {};
  return e.forEach((r) => {
    n[r] = t[r];
  }), n;
}
function Hs(t, e, n) {
  var r, i;
  const s = {
    peers: {}
  };
  if (e && "user" in e ? e.user : !0) {
    const a = kn((r = t.user) !== null && r !== void 0 ? r : {}, e == null ? void 0 : e.keys);
    s.user = Object.assign(Object.assign({}, a), { peerId: n });
  }
  for (const a of Object.keys((i = t.peers) !== null && i !== void 0 ? i : {})) {
    const c = (e == null ? void 0 : e.peers) === void 0, u = Array.isArray(e == null ? void 0 : e.peers) && (e == null ? void 0 : e.peers.includes(a));
    if (c || u) {
      const l = kn(t.peers[a], e == null ? void 0 : e.keys);
      s.peers[a] = Object.assign(Object.assign({}, l), { peerId: a });
    }
  }
  return s;
}
function Gs(t, e) {
  if (t.isLoading !== e.isLoading || t.error !== e.error || (t.user || e.user) && (!t.user || !e.user || !On(t.user, e.user)) || !Zn(t.peers, e.peers))
    return !0;
  for (const r of Object.keys(t.peers))
    if (!On(t.peers[r], e.peers[r]))
      return !0;
  return !1;
}
class Un {
  constructor() {
    this.promise = new Promise((e, n) => {
      this._resolve = e, this._reject = n;
    });
  }
  resolve(e) {
    this._resolve(e);
  }
  reject(e) {
    this._reject(e);
  }
}
var Ze = function(t, e, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function a(l) {
      try {
        u(r.next(l));
      } catch (f) {
        o(f);
      }
    }
    function c(l) {
      try {
        u(r.throw(l));
      } catch (f) {
        o(f);
      }
    }
    function u(l) {
      l.done ? s(l.value) : i(l.value).then(a, c);
    }
    u((r = r.apply(t, e || [])).next());
  });
};
class $n {
  constructor(e, n, r, i, s = (c) => JSON.stringify(c), o = (c) => JSON.parse(c), a = 100) {
    this._persister = e, this._key = n, this._onMerge = i, this._loadedCbs = [], this._isLoading = !0, this.currentValue = r, this.toJSON = s, this.fromJSON = o, this._saveThrottleMs = a, this._pendingSaveCbs = [], this._version = 0, this._load();
  }
  _load() {
    return Ze(this, void 0, void 0, function* () {
      const e = this.fromJSON(yield this._persister.getItem(this._key));
      this._isLoading = !1, this._onMerge(e, this.currentValue);
      for (const n of this._loadedCbs)
        n();
    });
  }
  waitForLoaded() {
    return Ze(this, void 0, void 0, function* () {
      if (!this._isLoading)
        return;
      yield new Promise((n) => {
        this._loadedCbs.push(n);
      });
    });
  }
  isLoading() {
    return this._isLoading;
  }
  version() {
    return this._version;
  }
  waitForSync() {
    return Ze(this, void 0, void 0, function* () {
      if (!this._nextSave)
        return;
      yield new Promise((n) => {
        this._pendingSaveCbs.push(n);
      });
    });
  }
  _writeToStorage() {
    this._persister.setItem(this._key, this.toJSON(this.currentValue));
    for (const e of this._pendingSaveCbs)
      e();
    this._pendingSaveCbs.length = 0;
  }
  flush() {
    return Ze(this, void 0, void 0, function* () {
      this._nextSave && (clearTimeout(this._nextSave), this._writeToStorage());
    });
  }
  _enqueuePersist(e) {
    if (this._nextSave) {
      e && this._pendingSaveCbs.push(e);
      return;
    }
    this._nextSave = setTimeout(() => {
      this._nextSave = null, this._writeToStorage();
    }, this._saveThrottleMs);
  }
  set(e, n) {
    this._version++, this.currentValue = e(this.currentValue), this._isLoading ? this._loadedCbs.push(() => this._enqueuePersist(n)) : this._enqueuePersist(n);
  }
}
function vr(t, e = []) {
  t.forEach((n) => {
    const { data: r } = n, { "datalog-result": i } = r, { "join-rows": s } = i;
    for (const o of s)
      for (const a of o)
        e.push(a);
    vr(n["child-nodes"], e);
  });
}
function Dn(t) {
  const e = [];
  return vr(t, e), e;
}
function Xs(t) {
  return Object.values(t.links).reduce((e, n) => {
    var r, i, s, o;
    return (r = e[s = n.forward.on]) !== null && r !== void 0 || (e[s] = {}), e[n.forward.on][n.forward.label] = {
      isForward: !0,
      isSingular: n.forward.has === "one",
      link: n
    }, (i = e[o = n.reverse.on]) !== null && i !== void 0 || (e[o] = {}), e[n.reverse.on][n.reverse.label] = {
      isForward: !1,
      isSingular: n.reverse.has === "one",
      link: n
    }, e;
  }, {});
}
const Zs = "v0.15.9";
var D = function(t, e, n, r) {
  function i(s) {
    return s instanceof n ? s : new n(function(o) {
      o(s);
    });
  }
  return new (n || (n = Promise))(function(s, o) {
    function a(l) {
      try {
        u(r.next(l));
      } catch (f) {
        o(f);
      }
    }
    function c(l) {
      try {
        u(r.throw(l));
      } catch (f) {
        o(f);
      }
    }
    function u(l) {
      l.done ? s(l.value) : i(l.value).then(a, c);
    }
    u((r = r.apply(t, e || [])).next());
  });
};
const ye = {
  CONNECTING: "connecting",
  OPENED: "opened",
  AUTHENTICATED: "authenticated",
  CLOSED: "closed",
  ERRORED: "errored"
}, eo = 3e4, to = 0, Ln = 1, no = {
  apiURI: "https://api.instantdb.com",
  websocketURI: "wss://api.instantdb.com/runtime/session"
}, Pt = "_instant_oauth_redirect", Tt = "currentUser";
let ro = 0;
function io(t) {
  const e = new WebSocket(t);
  return e._id = ro++, e;
}
function so() {
  return typeof window < "u" || typeof chrome < "u";
}
const qn = {
  "set-presence": !0,
  "set-presence-ok": !0,
  "refresh-presence": !0
};
function oo(t) {
  var e;
  const n = JSON.parse(t);
  for (const r in n) {
    const i = n[r];
    !((e = i == null ? void 0 : i.result) === null || e === void 0) && e.store && (i.result.store = _i(i.result.store));
  }
  return n;
}
function ao(t) {
  var e;
  const n = {};
  for (const r in t) {
    const i = t[r], s = Object.assign({}, i);
    !((e = i.result) === null || e === void 0) && e.store && (s.result = Object.assign(Object.assign({}, i.result), { store: vi(i.result.store) })), n[r] = s;
  }
  return JSON.stringify(n);
}
class _r {
  constructor(e, n = tn, r = nn, i) {
    this._isOnline = !0, this._isShutdown = !1, this.status = ye.CONNECTING, this.queryCbs = {}, this.queryOnceDfds = {}, this.authCbs = [], this.attrsCbs = [], this.mutationErrorCbs = [], this.mutationDeferredStore = /* @__PURE__ */ new Map(), this._reconnectTimeoutId = null, this._reconnectTimeoutMs = 0, this._localIdPromises = {}, this._errorMessage = null, this._oauthCallbackResponse = null, this._linkIndex = null, this._rooms = {}, this._roomsPendingLeave = {}, this._presence = {}, this._broadcastQueue = [], this._broadcastSubs = {}, this._currentUserCached = { isLoading: !0, error: void 0, user: void 0 }, this._beforeUnloadCbs = [], this._dataForQueryCache = {}, this._onMergeQuerySubs = (s, o) => {
      const a = s || {}, c = Object.assign({}, o);
      Object.entries(o).forEach(([l, f]) => {
        var p;
        const _ = (p = a == null ? void 0 : a[l]) === null || p === void 0 ? void 0 : p.result, g = f.result;
        _ && !g && (c[l].result = _);
      }), Object.keys(a).filter((l) => !o[l]).slice(0, 10).forEach((l) => {
        c[l] = a[l];
      }), this.querySubs.set((l) => c), this.loadedNotifyAll();
    }, this._onMergePendingMutations = (s, o) => {
      const a = new Map([...s.entries(), ...o.entries()]);
      this.pendingMutations.set((u) => a), this.loadedNotifyAll(), this._rewriteMutations(this.attrs, s).forEach((u, l) => {
        !o.has(l) && !u["tx-id"] && this._sendMutation(l, u);
      });
    }, this.getPreviousResult = (s) => {
      const o = V(s);
      return this.dataForQuery(o);
    }, this.notifyOne = (s) => {
      var o, a;
      const c = (o = this.queryCbs[s]) !== null && o !== void 0 ? o : [], u = (a = this._dataForQueryCache[s]) === null || a === void 0 ? void 0 : a.data, l = this.dataForQuery(s);
      l && (st(l, u) || c.forEach((f) => f.cb(l)));
    }, this.notifyOneQueryOnce = (s) => {
      var o;
      const a = (o = this.queryOnceDfds[s]) !== null && o !== void 0 ? o : [], c = this.dataForQuery(s);
      a.forEach((u) => {
        this._completeQueryOnce(u.q, s, u.dfd), u.dfd.resolve(c);
      });
    }, this.notifyQueryError = (s, o) => {
      (this.queryCbs[s] || []).forEach((c) => c.cb({ error: o }));
    }, this.pushTx = (s) => {
      try {
        const o = Ls({ attrs: this.optimisticAttrs(), schema: this.config.schema }, s);
        return this.pushOps(o);
      } catch (o) {
        return this.pushOps([], o);
      }
    }, this.pushOps = (s, o) => {
      const a = L(), c = {
        op: "transact",
        "tx-steps": s,
        error: o
      };
      this.pendingMutations.set((l) => (l.set(a, c), l));
      const u = new Un();
      return this.mutationDeferredStore.set(a, u), this._sendMutation(a, c), this.notifyAll(), u.promise;
    }, this._wsOnOpen = (s) => {
      const o = s.target;
      if (this._ws !== o) {
        N.info("[socket][open]", o._id, "skip; this is no longer the current ws");
        return;
      }
      N.info("[socket][open]", this._ws._id), this._setStatus(ye.OPENED), this.getCurrentUser().then((a) => {
        var c;
        this._trySend(L(), {
          op: "init",
          "app-id": this.config.appId,
          "refresh-token": (c = a.user) === null || c === void 0 ? void 0 : c.refresh_token,
          versions: this.versions,
          // If an admin token is provided for an app, we will
          // skip all permission checks. This is an advanced feature,
          // to let users write internal tools
          // This option is not exposed in `Config`, as it's
          // not ready for prme time
          "__admin-token": this.config.__adminToken
        });
      });
    }, this._wsOnMessage = (s) => {
      const o = s.target, a = JSON.parse(s.data.toString());
      if (this._ws !== o) {
        N.info("[socket][message]", o._id, a, "skip; this is no longer the current ws");
        return;
      }
      this._handleReceive(o._id, JSON.parse(s.data.toString()));
    }, this._wsOnError = (s) => {
      const o = s.target;
      if (this._ws !== o) {
        N.info("[socket][error]", o._id, "skip; this is no longer the current ws");
        return;
      }
      N.error("[socket][error]", o._id, s);
    }, this._wsOnClose = (s) => {
      const o = s.target;
      if (this._ws !== o) {
        N.info("[socket][close]", o._id, "skip; this is no longer the current ws");
        return;
      }
      this._setStatus(ye.CLOSED);
      for (const a of Object.values(this._rooms))
        a.isConnected = !1;
      if (this._isShutdown) {
        N.info("[socket][close]", o._id, "Reactor has been shut down and will not reconnect");
        return;
      }
      N.info("[socket][close]", o._id, "schedule reconnect, ms =", this._reconnectTimeoutMs), setTimeout(() => {
        if (this._reconnectTimeoutMs = Math.min(this._reconnectTimeoutMs + 1e3, 1e4), !this._isOnline) {
          N.info("[socket][close]", o._id, "we are offline, no need to start socket");
          return;
        }
        this._startSocket();
      }, this._reconnectTimeoutMs);
    }, this.config = Object.assign(Object.assign({}, no), e), this.versions = Object.assign(Object.assign({}, i || {}), { "@instantdb/core": Zs }), this.config.schema && (this._linkIndex = Xs(this.config.schema)), so() && (typeof BroadcastChannel == "function" && (this._broadcastChannel = new BroadcastChannel("@instantdb"), this._broadcastChannel.addEventListener("message", (s) => D(this, void 0, void 0, function* () {
      var o;
      if (((o = s.data) === null || o === void 0 ? void 0 : o.type) === "auth") {
        const a = yield this.getCurrentUser();
        this.updateUser(a.user);
      }
    }))), this._oauthCallbackResponse = this._oauthLoginInit(), this._initStorage(n), this.getCurrentUser(), r.getIsOnline().then((s) => {
      this._isOnline = s, this._startSocket(), r.listen((o) => {
        o !== this._isOnline && (N.info("[network] online =", o), this._isOnline = o, this._isOnline && this._startSocket());
      });
    }), typeof addEventListener < "u" && (this._beforeUnload = this._beforeUnload.bind(this), addEventListener("beforeunload", this._beforeUnload)));
  }
  _initStorage(e) {
    this._persister = new e(`instant_${this.config.appId}_5`), this.querySubs = new $n(this._persister, "querySubs", {}, this._onMergeQuerySubs, ao, oo), this.pendingMutations = new $n(this._persister, "pendingMutations", /* @__PURE__ */ new Map(), this._onMergePendingMutations, (n) => JSON.stringify([...n.entries()]), (n) => new Map(JSON.parse(n))), this._beforeUnloadCbs.push(() => {
      this.pendingMutations.flush(), this.querySubs.flush();
    });
  }
  _beforeUnload() {
    for (const e of this._beforeUnloadCbs)
      e();
  }
  /**
   * @param {'enqueued' | 'pending' | 'synced' | 'timeout' |  'error' } status
   * @param string clientId
   * @param {{message?: string, hint?: string, error?: Error}} [errDetails]
   */
  _finishTransaction(e, n, r) {
    const i = this.mutationDeferredStore.get(n);
    this.mutationDeferredStore.delete(n);
    const s = e !== "error" && e !== "timeout";
    !i && !s && console.error("Mutation failed", Object.assign({ status: e, clientId: n }, r)), i && (s ? i.resolve({ status: e, clientId: n }) : i.reject(Object.assign({ status: e, clientId: n }, r)));
  }
  _setStatus(e, n) {
    this.status = e, this._errorMessage = n;
  }
  /**
   * On refresh we clear out pending mutations that we know have been applied
   * by the server and thus those mutations are applied in the instaql result
   * returned by the server
   */
  _cleanPendingMutations(e) {
    this.pendingMutations.set((n) => {
      const r = new Map(n);
      return [...n.entries()].forEach(([i, s]) => {
        s["tx-id"] <= e && r.delete(i);
      }), r;
    });
  }
  _flushEnqueuedRoomData(e) {
    var n, r;
    const i = (r = (n = this._presence[e]) === null || n === void 0 ? void 0 : n.result) === null || r === void 0 ? void 0 : r.user, s = this._broadcastQueue[e];
    if (this._broadcastQueue[e] = [], i && this._trySetPresence(e, i), s)
      for (const o of s) {
        const { topic: a, roomType: c, data: u } = o;
        this._tryBroadcast(e, c, a, u);
      }
  }
  _handleReceive(e, n) {
    var r, i, s, o;
    const a = !!this.config.schema && ("cardinalityInference" in this.config ? !!this.config.cardinalityInference : !0);
    switch (qn[n.op] || N.info("[receive]", e, n.op, n), n.op) {
      case "init-ok":
        this._setStatus(ye.AUTHENTICATED), this._reconnectTimeoutMs = 0, this._setAttrs(n.attrs), this._flushPendingMessages(), this._sessionId = n["session-id"];
        for (const I of Object.keys(this._rooms))
          this._tryJoinRoom(I);
        break;
      case "add-query-exists":
        this.notifyOneQueryOnce(V(n.q));
        break;
      case "add-query-ok":
        const { q: c, result: u, "processed-tx-id": l } = n;
        this._cleanPendingMutations(l);
        const f = V(c), p = (i = (r = u == null ? void 0 : u[0]) === null || r === void 0 ? void 0 : r.data) === null || i === void 0 ? void 0 : i["page-info"], _ = (o = (s = u == null ? void 0 : u[0]) === null || s === void 0 ? void 0 : s.data) === null || o === void 0 ? void 0 : o.aggregate, g = Dn(u), m = Ut(this.attrs, g, a, this._linkIndex);
        this.querySubs.set((I) => (I[f].result = { store: m, pageInfo: p, aggregate: _ }, I)), this.notifyOne(f), this.notifyOneQueryOnce(f);
        break;
      case "refresh-ok":
        const { computations: O, attrs: S, "processed-tx-id": b } = n;
        this._cleanPendingMutations(b), this._setAttrs(S);
        const y = O.map((I) => {
          var ee, ve, _e, le;
          const Be = I["instaql-query"], se = I["instaql-result"], Qe = V(Be), _t = Dn(se), bt = Ut(this.attrs, _t, a, this._linkIndex), Me = (ve = (ee = se == null ? void 0 : se[0]) === null || ee === void 0 ? void 0 : ee.data) === null || ve === void 0 ? void 0 : ve["page-info"], Ie = (le = (_e = se == null ? void 0 : se[0]) === null || _e === void 0 ? void 0 : _e.data) === null || le === void 0 ? void 0 : le.aggregate;
          return { hash: Qe, store: bt, pageInfo: Me, aggregate: Ie };
        });
        y.forEach(({ hash: I, store: ee, pageInfo: ve, aggregate: _e }) => {
          this.querySubs.set((le) => (le[I].result = { store: ee, pageInfo: ve, aggregate: _e }, le));
        }), y.forEach(({ hash: I }) => {
          this.notifyOne(I);
        });
        break;
      case "transact-ok":
        const { "client-event-id": x, "tx-id": R } = n, $ = this._rewriteMutations(this.attrs, this.pendingMutations.currentValue).get(x);
        if (!$)
          break;
        const ce = Object.assign(Object.assign({}, $), { "tx-id": R });
        this.pendingMutations.set((I) => (I.set(x, ce), I)), this._finishTransaction("synced", x);
        const xe = $["tx-steps"].filter(([I, ...ee]) => I === "add-attr").map(([I, ee]) => ee).concat(Object.values(this.attrs));
        this._setAttrs(xe);
        break;
      case "refresh-presence":
        const Ae = n["room-id"];
        this._setPresencePeers(Ae, n.data), this._notifyPresenceSubs(Ae);
        break;
      case "server-broadcast":
        const pt = n["room-id"], vt = n.topic;
        this._notifyBroadcastSubs(pt, vt, n);
        break;
      case "join-room-ok":
        const ue = n["room-id"], J = this._rooms[ue];
        if (!J) {
          this._roomsPendingLeave[ue] && (this._tryLeaveRoom(ue), delete this._roomsPendingLeave[ue]);
          break;
        }
        J.isConnected = !0, this._notifyPresenceSubs(ue), this._flushEnqueuedRoomData(ue);
        break;
      case "join-room-error":
        const ie = n["room-id"], pe = this._rooms[ie];
        pe && (pe.error = n.error), this._notifyPresenceSubs(ie);
        break;
      case "error":
        this._handleReceiveError(n);
        break;
    }
  }
  /**
   * @param {'timeout' | 'error'} status
   * @param {string} eventId
   * @param {{message?: string, hint?: string, error?: Error}} errDetails
   */
  _handleMutationError(e, n, r) {
    const i = this.pendingMutations.currentValue.get(n);
    i && (e !== "timeout" || !i["tx-id"]) && (this.pendingMutations.set((s) => (s.delete(n), s)), this.notifyAll(), this.notifyAttrsSubs(), this.notifyMutationErrorSubs(r), this._finishTransaction(e, n, r));
  }
  _handleReceiveError(e) {
    var n, r, i;
    const s = e["client-event-id"], o = this.pendingMutations.currentValue.get(s), a = {
      message: e.message || "Uh-oh, something went wrong. Ping Joe & Stopa."
    };
    if (e.hint && (a.hint = e.hint), o) {
      const f = {
        message: e.message,
        hint: e.hint
      };
      this._handleMutationError("error", s, f);
      return;
    }
    const c = e.q || ((n = e["original-event"]) === null || n === void 0 ? void 0 : n.q);
    if (c) {
      const f = V(c);
      this.querySubs.set((p) => (delete p[f], p)), this.notifyQueryError(V(c), a), this.notifyQueryOnceError(f, s, a);
      return;
    }
    if (((r = e["original-event"]) === null || r === void 0 ? void 0 : r.op) === "init") {
      if (e.type === "record-not-found" && ((i = e.hint) === null || i === void 0 ? void 0 : i["record-type"]) === "app-user") {
        this.changeCurrentUser(null);
        return;
      }
      this._setStatus(ye.ERRORED, a), this.notifyAll();
      return;
    }
    const l = Object.assign({}, e);
    delete l.message, delete l.hint, console.error(e.message, l), e.hint && console.error(`This error comes with some debugging information. Here it is: 
`, e.hint);
  }
  notifyQueryOnceError(e, n, r) {
    var i;
    const s = (i = this.queryOnceDfds[e]) === null || i === void 0 ? void 0 : i.find((o) => o.eventId === n);
    s && s.dfd.reject(r);
  }
  _setAttrs(e) {
    this.attrs = e.reduce((n, r) => (n[r.id] = r, n), {}), this.notifyAttrsSubs();
  }
  _startQuerySub(e, n) {
    const r = L();
    return this.querySubs.set((i) => (i[n] = i[n] || { q: e, result: null, eventId: r }, i)), this._trySendAuthed(r, { op: "add-query", q: e }), r;
  }
  /**
   *  When a user subscribes to a query the following side effects occur:
   *
   *  - We update querySubs to include the new query
   *  - We update queryCbs to include the new cb
   *  - If we already have a result for the query we call cb immediately
   *  - We send the server an `add-query` message
   *
   *  Returns an unsubscribe function
   */
  subscribeQuery(e, n) {
    var r;
    const i = V(e), s = this.getPreviousResult(e);
    return s && n(s), this.queryCbs[i] = (r = this.queryCbs[i]) !== null && r !== void 0 ? r : [], this.queryCbs[i].push({ q: e, cb: n }), this._startQuerySub(e, i), () => {
      this._unsubQuery(e, i, n);
    };
  }
  queryOnce(e) {
    var n;
    const r = new Un();
    if (!this._isOnline)
      return r.reject(new Error("We can't run `queryOnce`, because the device is offline.")), r.promise;
    if (!this.querySubs)
      return r.reject(new Error("We can't run `queryOnce` on the backend. Use adminAPI.query instead: https://www.instantdb.com/docs/backend#query")), r.promise;
    const i = V(e), s = this._startQuerySub(e, i);
    return this.queryOnceDfds[i] = (n = this.queryOnceDfds[i]) !== null && n !== void 0 ? n : [], this.queryOnceDfds[i].push({ q: e, dfd: r, eventId: s }), setTimeout(() => r.reject(new Error("Query timed out")), eo), r.promise;
  }
  _completeQueryOnce(e, n, r) {
    this.queryOnceDfds[n] && (this.queryOnceDfds[n] = this.queryOnceDfds[n].filter((i) => i.dfd !== r), this._cleanupQuery(e, n));
  }
  _unsubQuery(e, n, r) {
    this.queryCbs[n] && (this.queryCbs[n] = this.queryCbs[n].filter((i) => i.cb !== r), this._cleanupQuery(e, n));
  }
  _cleanupQuery(e, n) {
    var r, i;
    !((r = this.queryCbs[n]) === null || r === void 0) && r.length || !((i = this.queryOnceDfds[n]) === null || i === void 0) && i.length || (delete this.queryCbs[n], delete this.queryOnceDfds[n], this._trySendAuthed(L(), { op: "remove-query", q: e }));
  }
  // When we `pushTx`, it's possible that we don't yet have `this.attrs`
  // This means that `tx-steps` in `pendingMutations` will include `add-attr`
  // commands for attrs that already exist.
  //
  // This will also affect `add-triple` and `retract-triple` which
  // reference attr-ids that do not match the server.
  //
  // We fix this by rewriting `tx-steps` in each `pendingMutation`.
  // We remove `add-attr` commands for attrs that already exist.
  // We update `add-triple` and `retract-triple` commands to use the
  // server attr-ids.
  _rewriteMutations(e, n) {
    if (!e)
      return n;
    const r = (c) => {
      const [u, l, f] = c["forward-identity"];
      return Q(e, l, f);
    }, i = (c) => {
      const [u, l, f] = c["forward-identity"];
      return we(e, l, f);
    }, s = { attrIdMap: {}, refSwapAttrIds: /* @__PURE__ */ new Set() }, o = (c) => {
      const u = [];
      for (const l of c) {
        const [f] = l;
        if (f === "add-attr") {
          const [_, g] = l, m = r(g);
          if (m) {
            s.attrIdMap[g.id] = m.id;
            continue;
          }
          if (g["value-type"] === "ref") {
            const O = i(g);
            if (O) {
              s.attrIdMap[g.id] = O.id, s.refSwapAttrIds.add(g.id);
              continue;
            }
          }
        }
        const p = bs(s, l);
        u.push(p);
      }
      return u;
    }, a = /* @__PURE__ */ new Map();
    for (const [c, u] of n.entries())
      a.set(c, Object.assign(Object.assign({}, u), { "tx-steps": o(u["tx-steps"]) }));
    return a;
  }
  // ---------------------------
  // Transact
  optimisticAttrs() {
    var e;
    const n = [
      ...this.pendingMutations.currentValue.values()
    ].flatMap((a) => a["tx-steps"]), r = new Set(n.filter(([a, c]) => a === "delete-attr").map(([a, c]) => c)), i = [];
    for (const [a, c] of n)
      if (a === "add-attr")
        i.push(c);
      else if (a === "update-attr" && c.id && (!((e = this.attrs) === null || e === void 0) && e[c.id])) {
        const u = Object.assign(Object.assign({}, this.attrs[c.id]), c);
        i.push(u);
      }
    const s = [
      ...Object.values(this.attrs || {}),
      ...i
    ].filter((a) => !r.has(a.id));
    return Object.fromEntries(s.map((a) => [a.id, a]));
  }
  /** Runs instaql on a query and a store */
  dataForQuery(e) {
    const n = this._errorMessage;
    if (n)
      return { error: n };
    if (!this.querySubs || !this.pendingMutations)
      return;
    const r = this.querySubs.version(), i = this.querySubs.currentValue, s = this.pendingMutations.version(), o = this.pendingMutations.currentValue, { q: a, result: c } = i[e] || {};
    if (!c)
      return;
    const u = this._dataForQueryCache[e];
    if (u && r === u.querySubVersion && s === u.pendingMutationsVersion)
      return u.data;
    const { store: l, pageInfo: f, aggregate: p } = c, g = [...this._rewriteMutations(l.attrs, o).values()].flatMap((S) => S["tx-steps"]), m = xi(l, g), O = ps({ store: m, pageInfo: f, aggregate: p }, a);
    return this._dataForQueryCache[e] = {
      querySubVersion: r,
      pendingMutationsVersion: s,
      data: O
    }, O;
  }
  /** Re-compute all subscriptions */
  notifyAll() {
    Object.keys(this.queryCbs).forEach((e) => {
      this.notifyOne(e);
    });
  }
  loadedNotifyAll() {
    this.pendingMutations.isLoading() || this.querySubs.isLoading() || this.notifyAll();
  }
  shutdown() {
    this._isShutdown = !0, this._ws.close();
  }
  /**
   * Sends mutation to server and schedules a timeout to cancel it if
   * we don't hear back in time.
   * Note: If we're offline we don't schedule a timeout, we'll schedule it
   * later once we're back online and send the mutation again
   *
   */
  _sendMutation(e, n) {
    if (n.error) {
      this._handleMutationError("error", e, {
        error: n.error,
        message: n.error.message
      });
      return;
    }
    if (this.status !== ye.AUTHENTICATED) {
      this._finishTransaction("enqueued", e);
      return;
    }
    const r = Math.max(5e3, this.pendingMutations.currentValue.size * 5e3);
    this._isOnline ? (this._trySend(e, n), setTimeout(() => {
      this._finishTransaction("pending", e);
    }, 3e3), setTimeout(() => {
      this._isOnline && this._handleMutationError("timeout", e, {
        message: "transaction timed out"
      });
    }, r)) : this._finishTransaction("enqueued", e);
  }
  // ---------------------------
  // Websocket
  /** Send messages we accumulated while we were connecting */
  _flushPendingMessages() {
    Object.keys(this.queryCbs).map((i) => this.querySubs.currentValue[i]).filter((i) => i).forEach(({ eventId: i, q: s }) => {
      this._trySendAuthed(i, { op: "add-query", q: s });
    }), Object.values(this.queryOnceDfds).flat().forEach(({ eventId: i, q: s }) => {
      this._trySendAuthed(i, { op: "add-query", q: s });
    }), this._rewriteMutations(this.attrs, this.pendingMutations.currentValue).forEach((i, s) => {
      i["tx-id"] || this._sendMutation(s, i);
    });
  }
  _trySendAuthed(...e) {
    this.status === ye.AUTHENTICATED && this._trySend(...e);
  }
  _trySend(e, n, r) {
    this._ws.readyState === Ln && (qn[n.op] || N.info("[send]", this._ws._id, n.op, n), this._ws.send(JSON.stringify(Object.assign({ "client-event-id": e }, n))));
  }
  _startSocket() {
    if (this._ws && this._ws.readyState == to) {
      N.info("[socket][start]", this._ws._id, "maintained as current ws, we were still in a connecting state");
      return;
    }
    const e = this._ws;
    this._ws = io(`${this.config.websocketURI}?app_id=${this.config.appId}`), this._ws.onopen = this._wsOnOpen, this._ws.onmessage = this._wsOnMessage, this._ws.onclose = this._wsOnClose, this._ws.onerror = this._wsOnError, N.info("[socket][start]", this._ws._id), (e == null ? void 0 : e.readyState) === Ln && (N.info("[socket][start]", this._ws._id, "close previous ws id = ", e._id), e.close());
  }
  /**
   * Given a key, returns a stable local id, unique to this device and app.
   *
   * This can be useful if you want to create guest ids for example.
   *
   * Note: If the user deletes their local storage, this id will change.
   *
   * We use this._localIdPromises to ensure that we only generate a local
   * id once, even if multiple callers call this function concurrently.
   */
  getLocalId(e) {
    return D(this, void 0, void 0, function* () {
      const n = `localToken_${e}`, r = yield this._persister.getItem(n);
      if (r)
        return r;
      if (this._localIdPromises[n])
        return this._localIdPromises[n];
      const i = L();
      return this._localIdPromises[n] = this._persister.setItem(n, i).then(() => i), this._localIdPromises[n];
    });
  }
  // ----
  // Auth
  _replaceUrlAfterOAuth() {
    if (typeof URL > "u")
      return;
    const e = new URL(window.location.href);
    if (e.searchParams.get(Pt)) {
      const n = e.toString();
      e.searchParams.delete(Pt), e.searchParams.delete("code"), e.searchParams.delete("error");
      const r = e.pathname + (e.searchParams.size ? "?" + e.searchParams : "") + e.hash;
      if (history.replaceState(history.state, "", r), // @ts-ignore (waiting for ts support)
      typeof navigation == "object" && // @ts-ignore (waiting for ts support)
      typeof navigation.addEventListener == "function" && // @ts-ignore (waiting for ts support)
      typeof navigation.removeEventListener == "function") {
        let i = !1;
        const s = (o) => {
          var a;
          i || (i = !0, navigation.removeEventListener("navigate", s), !o.userInitiated && o.navigationType === "replace" && ((a = o.destination) === null || a === void 0 ? void 0 : a.url) === n && history.replaceState(history.state, "", r));
        };
        navigation.addEventListener("navigate", s);
      }
    }
  }
  /**
   *
   * @returns Promise<null | {error: {message: string}}>
   */
  _oauthLoginInit() {
    return D(this, void 0, void 0, function* () {
      var e, n, r, i;
      if (typeof window > "u" || typeof window.location > "u" || typeof URLSearchParams > "u")
        return null;
      const s = new URLSearchParams(window.location.search);
      if (!s.get(Pt))
        return null;
      const o = s.get("error");
      if (o)
        return this._replaceUrlAfterOAuth(), { error: { message: o } };
      const a = s.get("code");
      if (!a)
        return null;
      this._replaceUrlAfterOAuth();
      try {
        const { user: c } = yield In({
          apiURI: this.config.apiURI,
          appId: this.config.appId,
          code: a
        });
        return this.setCurrentUser(c), null;
      } catch (c) {
        return ((e = c == null ? void 0 : c.body) === null || e === void 0 ? void 0 : e.type) === "record-not-found" && ((r = (n = c == null ? void 0 : c.body) === null || n === void 0 ? void 0 : n.hint) === null || r === void 0 ? void 0 : r["record-type"]) === "app-oauth-code" && (yield this._hasCurrentUser()) ? null : { error: { message: ((i = c == null ? void 0 : c.body) === null || i === void 0 ? void 0 : i.message) || "Error logging in." } };
      }
    });
  }
  _waitForOAuthCallbackResponse() {
    return D(this, void 0, void 0, function* () {
      return yield this._oauthCallbackResponse;
    });
  }
  __subscribeMutationErrors(e) {
    return this.mutationErrorCbs.push(e), () => {
      this.mutationErrorCbs = this.mutationErrorCbs.filter((n) => n !== e);
    };
  }
  subscribeAuth(e) {
    this.authCbs.push(e);
    const n = this._currentUserCached;
    n.isLoading || e(this._currentUserCached);
    let r = !1;
    return this.getCurrentUser().then((i) => {
      r || st(i, n) || e(i);
    }), () => {
      r = !0, this.authCbs = this.authCbs.filter((i) => i !== e);
    };
  }
  subscribeAttrs(e) {
    return this.attrsCbs.push(e), this.attrs && e(this.attrs), () => {
      this.attrsCbs = this.attrsCbs.filter((n) => n !== e);
    };
  }
  notifyAuthSubs(e) {
    this.authCbs.forEach((n) => n(e));
  }
  notifyMutationErrorSubs(e) {
    this.mutationErrorCbs.forEach((n) => n(e));
  }
  notifyAttrsSubs() {
    if (!this.attrs)
      return;
    const e = this.optimisticAttrs();
    this.attrsCbs.forEach((n) => n(e));
  }
  setCurrentUser(e) {
    return D(this, void 0, void 0, function* () {
      yield this._persister.setItem(Tt, JSON.stringify(e));
    });
  }
  getCurrentUserCached() {
    return this._currentUserCached;
  }
  getCurrentUser() {
    return D(this, void 0, void 0, function* () {
      const e = yield this._waitForOAuthCallbackResponse();
      if (e != null && e.error) {
        const i = { error: e.error, user: void 0 };
        return this._currentUserCached = Object.assign({ isLoading: !1 }, i), i;
      }
      const n = yield this._persister.getItem(Tt), r = { user: JSON.parse(n), error: void 0 };
      return this._currentUserCached = Object.assign({ isLoading: !1 }, r), r;
    });
  }
  _hasCurrentUser() {
    return D(this, void 0, void 0, function* () {
      const e = yield this._persister.getItem(Tt);
      return JSON.parse(e) != null;
    });
  }
  changeCurrentUser(e) {
    return D(this, void 0, void 0, function* () {
      var n;
      const { user: r } = yield this.getCurrentUser();
      if (!st(r, e)) {
        yield this.setCurrentUser(e), this.updateUser(e);
        try {
          (n = this._broadcastChannel) === null || n === void 0 || n.postMessage({ type: "auth" });
        } catch (i) {
          console.error("Error posting message to broadcast channel", i);
        }
      }
    });
  }
  updateUser(e) {
    const n = { error: void 0, user: e };
    this._currentUserCached = Object.assign({ isLoading: !1 }, n), this._dataForQueryCache = {}, this.querySubs.set((r) => (Object.keys(r).forEach((i) => {
      delete r[i].result;
    }), r)), this._reconnectTimeoutMs = 0, this._ws.close(), this._oauthCallbackResponse = null, this.notifyAuthSubs(n);
  }
  sendMagicCode({ email: e }) {
    return Fs({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      email: e
    });
  }
  signInWithMagicCode(e) {
    return D(this, arguments, void 0, function* ({ email: n, code: r }) {
      const i = yield zs({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        email: n,
        code: r
      });
      return yield this.changeCurrentUser(i.user), i;
    });
  }
  signInWithCustomToken(e) {
    return D(this, void 0, void 0, function* () {
      const n = yield Ws({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        refreshToken: e
      });
      return yield this.changeCurrentUser(n.user), n;
    });
  }
  signOut() {
    return D(this, void 0, void 0, function* () {
      var e;
      const n = yield this.getCurrentUser(), r = (e = n == null ? void 0 : n.user) === null || e === void 0 ? void 0 : e.refresh_token;
      if (r)
        try {
          yield Bs({
            apiURI: this.config.apiURI,
            appId: this.config.appId,
            refreshToken: r
          });
        } catch {
        }
      yield this.changeCurrentUser(null);
    });
  }
  /**
   * Creates an OAuth authorization URL.
   * @param {Object} params - The parameters to create the authorization URL.
   * @param {string} params.clientName - The name of the client requesting authorization.
   * @param {string} params.redirectURL - The URL to redirect users to after authorization.
   * @returns {string} The created authorization URL.
   */
  createAuthorizationURL({ clientName: e, redirectURL: n }) {
    const { apiURI: r, appId: i } = this.config;
    return `${r}/runtime/oauth/start?app_id=${i}&client_name=${e}&redirect_uri=${n}`;
  }
  exchangeCodeForToken(e) {
    return D(this, arguments, void 0, function* ({ code: n, codeVerifier: r }) {
      const i = yield In({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        code: n,
        codeVerifier: r
      });
      return yield this.changeCurrentUser(i.user), i;
    });
  }
  issuerURI() {
    const { apiURI: e, appId: n } = this.config;
    return `${e}/runtime/${n}`;
  }
  /**
   * @param {Object} params
   * @param {string} params.clientName - The name of the client requesting authorization.
   * @param {string} params.idToken - The id_token from the external service
   * @param {string | null | undefined} [params.nonce] - The nonce used when requesting the id_token from the external service
   */
  signInWithIdToken(e) {
    return D(this, arguments, void 0, function* ({ idToken: n, clientName: r, nonce: i }) {
      var s;
      const o = yield this.getCurrentUser(), a = (s = o == null ? void 0 : o.user) === null || s === void 0 ? void 0 : s.refresh_token, c = yield Vs({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        idToken: n,
        clientName: r,
        nonce: i,
        refreshToken: a
      });
      return yield this.changeCurrentUser(c.user), c;
    });
  }
  // --------
  // Rooms
  joinRoom(e) {
    return this._rooms[e] || (this._rooms[e] = {
      isConnected: !1,
      error: void 0
    }), this._presence[e] = this._presence[e] || {}, this._tryJoinRoom(e), () => {
      this._cleanupRoom(e);
    };
  }
  _cleanupRoom(e) {
    var n, r, i, s;
    if (!(!((r = (n = this._presence[e]) === null || n === void 0 ? void 0 : n.handlers) === null || r === void 0) && r.length) && !Object.keys((i = this._broadcastSubs[e]) !== null && i !== void 0 ? i : {}).length) {
      const o = (s = this._rooms[e]) === null || s === void 0 ? void 0 : s.isConnected;
      delete this._rooms[e], delete this._presence[e], delete this._broadcastSubs[e], o ? this._tryLeaveRoom(e) : this._roomsPendingLeave[e] = !0;
    }
  }
  // --------
  // Presence
  // TODO: look into typing again
  getPresence(e, n, r = {}) {
    const i = this._rooms[n], s = this._presence[n];
    return !i || !s || !s.result ? null : Object.assign(Object.assign({}, Hs(s.result, r, this._sessionId)), { isLoading: !i.isConnected, error: i.error });
  }
  // TODO: look into typing again
  publishPresence(e, n, r) {
    const i = this._rooms[n], s = this._presence[n];
    if (!i || !s)
      return;
    s.result = s.result || {};
    const o = Object.assign(Object.assign({}, s.result.user), r);
    s.result.user = o, i.isConnected && (this._trySetPresence(n, o), this._notifyPresenceSubs(n));
  }
  _trySetPresence(e, n) {
    this._trySendAuthed(L(), {
      op: "set-presence",
      "room-id": e,
      data: n
    });
  }
  _tryJoinRoom(e) {
    this._trySendAuthed(L(), { op: "join-room", "room-id": e }), delete this._roomsPendingLeave[e];
  }
  _tryLeaveRoom(e) {
    this._trySendAuthed(L(), { op: "leave-room", "room-id": e });
  }
  // TODO: look into typing again
  subscribePresence(e, n, r, i) {
    const s = this.joinRoom(n), o = Object.assign(Object.assign({}, r), { roomId: n, cb: i, prev: null });
    return this._presence[n] = this._presence[n] || {}, this._presence[n].handlers = this._presence[n].handlers || [], this._presence[n].handlers.push(o), this._notifyPresenceSub(n, o), () => {
      var a, c, u;
      this._presence[n].handlers = (u = (c = (a = this._presence[n]) === null || a === void 0 ? void 0 : a.handlers) === null || c === void 0 ? void 0 : c.filter((l) => l !== o)) !== null && u !== void 0 ? u : [], s();
    };
  }
  _notifyPresenceSubs(e) {
    var n, r;
    (r = (n = this._presence[e]) === null || n === void 0 ? void 0 : n.handlers) === null || r === void 0 || r.forEach((i) => {
      this._notifyPresenceSub(e, i);
    });
  }
  _notifyPresenceSub(e, n) {
    const r = this.getPresence("", e, n);
    r && (n.prev && !Gs(r, n.prev) || (n.prev = r, n.cb(r)));
  }
  _setPresencePeers(e, n) {
    const r = Object.assign({}, n);
    delete r[this._sessionId];
    const i = Object.fromEntries(Object.entries(r).map(([s, o]) => [s, o.data]));
    this._presence[e] = this._presence[e] || {}, this._presence[e].result = this._presence[e].result || {}, this._presence[e].result.peers = i;
  }
  // --------
  // Broadcast
  publishTopic({ roomType: e, roomId: n, topic: r, data: i }) {
    var s;
    const o = this._rooms[n];
    if (o) {
      if (!o.isConnected) {
        this._broadcastQueue[n] = (s = this._broadcastQueue[n]) !== null && s !== void 0 ? s : [], this._broadcastQueue[n].push({ topic: r, roomType: e, data: i });
        return;
      }
      this._tryBroadcast(n, e, r, i);
    }
  }
  _tryBroadcast(e, n, r, i) {
    this._trySendAuthed(L(), {
      op: "client-broadcast",
      "room-id": e,
      roomType: n,
      topic: r,
      data: i
    });
  }
  subscribeTopic(e, n, r) {
    const i = this.joinRoom(e);
    return this._broadcastSubs[e] = this._broadcastSubs[e] || {}, this._broadcastSubs[e][n] = this._broadcastSubs[e][n] || [], this._broadcastSubs[e][n].push(r), this._presence[e] = this._presence[e] || {}, () => {
      this._broadcastSubs[e][n] = this._broadcastSubs[e][n].filter((s) => s !== r), this._broadcastSubs[e][n].length || delete this._broadcastSubs[e][n], i();
    };
  }
  _notifyBroadcastSubs(e, n, r) {
    var i, s, o;
    (o = (s = (i = this._broadcastSubs) === null || i === void 0 ? void 0 : i[e]) === null || s === void 0 ? void 0 : s[n]) === null || o === void 0 || o.forEach((a) => {
      var c, u, l, f, p, _;
      const g = (c = r.data) === null || c === void 0 ? void 0 : c.data, m = r.data["peer-id"] === this._sessionId ? (l = (u = this._presence[e]) === null || u === void 0 ? void 0 : u.result) === null || l === void 0 ? void 0 : l.user : (_ = (p = (f = this._presence[e]) === null || f === void 0 ? void 0 : f.result) === null || p === void 0 ? void 0 : p.peers) === null || _ === void 0 ? void 0 : _[r.data["peer-id"]];
      return a(g, m);
    });
  }
  // --------
  // Storage
  upload(e, n) {
    return D(this, void 0, void 0, function* () {
      var r;
      const i = yield this.getCurrentUser(), s = (r = i == null ? void 0 : i.user) === null || r === void 0 ? void 0 : r.refresh_token, o = e || n.name, a = yield Qs({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        fileName: o,
        refreshToken: s
      });
      return yield Js(a, n);
    });
  }
  getDownloadUrl(e) {
    return D(this, void 0, void 0, function* () {
      var n;
      const r = yield this.getCurrentUser(), i = (n = r == null ? void 0 : r.user) === null || n === void 0 ? void 0 : n.refresh_token;
      return yield Ks({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        path: e,
        refreshToken: i
      });
    });
  }
  deleteFile(e) {
    return D(this, void 0, void 0, function* () {
      var n;
      const r = yield this.getCurrentUser(), i = (n = r == null ? void 0 : r.user) === null || n === void 0 ? void 0 : n.refresh_token;
      return yield Ys({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        path: e,
        refreshToken: i
      });
    });
  }
}
class ae {
  constructor(e, n, r = { indexed: !1, unique: !1 }) {
    this.valueType = e, this.required = n, this.config = r;
  }
  optional() {
    return new ae(this.valueType, !1);
  }
  unique() {
    return new ae(this.valueType, this.required, Object.assign(Object.assign({}, this.config), { unique: !0 }));
  }
  indexed() {
    return new ae(this.valueType, this.required, Object.assign(Object.assign({}, this.config), { indexed: !0 }));
  }
}
class ft {
  constructor(e, n) {
    this.attrs = e, this.links = n;
  }
  asType() {
    return new ft(this.attrs, this.links);
  }
}
class ht {
  constructor(e, n, r) {
    this.entities = e, this.links = n, this.rooms = r;
  }
  withRoomSchema() {
    return new ht(this.entities, this.links, {});
  }
}
function co(t, e) {
  return new ht(
    br(t, e),
    // (XXX): LinksDef<any> stems from TypeScript’s inability to reconcile the
    // type EntitiesWithLinks<EntitiesWithoutLinks, Links> with
    // EntitiesWithoutLinks. TypeScript is strict about ensuring that types are
    // correctly aligned and does not allow for substituting a type that might
    // be broader or have additional properties.
    e,
    void 0
  );
}
function uo(t) {
  return new ft(t, {});
}
function lo() {
  return new ae("string", !0);
}
function fo() {
  return new ae("number", !0);
}
function ho() {
  return new ae("boolean", !0);
}
function po() {
  return new ae("date", !0);
}
function vo() {
  return new ae("json", !0);
}
function _o() {
  return new ae("json", !0);
}
function br(t, e) {
  var n, r, i, s;
  const o = { fwd: {}, rev: {} };
  for (const c of Object.values(e))
    (n = o.fwd)[r = c.forward.on] || (n[r] = {}), (i = o.rev)[s = c.reverse.on] || (i[s] = {}), o.fwd[c.forward.on][c.forward.label] = {
      entityName: c.reverse.on,
      cardinality: c.forward.has
    }, o.rev[c.reverse.on][c.reverse.label] = {
      entityName: c.forward.on,
      cardinality: c.reverse.has
    };
  return Object.fromEntries(Object.entries(t).map(([c, u]) => [
    c,
    new ft(u.attrs, Object.assign(Object.assign({}, o.fwd[c]), o.rev[c]))
  ]));
}
function bo({ entities: t, links: e, rooms: n }) {
  return new ht(
    br(t, e),
    // (XXX): LinksDef<any> stems from TypeScript’s inability to reconcile the
    // type EntitiesWithLinks<EntitiesWithoutLinks, Links> with
    // EntitiesWithoutLinks. TypeScript is strict about ensuring that types are
    // correctly aligned and does not allow for substituting a type that might
    // be broader or have additional properties.
    e,
    n
  );
}
const Yo = {
  // constructs
  graph: co,
  schema: bo,
  entity: uo,
  // value types
  string: lo,
  number: fo,
  boolean: ho,
  date: po,
  json: vo,
  any: _o
};
let et;
function yr(t) {
  et == null || et.dispose();
  const e = mo(), n = wo(o), r = go(yo(t));
  function i(u) {
    var l;
    u.source === r.element.contentWindow && ((l = u.data) === null || l === void 0 ? void 0 : l.type) === "close" && e.isVisible() && o();
  }
  function s(u) {
    const l = u.shiftKey && u.ctrlKey && u.key === "0", f = u.key === "Escape" || u.key === "Esc";
    (l || f && e.isVisible()) && o();
  }
  function o() {
    e.isVisible() ? e.element.style.display = "none" : (e.element.style.display = "block", e.element.contains(r.element) || e.element.appendChild(r.element));
  }
  function a() {
    e.element.remove(), n.element.remove(), removeEventListener("keydown", s), removeEventListener("message", i);
  }
  function c() {
    document.body.appendChild(e.element), document.body.appendChild(n.element), addEventListener("keydown", s), addEventListener("message", i), et = {
      dispose: a
    };
  }
  return c();
}
function yo(t) {
  return `${window.DEV_DEVTOOL ? "http://localhost:3000" : "https://instantdb.com"}/_devtool?appId=${t}`;
}
function go(t) {
  const e = document.createElement("iframe");
  return e.src = t, e.className = "instant-devtool-iframe", Object.assign(e.style, {
    width: "100%",
    height: "100%",
    borderRadius: "4px",
    backgroundColor: "white",
    border: "none"
  }), { element: e };
}
function wo(t) {
  const e = `
    <svg width="32" height="32" viewBox="0 0 512 512" fill="none" xmlns="http://www.w3.org/2000/svg">
      <rect width="512" height="512" fill="black"/>
      <rect x="97.0973" y="91.3297" width="140" height="330" fill="white"/>
    </svg>
  `, n = document.createElement("button");
  return n.innerHTML = e, n.className = "instant-devtool-toggler", Object.assign(n.style, {
    // pos
    position: "fixed",
    bottom: "24px",
    left: "24px",
    height: "32px",
    width: "32px",
    // layout
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    // look
    padding: "0",
    margin: "0",
    border: "none",
    cursor: "pointer"
  }), n.addEventListener("click", t), { element: n };
}
function mo() {
  const t = document.createElement("div");
  Object.assign(t.style, {
    position: "fixed",
    bottom: "24px",
    right: "24px",
    left: "60px",
    top: "72px",
    display: "block",
    borderRadius: "4px",
    border: "1px #ccc solid",
    boxShadow: "0px 0px 8px #00000044",
    backgroundColor: "#eee",
    zIndex: "999990"
  }), t.style.display = "none", t.className = "instant-devtool-container";
  function e() {
    return t.style.display !== "none";
  }
  return { element: t, isVisible: e };
}
const gr = !0, wr = {
  apiURI: "https://api.instantdb.com",
  websocketURI: "wss://api.instantdb.com/runtime/session"
};
function Oo() {
  var t;
  return globalThis.__instantDbStore = (t = globalThis.__instantDbStore) !== null && t !== void 0 ? t : {}, globalThis.__instantDbStore;
}
const Te = Oo();
function So(t, e, n, r) {
  const i = Te[t.appId];
  if (i)
    return i;
  const s = new _r(Object.assign(Object.assign({}, wr), t), e || tn, n || nn, r), o = new Ro(s);
  return Te[t.appId] = o, typeof window < "u" && typeof window.location < "u" && // show widget by default?
  ("devtool" in t ? t.devtool : gr) && // only run on localhost (dev env)
  window.location.hostname === "localhost" && // used by dash and other internal consumers
  !globalThis._nodevtool && yr(t.appId), o;
}
class Ro {
  constructor(e) {
    this.tx = We(), this._reactor = e, this.auth = new mr(this._reactor), this.storage = new Or(this._reactor);
  }
  /**
   * Use this to write data! You can create, update, delete, and link objects
   *
   * @see https://instantdb.com/docs/instaml
   *
   * @example
   *   // Create a new object in the `goals` namespace
   *   const goalId = id();
   *   db.transact(tx.goals[goalId].update({title: "Get fit"}))
   *
   *   // Update the title
   *   db.transact(tx.goals[goalId].update({title: "Get super fit"}))
   *
   *   // Delete it
   *   db.transact(tx.goals[goalId].delete())
   *
   *   // Or create an association:
   *   todoId = id();
   *   db.transact([
   *    tx.todos[todoId].update({ title: 'Go on a run' }),
   *    tx.goals[goalId].link({todos: todoId}),
   *  ])
   */
  transact(e) {
    return this._reactor.pushTx(e);
  }
  getLocalId(e) {
    return this._reactor.getLocalId(e);
  }
  /**
   * Use this to query your data!
   *
   * @see https://instantdb.com/docs/instaql
   *
   * @example
   *  // listen to all goals
   *  db.subscribeQuery({ goals: {} }, (resp) => {
   *    console.log(resp.data.goals)
   *  })
   *
   *  // goals where the title is "Get Fit"
   *  db.subscribeQuery(
   *    { goals: { $: { where: { title: "Get Fit" } } } },
   *    (resp) => {
   *      console.log(resp.data.goals)
   *    }
   *  )
   *
   *  // all goals, _alongside_ their todos
   *  db.subscribeQuery({ goals: { todos: {} } }, (resp) => {
   *    console.log(resp.data.goals)
   *  });
   */
  subscribeQuery(e, n) {
    return this._reactor.subscribeQuery(e, n);
  }
  /**
   * Listen for the logged in state. This is useful
   * for deciding when to show a login screen.
   *
   * @see https://instantdb.com/docs/auth
   * @example
   *   const unsub = db.subscribeAuth((auth) => {
   *     if (auth.user) {
   *     console.log('logged in as', auth.user.email)
   *    } else {
   *      console.log('logged out')
   *    }
   *  })
   */
  subscribeAuth(e) {
    return this._reactor.subscribeAuth(e);
  }
  /**
   * Join a room to publish and subscribe to topics and presence.
   *
   * @see https://instantdb.com/docs/presence-and-topics
   * @example
   * // init
   * const db = init();
   * const room = db.joinRoom(roomType, roomId);
   * // usage
   * const unsubscribeTopic = room.subscribeTopic("foo", console.log);
   * const unsubscribePresence = room.subscribePresence({}, console.log);
   * room.publishTopic("hello", { message: "hello world!" });
   * room.publishPresence({ name: "joe" });
   * // later
   * unsubscribePresence();
   * unsubscribeTopic();
   * room.leaveRoom();
   */
  joinRoom(e = "_defaultRoomType", n = "_defaultRoomId") {
    return {
      leaveRoom: this._reactor.joinRoom(n),
      subscribeTopic: (i, s) => this._reactor.subscribeTopic(n, i, s),
      subscribePresence: (i, s) => this._reactor.subscribePresence(e, n, i, s),
      publishTopic: (i, s) => this._reactor.publishTopic({ roomType: e, roomId: n, topic: i, data: s }),
      publishPresence: (i) => this._reactor.publishPresence(e, n, i),
      getPresence: (i) => this._reactor.getPresence(e, n, i)
    };
  }
  shutdown() {
    delete Te[this._reactor.config.appId], this._reactor.shutdown();
  }
  /**
   * Use this for one-off queries.
   * Returns local data if available, otherwise fetches from the server.
   * Because we want to avoid stale data, this method will throw an error
   * if the user is offline or there is no active connection to the server.
   *
   * @see https://instantdb.com/docs/instaql
   *
   * @example
   *
   *  const resp = await db.queryOnce({ goals: {} });
   *  console.log(resp.data.goals)
   */
  queryOnce(e) {
    return this._reactor.queryOnce(e);
  }
}
class mr {
  constructor(e) {
    this.db = e, this.sendMagicCode = (n) => this.db.sendMagicCode(n), this.signInWithMagicCode = (n) => this.db.signInWithMagicCode(n), this.signInWithToken = (n) => this.db.signInWithCustomToken(n), this.createAuthorizationURL = (n) => this.db.createAuthorizationURL(n), this.signInWithIdToken = (n) => this.db.signInWithIdToken(n), this.exchangeOAuthCode = (n) => this.db.exchangeCodeForToken(n), this.issuerURI = () => this.db.issuerURI(), this.signOut = () => this.db.signOut();
  }
}
class Or {
  constructor(e) {
    this.db = e, this.upload = (n, r) => this.db.upload(n, r), this.put = this.upload, this.getDownloadUrl = (n) => this.db.getDownloadUrl(n), this.delete = (n) => this.db.deleteFile(n);
  }
}
function Sr(t) {
  return JSON.parse(JSON.stringify(t));
}
class Eo {
  constructor(e) {
    this.tx = We(), this._reactor = e, this.auth = new mr(this._reactor), this.storage = new Or(this._reactor);
  }
  /**
   * Use this to write data! You can create, update, delete, and link objects
   *
   * @see https://instantdb.com/docs/instaml
   *
   * @example
   *   // Create a new object in the `goals` namespace
   *   const goalId = id();
   *   db.transact(tx.goals[goalId].update({title: "Get fit"}))
   *
   *   // Update the title
   *   db.transact(tx.goals[goalId].update({title: "Get super fit"}))
   *
   *   // Delete it
   *   db.transact(tx.goals[goalId].delete())
   *
   *   // Or create an association:
   *   todoId = id();
   *   db.transact([
   *    tx.todos[todoId].update({ title: 'Go on a run' }),
   *    tx.goals[goalId].link({todos: todoId}),
   *  ])
   */
  transact(e) {
    return this._reactor.pushTx(e);
  }
  getLocalId(e) {
    return this._reactor.getLocalId(e);
  }
  /**
   * Use this to query your data!
   *
   * @see https://instantdb.com/docs/instaql
   *
   * @example
   *  // listen to all goals
   *  db.subscribeQuery({ goals: {} }, (resp) => {
   *    console.log(resp.data.goals)
   *  })
   *
   *  // goals where the title is "Get Fit"
   *  db.subscribeQuery(
   *    { goals: { $: { where: { title: "Get Fit" } } } },
   *    (resp) => {
   *      console.log(resp.data.goals)
   *    }
   *  )
   *
   *  // all goals, _alongside_ their todos
   *  db.subscribeQuery({ goals: { todos: {} } }, (resp) => {
   *    console.log(resp.data.goals)
   *  });
   */
  subscribeQuery(e, n) {
    return this._reactor.subscribeQuery(e, n);
  }
  /**
   * Listen for the logged in state. This is useful
   * for deciding when to show a login screen.
   *
   * @see https://instantdb.com/docs/auth
   * @example
   *   const unsub = db.subscribeAuth((auth) => {
   *     if (auth.user) {
   *     console.log('logged in as', auth.user.email)
   *    } else {
   *      console.log('logged out')
   *    }
   *  })
   */
  subscribeAuth(e) {
    return this._reactor.subscribeAuth(e);
  }
  /**
   * Join a room to publish and subscribe to topics and presence.
   *
   * @see https://instantdb.com/docs/presence-and-topics
   * @example
   * // init
   * const db = init();
   * const room = db.joinRoom(roomType, roomId);
   * // usage
   * const unsubscribeTopic = room.subscribeTopic("foo", console.log);
   * const unsubscribePresence = room.subscribePresence({}, console.log);
   * room.publishTopic("hello", { message: "hello world!" });
   * room.publishPresence({ name: "joe" });
   * // later
   * unsubscribePresence();
   * unsubscribeTopic();
   * room.leaveRoom();
   */
  joinRoom(e = "_defaultRoomType", n = "_defaultRoomId") {
    return {
      leaveRoom: this._reactor.joinRoom(n),
      subscribeTopic: (i, s) => this._reactor.subscribeTopic(n, i, s),
      subscribePresence: (i, s) => this._reactor.subscribePresence(e, n, i, s),
      publishTopic: (i, s) => this._reactor.publishTopic({ roomType: e, roomId: n, topic: i, data: s }),
      publishPresence: (i) => this._reactor.publishPresence(e, n, i),
      getPresence: (i) => this._reactor.getPresence(e, n, i)
    };
  }
  shutdown() {
    delete Te[this._reactor.config.appId], this._reactor.shutdown();
  }
  /**
   * Use this for one-off queries.
   * Returns local data if available, otherwise fetches from the server.
   * Because we want to avoid stale data, this method will throw an error
   * if the user is offline or there is no active connection to the server.
   *
   * @see https://instantdb.com/docs/instaql
   *
   * @example
   *
   *  const resp = await db.queryOnce({ goals: {} });
   *  console.log(resp.data.goals)
   */
  queryOnce(e) {
    return this._reactor.queryOnce(e);
  }
}
function jo(t, e, n) {
  const r = Te[t.appId];
  if (r)
    return r;
  const i = new _r(Object.assign(Object.assign(Object.assign({}, wr), t), { cardinalityInference: !!t.schema }), e || tn, n || nn), s = new Eo(i);
  return Te[t.appId] = s, typeof window < "u" && typeof window.location < "u" && // show widget by default?
  ("devtool" in t ? t.devtool : gr) && // only run on localhost (dev env)
  window.location.hostname === "localhost" && // used by dash and other internal consumers
  !globalThis._nodevtool && yr(t.appId), s;
}
const Rr = {
  isLoading: !0,
  data: void 0,
  pageInfo: void 0,
  error: void 0
};
function Er(t) {
  return {
    isLoading: !t,
    data: void 0,
    pageInfo: void 0,
    error: void 0,
    ...t || {}
  };
}
function Po(t, e) {
  const n = e ? Sr(e) : null, r = V(n), i = Fe(Er(t._reactor.getPreviousResult(n))), s = Ce(
    (a) => n ? t.subscribeQuery(n, (u) => {
      i.current = {
        isLoading: !u,
        data: void 0,
        pageInfo: void 0,
        error: void 0,
        ...u
      }, a();
    }) : () => {
    },
    // Build a new subscribe function if the query changes
    [r, t]
  );
  return { state: at(
    s,
    () => i.current,
    () => Rr
  ), query: n };
}
function To(t, e) {
  const n = e ? Sr(e) : null, r = V(n), i = Fe(Er(t._reactor.getPreviousResult(n))), s = Ce(
    (a) => n ? t.subscribeQuery(n, (u) => {
      i.current = {
        isLoading: !u,
        data: void 0,
        pageInfo: void 0,
        error: void 0,
        ...u
      }, a();
    }) : () => {
    },
    // Build a new subscribe function if the query changes
    [r]
  );
  return { state: at(
    s,
    () => i.current,
    () => Rr
  ), query: n };
}
function jr() {
  const t = Fe(null);
  ne(() => {
    n();
  }, []);
  function e(r, i) {
    clearTimeout(t.current), t.current = setTimeout(i, r);
  }
  function n() {
    clearTimeout(t.current);
  }
  return { set: e, clear: n };
}
const Co = "v0.15.9", xo = 1e3;
let Ao = class {
  constructor(e, n, r) {
    this.useTopicEffect = (i, s) => {
      ne(() => this._core._reactor.subscribeTopic(
        this.id,
        i,
        (a, c) => {
          s(a, c);
        }
      ), [this.id, i]);
    }, this.usePublishTopic = (i) => (ne(() => this._core._reactor.joinRoom(this.id), [this.id]), Ce(
      (o) => {
        this._core._reactor.publishTopic({
          roomType: this.type,
          roomId: this.id,
          topic: i,
          data: o
        });
      },
      [this.id, i]
    )), this.usePresence = (i = {}) => {
      var a, c;
      const [s, o] = Vn(() => this._core._reactor.getPresence(this.type, this.id, i) ?? {
        peers: {},
        isLoading: !0
      });
      return ne(() => this._core._reactor.subscribePresence(
        this.type,
        this.id,
        i,
        (l) => {
          o(l);
        }
      ), [this.id, i.user, (a = i.peers) == null ? void 0 : a.join(), (c = i.keys) == null ? void 0 : c.join()]), {
        ...s,
        publishPresence: (u) => {
          this._core._reactor.publishPresence(this.type, this.id, u);
        }
      };
    }, this.useSyncPresence = (i, s) => {
      ne(() => this._core._reactor.joinRoom(this.id), [this.id]), ne(() => this._core._reactor.publishPresence(this.type, this.id, i), [this.type, this.id, s ?? JSON.stringify(i)]);
    }, this.useTypingIndicator = (i, s = {}) => {
      const o = jr(), a = this.usePresence({
        keys: [i]
      }), c = Bn(() => {
        const l = this._core._reactor.getPresence(
          this.type,
          this.id
        );
        return s != null && s.writeOnly ? [] : Object.values((l == null ? void 0 : l.peers) ?? {}).filter(
          (f) => f[i] === !0
        );
      }, [s == null ? void 0 : s.writeOnly, a]), u = (l) => {
        this._core._reactor.publishPresence(this.type, this.id, {
          [i]: l
        }), l && ((s == null ? void 0 : s.timeout) === null || (s == null ? void 0 : s.timeout) === 0 || o.set((s == null ? void 0 : s.timeout) ?? xo, () => {
          this._core._reactor.publishPresence(this.type, this.id, {
            [i]: null
          });
        }));
      };
      return {
        active: c,
        setActive: (l) => {
          u(l);
        },
        inputProps: {
          onKeyDown: (l) => {
            const p = !((s == null ? void 0 : s.stopOnEnter) && l.key === "Enter");
            u(p);
          },
          onBlur: () => {
            u(!1);
          }
        }
      };
    }, this._core = e, this.type = n, this.id = r;
  }
};
const Mo = {
  isLoading: !0,
  user: void 0,
  error: void 0
};
class Io {
  constructor(e, n) {
    this.tx = We(), this.getLocalId = (r) => this._core.getLocalId(r), this.transact = (r) => this._core.transact(r), this.useQuery = (r) => Po(this._core, r).state, this.useAuth = () => {
      const r = Fe(
        this._core._reactor._currentUserCached
      ), i = Ce((o) => this._core.subscribeAuth((c) => {
        r.current = { isLoading: !1, ...c }, o();
      }), []);
      return at(
        i,
        () => r.current,
        () => Mo
      );
    }, this.queryOnce = (r) => this._core.queryOnce(r), this._core = So(
      e,
      // @ts-expect-error because TS can't resolve subclass statics
      this.constructor.Storage,
      // @ts-expect-error because TS can't resolve subclass statics
      this.constructor.NetworkListener,
      { ...n || {}, "@instantdb/react": Co }
    ), this.auth = this._core.auth, this.storage = this._core.storage;
  }
  /**
   * Obtain a handle to a room, which allows you to listen to topics and presence data
   *
   * If you don't provide a `type` or `id`, Instant will default to `_defaultRoomType` and `_defaultRoomId`
   * as the room type and id, respectively.
   *
   * @see https://instantdb.com/docs/presence-and-topics
   *
   * @example
   *  const {
   *   useTopicEffect,
   *   usePublishTopic,
   *   useSyncPresence,
   *   useTypingIndicator,
   * } = db.room(roomType, roomId);
   */
  room(e = "_defaultRoomType", n = "_defaultRoomId") {
    return new Ao(
      this._core,
      e,
      n
    );
  }
}
const ko = 1e3;
class Uo {
  constructor(e, n, r) {
    this.useTopicEffect = (i, s) => {
      ne(() => this._core._reactor.subscribeTopic(
        this.id,
        i,
        (a, c) => {
          s(a, c);
        }
      ), [this.id, i]);
    }, this.usePublishTopic = (i) => (ne(() => this._core._reactor.joinRoom(this.id), [this.id]), Ce(
      (o) => {
        this._core._reactor.publishTopic({
          roomType: this.type,
          roomId: this.id,
          topic: i,
          data: o
        });
      },
      [this.id, i]
    )), this.usePresence = (i = {}) => {
      var a, c;
      const [s, o] = Vn(() => this._core._reactor.getPresence(this.type, this.id, i) ?? {
        peers: {},
        isLoading: !0
      });
      return ne(() => this._core._reactor.subscribePresence(
        this.type,
        this.id,
        i,
        (l) => {
          o(l);
        }
      ), [this.id, i.user, (a = i.peers) == null ? void 0 : a.join(), (c = i.keys) == null ? void 0 : c.join()]), {
        ...s,
        publishPresence: (u) => {
          this._core._reactor.publishPresence(this.type, this.id, u);
        }
      };
    }, this.useSyncPresence = (i, s) => {
      ne(() => this._core._reactor.joinRoom(this.id), [this.id]), ne(() => this._core._reactor.publishPresence(this.type, this.id, i), [this.type, this.id, s ?? JSON.stringify(i)]);
    }, this.useTypingIndicator = (i, s = {}) => {
      const o = jr(), a = this.usePresence({
        keys: [i]
      }), c = Bn(() => {
        const l = this._core._reactor.getPresence(
          this.type,
          this.id
        );
        return s != null && s.writeOnly ? [] : Object.values((l == null ? void 0 : l.peers) ?? {}).filter(
          (f) => f[i] === !0
        );
      }, [s == null ? void 0 : s.writeOnly, a]), u = (l) => {
        this._core._reactor.publishPresence(this.type, this.id, {
          [i]: l
        }), l && ((s == null ? void 0 : s.timeout) === null || (s == null ? void 0 : s.timeout) === 0 || o.set((s == null ? void 0 : s.timeout) ?? ko, () => {
          this._core._reactor.publishPresence(this.type, this.id, {
            [i]: null
          });
        }));
      };
      return {
        active: c,
        setActive: (l) => {
          u(l);
        },
        inputProps: {
          onKeyDown: (l) => {
            const p = !((s == null ? void 0 : s.stopOnEnter) && l.key === "Enter");
            u(p);
          },
          onBlur: () => {
            u(!1);
          }
        }
      };
    }, this._core = e, this.type = n, this.id = r;
  }
}
const $o = {
  isLoading: !0,
  user: void 0,
  error: void 0
};
class Do {
  constructor(e) {
    this.tx = We(), this.getLocalId = (n) => this._core.getLocalId(n), this.transact = (n) => this._core.transact(n), this.useQuery = (n) => To(this._core, n).state, this.useAuth = () => {
      const n = Fe(
        this._core._reactor._currentUserCached
      ), r = Ce((s) => this._core.subscribeAuth((a) => {
        n.current = { isLoading: !1, ...a }, s();
      }), []);
      return at(
        r,
        () => n.current,
        () => $o
      );
    }, this.queryOnce = (n) => this._core.queryOnce(n), this._core = jo(
      e,
      // @ts-expect-error because TS can't resolve subclass statics
      this.constructor.Storage,
      // @ts-expect-error because TS can't resolve subclass statics
      this.constructor.NetworkListener
    ), this.auth = this._core.auth, this.storage = this._core.storage;
  }
  /**
   * Obtain a handle to a room, which allows you to listen to topics and presence data
   *
   * If you don't provide a `type` or `id`, Instant will default to `_defaultRoomType` and `_defaultRoomId`
   * as the room type and id, respectively.
   *
   * @see https://instantdb.com/docs/presence-and-topics
   *
   * @example
   *  const {
   *   useTopicEffect,
   *   usePublishTopic,
   *   useSyncPresence,
   *   useTypingIndicator,
   * } = db.room(roomType, roomId);
   */
  room(e = "_defaultRoomType", n = "_defaultRoomId") {
    return new Uo(this._core, e, n);
  }
}
class Lo extends Io {
}
class qo extends Do {
}
function Go(t) {
  return new Lo(t);
}
function Xo(t) {
  return new qo(t);
}
var Bt = { exports: {} }, $e = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Nn;
function No() {
  if (Nn) return $e;
  Nn = 1;
  var t = Wn, e = Symbol.for("react.element"), n = Symbol.for("react.fragment"), r = Object.prototype.hasOwnProperty, i = t.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, s = { key: !0, ref: !0, __self: !0, __source: !0 };
  function o(a, c, u) {
    var l, f = {}, p = null, _ = null;
    u !== void 0 && (p = "" + u), c.key !== void 0 && (p = "" + c.key), c.ref !== void 0 && (_ = c.ref);
    for (l in c) r.call(c, l) && !s.hasOwnProperty(l) && (f[l] = c[l]);
    if (a && a.defaultProps) for (l in c = a.defaultProps, c) f[l] === void 0 && (f[l] = c[l]);
    return { $$typeof: e, type: a, key: p, ref: _, props: f, _owner: i.current };
  }
  return $e.Fragment = n, $e.jsx = o, $e.jsxs = o, $e;
}
var De = {}, Fn;
function Fo() {
  if (Fn) return De;
  Fn = 1;
  var t = {};
  /**
   * @license React
   * react-jsx-runtime.development.js
   *
   * Copyright (c) Facebook, Inc. and its affiliates.
   *
   * This source code is licensed under the MIT license found in the
   * LICENSE file in the root directory of this source tree.
   */
  return t.NODE_ENV !== "production" && function() {
    var e = Wn, n = Symbol.for("react.element"), r = Symbol.for("react.portal"), i = Symbol.for("react.fragment"), s = Symbol.for("react.strict_mode"), o = Symbol.for("react.profiler"), a = Symbol.for("react.provider"), c = Symbol.for("react.context"), u = Symbol.for("react.forward_ref"), l = Symbol.for("react.suspense"), f = Symbol.for("react.suspense_list"), p = Symbol.for("react.memo"), _ = Symbol.for("react.lazy"), g = Symbol.for("react.offscreen"), m = Symbol.iterator, O = "@@iterator";
    function S(d) {
      if (d === null || typeof d != "object")
        return null;
      var h = m && d[m] || d[O];
      return typeof h == "function" ? h : null;
    }
    var b = e.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    function y(d) {
      {
        for (var h = arguments.length, v = new Array(h > 1 ? h - 1 : 0), w = 1; w < h; w++)
          v[w - 1] = arguments[w];
        x("error", d, v);
      }
    }
    function x(d, h, v) {
      {
        var w = b.ReactDebugCurrentFrame, T = w.getStackAddendum();
        T !== "" && (h += "%s", v = v.concat([T]));
        var C = v.map(function(P) {
          return String(P);
        });
        C.unshift("Warning: " + h), Function.prototype.apply.call(console[d], console, C);
      }
    }
    var R = !1, z = !1, $ = !1, ce = !1, xe = !1, Ae;
    Ae = Symbol.for("react.module.reference");
    function pt(d) {
      return !!(typeof d == "string" || typeof d == "function" || d === i || d === o || xe || d === s || d === l || d === f || ce || d === g || R || z || $ || typeof d == "object" && d !== null && (d.$$typeof === _ || d.$$typeof === p || d.$$typeof === a || d.$$typeof === c || d.$$typeof === u || // This needs to include all possible module reference object
      // types supported by any Flight configuration anywhere since
      // we don't know which Flight build this will end up being used
      // with.
      d.$$typeof === Ae || d.getModuleId !== void 0));
    }
    function vt(d, h, v) {
      var w = d.displayName;
      if (w)
        return w;
      var T = h.displayName || h.name || "";
      return T !== "" ? v + "(" + T + ")" : v;
    }
    function ue(d) {
      return d.displayName || "Context";
    }
    function J(d) {
      if (d == null)
        return null;
      if (typeof d.tag == "number" && y("Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."), typeof d == "function")
        return d.displayName || d.name || null;
      if (typeof d == "string")
        return d;
      switch (d) {
        case i:
          return "Fragment";
        case r:
          return "Portal";
        case o:
          return "Profiler";
        case s:
          return "StrictMode";
        case l:
          return "Suspense";
        case f:
          return "SuspenseList";
      }
      if (typeof d == "object")
        switch (d.$$typeof) {
          case c:
            var h = d;
            return ue(h) + ".Consumer";
          case a:
            var v = d;
            return ue(v._context) + ".Provider";
          case u:
            return vt(d, d.render, "ForwardRef");
          case p:
            var w = d.displayName || null;
            return w !== null ? w : J(d.type) || "Memo";
          case _: {
            var T = d, C = T._payload, P = T._init;
            try {
              return J(P(C));
            } catch {
              return null;
            }
          }
        }
      return null;
    }
    var ie = Object.assign, pe = 0, I, ee, ve, _e, le, Be, se;
    function Qe() {
    }
    Qe.__reactDisabledLog = !0;
    function _t() {
      {
        if (pe === 0) {
          I = console.log, ee = console.info, ve = console.warn, _e = console.error, le = console.group, Be = console.groupCollapsed, se = console.groupEnd;
          var d = {
            configurable: !0,
            enumerable: !0,
            value: Qe,
            writable: !0
          };
          Object.defineProperties(console, {
            info: d,
            log: d,
            warn: d,
            error: d,
            group: d,
            groupCollapsed: d,
            groupEnd: d
          });
        }
        pe++;
      }
    }
    function bt() {
      {
        if (pe--, pe === 0) {
          var d = {
            configurable: !0,
            enumerable: !0,
            writable: !0
          };
          Object.defineProperties(console, {
            log: ie({}, d, {
              value: I
            }),
            info: ie({}, d, {
              value: ee
            }),
            warn: ie({}, d, {
              value: ve
            }),
            error: ie({}, d, {
              value: _e
            }),
            group: ie({}, d, {
              value: le
            }),
            groupCollapsed: ie({}, d, {
              value: Be
            }),
            groupEnd: ie({}, d, {
              value: se
            })
          });
        }
        pe < 0 && y("disabledDepth fell below zero. This is a bug in React. Please file an issue.");
      }
    }
    var Me = b.ReactCurrentDispatcher, Ie;
    function Je(d, h, v) {
      {
        if (Ie === void 0)
          try {
            throw Error();
          } catch (T) {
            var w = T.stack.trim().match(/\n( *(at )?)/);
            Ie = w && w[1] || "";
          }
        return `
` + Ie + d;
      }
    }
    var yt = !1, Ke;
    {
      var Pr = typeof WeakMap == "function" ? WeakMap : Map;
      Ke = new Pr();
    }
    function rn(d, h) {
      if (!d || yt)
        return "";
      {
        var v = Ke.get(d);
        if (v !== void 0)
          return v;
      }
      var w;
      yt = !0;
      var T = Error.prepareStackTrace;
      Error.prepareStackTrace = void 0;
      var C;
      C = Me.current, Me.current = null, _t();
      try {
        if (h) {
          var P = function() {
            throw Error();
          };
          if (Object.defineProperty(P.prototype, "props", {
            set: function() {
              throw Error();
            }
          }), typeof Reflect == "object" && Reflect.construct) {
            try {
              Reflect.construct(P, []);
            } catch (W) {
              w = W;
            }
            Reflect.construct(d, [], P);
          } else {
            try {
              P.call();
            } catch (W) {
              w = W;
            }
            d.call(P.prototype);
          }
        } else {
          try {
            throw Error();
          } catch (W) {
            w = W;
          }
          d();
        }
      } catch (W) {
        if (W && w && typeof W.stack == "string") {
          for (var E = W.stack.split(`
`), q = w.stack.split(`
`), A = E.length - 1, k = q.length - 1; A >= 1 && k >= 0 && E[A] !== q[k]; )
            k--;
          for (; A >= 1 && k >= 0; A--, k--)
            if (E[A] !== q[k]) {
              if (A !== 1 || k !== 1)
                do
                  if (A--, k--, k < 0 || E[A] !== q[k]) {
                    var G = `
` + E[A].replace(" at new ", " at ");
                    return d.displayName && G.includes("<anonymous>") && (G = G.replace("<anonymous>", d.displayName)), typeof d == "function" && Ke.set(d, G), G;
                  }
                while (A >= 1 && k >= 0);
              break;
            }
        }
      } finally {
        yt = !1, Me.current = C, bt(), Error.prepareStackTrace = T;
      }
      var Re = d ? d.displayName || d.name : "", be = Re ? Je(Re) : "";
      return typeof d == "function" && Ke.set(d, be), be;
    }
    function Tr(d, h, v) {
      return rn(d, !1);
    }
    function Cr(d) {
      var h = d.prototype;
      return !!(h && h.isReactComponent);
    }
    function Ye(d, h, v) {
      if (d == null)
        return "";
      if (typeof d == "function")
        return rn(d, Cr(d));
      if (typeof d == "string")
        return Je(d);
      switch (d) {
        case l:
          return Je("Suspense");
        case f:
          return Je("SuspenseList");
      }
      if (typeof d == "object")
        switch (d.$$typeof) {
          case u:
            return Tr(d.render);
          case p:
            return Ye(d.type, h, v);
          case _: {
            var w = d, T = w._payload, C = w._init;
            try {
              return Ye(C(T), h, v);
            } catch {
            }
          }
        }
      return "";
    }
    var ke = Object.prototype.hasOwnProperty, sn = {}, on = b.ReactDebugCurrentFrame;
    function He(d) {
      if (d) {
        var h = d._owner, v = Ye(d.type, d._source, h ? h.type : null);
        on.setExtraStackFrame(v);
      } else
        on.setExtraStackFrame(null);
    }
    function xr(d, h, v, w, T) {
      {
        var C = Function.call.bind(ke);
        for (var P in d)
          if (C(d, P)) {
            var E = void 0;
            try {
              if (typeof d[P] != "function") {
                var q = Error((w || "React class") + ": " + v + " type `" + P + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof d[P] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.");
                throw q.name = "Invariant Violation", q;
              }
              E = d[P](h, P, w, v, null, "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED");
            } catch (A) {
              E = A;
            }
            E && !(E instanceof Error) && (He(T), y("%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).", w || "React class", v, P, typeof E), He(null)), E instanceof Error && !(E.message in sn) && (sn[E.message] = !0, He(T), y("Failed %s type: %s", v, E.message), He(null));
          }
      }
    }
    var Ar = Array.isArray;
    function gt(d) {
      return Ar(d);
    }
    function Mr(d) {
      {
        var h = typeof Symbol == "function" && Symbol.toStringTag, v = h && d[Symbol.toStringTag] || d.constructor.name || "Object";
        return v;
      }
    }
    function Ir(d) {
      try {
        return an(d), !1;
      } catch {
        return !0;
      }
    }
    function an(d) {
      return "" + d;
    }
    function cn(d) {
      if (Ir(d))
        return y("The provided key is an unsupported type %s. This value must be coerced to a string before before using it here.", Mr(d)), an(d);
    }
    var Ue = b.ReactCurrentOwner, kr = {
      key: !0,
      ref: !0,
      __self: !0,
      __source: !0
    }, un, ln, wt;
    wt = {};
    function Ur(d) {
      if (ke.call(d, "ref")) {
        var h = Object.getOwnPropertyDescriptor(d, "ref").get;
        if (h && h.isReactWarning)
          return !1;
      }
      return d.ref !== void 0;
    }
    function $r(d) {
      if (ke.call(d, "key")) {
        var h = Object.getOwnPropertyDescriptor(d, "key").get;
        if (h && h.isReactWarning)
          return !1;
      }
      return d.key !== void 0;
    }
    function Dr(d, h) {
      if (typeof d.ref == "string" && Ue.current && h && Ue.current.stateNode !== h) {
        var v = J(Ue.current.type);
        wt[v] || (y('Component "%s" contains the string ref "%s". Support for string refs will be removed in a future major release. This case cannot be automatically converted to an arrow function. We ask you to manually fix this case by using useRef() or createRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref', J(Ue.current.type), d.ref), wt[v] = !0);
      }
    }
    function Lr(d, h) {
      {
        var v = function() {
          un || (un = !0, y("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", h));
        };
        v.isReactWarning = !0, Object.defineProperty(d, "key", {
          get: v,
          configurable: !0
        });
      }
    }
    function qr(d, h) {
      {
        var v = function() {
          ln || (ln = !0, y("%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", h));
        };
        v.isReactWarning = !0, Object.defineProperty(d, "ref", {
          get: v,
          configurable: !0
        });
      }
    }
    var Nr = function(d, h, v, w, T, C, P) {
      var E = {
        // This tag allows us to uniquely identify this as a React Element
        $$typeof: n,
        // Built-in properties that belong on the element
        type: d,
        key: h,
        ref: v,
        props: P,
        // Record the component responsible for creating this element.
        _owner: C
      };
      return E._store = {}, Object.defineProperty(E._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: !1
      }), Object.defineProperty(E, "_self", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: w
      }), Object.defineProperty(E, "_source", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: T
      }), Object.freeze && (Object.freeze(E.props), Object.freeze(E)), E;
    };
    function Fr(d, h, v, w, T) {
      {
        var C, P = {}, E = null, q = null;
        v !== void 0 && (cn(v), E = "" + v), $r(h) && (cn(h.key), E = "" + h.key), Ur(h) && (q = h.ref, Dr(h, T));
        for (C in h)
          ke.call(h, C) && !kr.hasOwnProperty(C) && (P[C] = h[C]);
        if (d && d.defaultProps) {
          var A = d.defaultProps;
          for (C in A)
            P[C] === void 0 && (P[C] = A[C]);
        }
        if (E || q) {
          var k = typeof d == "function" ? d.displayName || d.name || "Unknown" : d;
          E && Lr(P, k), q && qr(P, k);
        }
        return Nr(d, E, q, T, w, Ue.current, P);
      }
    }
    var mt = b.ReactCurrentOwner, dn = b.ReactDebugCurrentFrame;
    function Se(d) {
      if (d) {
        var h = d._owner, v = Ye(d.type, d._source, h ? h.type : null);
        dn.setExtraStackFrame(v);
      } else
        dn.setExtraStackFrame(null);
    }
    var Ot;
    Ot = !1;
    function St(d) {
      return typeof d == "object" && d !== null && d.$$typeof === n;
    }
    function fn() {
      {
        if (mt.current) {
          var d = J(mt.current.type);
          if (d)
            return `

Check the render method of \`` + d + "`.";
        }
        return "";
      }
    }
    function zr(d) {
      return "";
    }
    var hn = {};
    function Wr(d) {
      {
        var h = fn();
        if (!h) {
          var v = typeof d == "string" ? d : d.displayName || d.name;
          v && (h = `

Check the top-level render call using <` + v + ">.");
        }
        return h;
      }
    }
    function pn(d, h) {
      {
        if (!d._store || d._store.validated || d.key != null)
          return;
        d._store.validated = !0;
        var v = Wr(h);
        if (hn[v])
          return;
        hn[v] = !0;
        var w = "";
        d && d._owner && d._owner !== mt.current && (w = " It was passed a child from " + J(d._owner.type) + "."), Se(d), y('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.', v, w), Se(null);
      }
    }
    function vn(d, h) {
      {
        if (typeof d != "object")
          return;
        if (gt(d))
          for (var v = 0; v < d.length; v++) {
            var w = d[v];
            St(w) && pn(w, h);
          }
        else if (St(d))
          d._store && (d._store.validated = !0);
        else if (d) {
          var T = S(d);
          if (typeof T == "function" && T !== d.entries)
            for (var C = T.call(d), P; !(P = C.next()).done; )
              St(P.value) && pn(P.value, h);
        }
      }
    }
    function Vr(d) {
      {
        var h = d.type;
        if (h == null || typeof h == "string")
          return;
        var v;
        if (typeof h == "function")
          v = h.propTypes;
        else if (typeof h == "object" && (h.$$typeof === u || // Note: Memo only checks outer props here.
        // Inner props are checked in the reconciler.
        h.$$typeof === p))
          v = h.propTypes;
        else
          return;
        if (v) {
          var w = J(h);
          xr(v, d.props, "prop", w, d);
        } else if (h.PropTypes !== void 0 && !Ot) {
          Ot = !0;
          var T = J(h);
          y("Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?", T || "Unknown");
        }
        typeof h.getDefaultProps == "function" && !h.getDefaultProps.isReactClassApproved && y("getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.");
      }
    }
    function Br(d) {
      {
        for (var h = Object.keys(d.props), v = 0; v < h.length; v++) {
          var w = h[v];
          if (w !== "children" && w !== "key") {
            Se(d), y("Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.", w), Se(null);
            break;
          }
        }
        d.ref !== null && (Se(d), y("Invalid attribute `ref` supplied to `React.Fragment`."), Se(null));
      }
    }
    var _n = {};
    function bn(d, h, v, w, T, C) {
      {
        var P = pt(d);
        if (!P) {
          var E = "";
          (d === void 0 || typeof d == "object" && d !== null && Object.keys(d).length === 0) && (E += " You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");
          var q = zr();
          q ? E += q : E += fn();
          var A;
          d === null ? A = "null" : gt(d) ? A = "array" : d !== void 0 && d.$$typeof === n ? (A = "<" + (J(d.type) || "Unknown") + " />", E = " Did you accidentally export a JSX literal instead of a component?") : A = typeof d, y("React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s", A, E);
        }
        var k = Fr(d, h, v, T, C);
        if (k == null)
          return k;
        if (P) {
          var G = h.children;
          if (G !== void 0)
            if (w)
              if (gt(G)) {
                for (var Re = 0; Re < G.length; Re++)
                  vn(G[Re], d);
                Object.freeze && Object.freeze(G);
              } else
                y("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");
            else
              vn(G, d);
        }
        if (ke.call(h, "key")) {
          var be = J(d), W = Object.keys(h).filter(function(Gr) {
            return Gr !== "key";
          }), Rt = W.length > 0 ? "{key: someKey, " + W.join(": ..., ") + ": ...}" : "{key: someKey}";
          if (!_n[be + Rt]) {
            var Hr = W.length > 0 ? "{" + W.join(": ..., ") + ": ...}" : "{}";
            y(`A props object containing a "key" prop is being spread into JSX:
  let props = %s;
  <%s {...props} />
React keys must be passed directly to JSX without using spread:
  let props = %s;
  <%s key={someKey} {...props} />`, Rt, be, Hr, be), _n[be + Rt] = !0;
          }
        }
        return d === i ? Br(k) : Vr(k), k;
      }
    }
    function Qr(d, h, v) {
      return bn(d, h, v, !0);
    }
    function Jr(d, h, v) {
      return bn(d, h, v, !1);
    }
    var Kr = Jr, Yr = Qr;
    De.Fragment = i, De.jsx = Kr, De.jsxs = Yr;
  }(), De;
}
var zo = {};
zo.NODE_ENV === "production" ? Bt.exports = No() : Bt.exports = Fo();
var K = Bt.exports;
function Zo({
  as: t = "div",
  spaceId: e,
  room: n,
  className: r,
  style: i,
  userCursorColor: s,
  children: o,
  renderCursor: a,
  propagate: c,
  zIndex: u
}) {
  const l = e || `cursors-space-default--${String(n.type)}-${n.id}`, f = n.usePresence({
    keys: [l]
  }), p = n._core._reactor.getPresence(n.type, n.id);
  function _(b, y) {
    const x = y.clientX, R = y.clientY, z = (x - b.left) / b.width * 100, $ = (R - b.top) / b.height * 100;
    f.publishPresence({
      [l]: {
        x,
        y: R,
        xPercent: z,
        yPercent: $,
        color: s
      }
    });
  }
  function g(b) {
    c || b.stopPropagation();
    const y = b.currentTarget.getBoundingClientRect();
    _(y, b);
  }
  function m(b) {
    f.publishPresence({
      [l]: void 0
    });
  }
  function O(b) {
    if (b.touches.length !== 1)
      return;
    const y = b.touches[0];
    if (y.target instanceof Element) {
      c || b.stopPropagation();
      const x = y.target.getBoundingClientRect();
      _(x, y);
    }
  }
  function S(b) {
    f.publishPresence({
      [l]: void 0
    });
  }
  return Xr(
    t,
    {
      onMouseMove: g,
      onMouseOut: m,
      onTouchMove: O,
      onTouchEnd: S,
      className: r,
      style: {
        position: "relative",
        ...i
      }
    },
    [
      o,
      /* @__PURE__ */ K.jsx(
        "div",
        {
          style: {
            ...zn,
            ...Vo,
            zIndex: u !== void 0 ? u : Bo
          },
          children: Object.entries(f.peers).map(([b, y]) => {
            const x = y[l];
            return x ? /* @__PURE__ */ K.jsx(
              "div",
              {
                style: {
                  ...zn,
                  transform: `translate(${x.xPercent}%, ${x.yPercent}%)`,
                  transformOrigin: "0 0",
                  transition: "transform 100ms"
                },
                children: a ? a({
                  color: x.color,
                  presence: p.peers[b]
                }) : /* @__PURE__ */ K.jsx(Wo, { ...x })
              },
              b
            ) : null;
          })
        },
        l
      )
    ]
  );
}
function Wo({ color: t }) {
  const n = t || "black";
  return /* @__PURE__ */ K.jsxs(
    "svg",
    {
      style: { height: 35, width: 35 },
      viewBox: "0 0 35 35",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg",
      children: [
        /* @__PURE__ */ K.jsxs(
          "g",
          {
            fill: "rgba(0,0,0,.2)",
            transform: "matrix(1, 0, 0, 1, -11.999999046325684, -8.406899452209473)",
            children: [
              /* @__PURE__ */ K.jsx("path", { d: "m12 24.4219v-16.015l11.591 11.619h-6.781l-.411.124z" }),
              /* @__PURE__ */ K.jsx("path", { d: "m21.0845 25.0962-3.605 1.535-4.682-11.089 3.686-1.553z" })
            ]
          }
        ),
        /* @__PURE__ */ K.jsxs(
          "g",
          {
            fill: "white",
            transform: "matrix(1, 0, 0, 1, -11.999999046325684, -8.406899452209473)",
            children: [
              /* @__PURE__ */ K.jsx("path", { d: "m12 24.4219v-16.015l11.591 11.619h-6.781l-.411.124z" }),
              /* @__PURE__ */ K.jsx("path", { d: "m21.0845 25.0962-3.605 1.535-4.682-11.089 3.686-1.553z" })
            ]
          }
        ),
        /* @__PURE__ */ K.jsxs(
          "g",
          {
            fill: n,
            transform: "matrix(1, 0, 0, 1, -11.999999046325684, -8.406899452209473)",
            children: [
              /* @__PURE__ */ K.jsx("path", { d: "m19.751 24.4155-1.844.774-3.1-7.374 1.841-.775z" }),
              /* @__PURE__ */ K.jsx("path", { d: "m13 10.814v11.188l2.969-2.866.428-.139h4.768z" })
            ]
          }
        )
      ]
    }
  );
}
const zn = {
  position: "absolute",
  top: 0,
  left: 0,
  bottom: 0,
  right: 0
}, Vo = {
  overflow: "hidden",
  pointerEvents: "none",
  userSelect: "none"
}, Bo = 99999;
export {
  Zo as Cursors,
  Io as InstantReact,
  Do as InstantReactAbstractDatabase,
  Lo as InstantReactWeb,
  qo as InstantReactWebDatabase,
  Yo as i,
  L as id,
  Go as init,
  Xo as init_experimental,
  Jo as lookup,
  Ko as tx
};
