// @instantdb/core/dist/standalone/index.mjs
var _n = Object.defineProperty;
var bn = (t, e, n) => e in t ? _n(t, e, { enumerable: !0, configurable: !0, writable: !0, value: n }) : t[e] = n;
var p = (t, e, n) => bn(t, typeof e != "symbol" ? e + "" : e, n);
let he = !1;
typeof window < "u" && typeof window.localStorage < "u" && (he = !!window.localStorage.getItem("devBackend") || !!window.localStorage.getItem("__instantLogging"));
const E = {
  info: he ? console.info.bind(console) : () => {
  },
  debug: he ? console.debug.bind(console) : () => {
  },
  error: he ? console.error.bind(console) : () => {
  }
};
function U(t) {
  if (typeof t == "number")
    return (Math.abs(t * 2654435761) >>> 0).toString(16);
  if (typeof t == "boolean") return t ? "1" : "0";
  if (t === null) return "null";
  if (t === void 0) return "undefined";
  if (typeof t == "string") {
    let e = 2166136261;
    for (let n = 0; n < t.length; n++)
      e ^= t.charCodeAt(n), e += (e << 1) + (e << 4) + (e << 7) + (e << 8) + (e << 24), e = e >>> 0;
    return e.toString(16);
  }
  if (Array.isArray(t)) {
    let e = 2166136261;
    for (let n = 0; n < t.length; n++) {
      e ^= (n + 1) * 2654435761;
      const r = U(t[n]);
      for (let s = 0; s < r.length; s++)
        e ^= r.charCodeAt(s), e *= 16777619, e = e >>> 0;
    }
    return e.toString(16);
  }
  if (typeof t == "object") {
    let e = 2166136261;
    const n = Object.keys(t).sort();
    for (let r = 0; r < n.length; r++) {
      const s = n[r], i = U(s);
      e ^= parseInt(i, 16), e *= 16777619, e = e >>> 0;
      const o = U(t[s]);
      e ^= parseInt(o, 16), e *= 16777619, e = e >>> 0;
    }
    return e.toString(16);
  }
  return U(String(t));
}
const v = {
  Remove: "remove",
  Replace: "replace",
  Add: "add"
}, kt = Symbol.for("__MUTATIVE_PROXY_DRAFT__"), wn = Symbol("__MUTATIVE_RAW_RETURN_SYMBOL__"), pe = Symbol.iterator, I = {
  mutable: "mutable",
  immutable: "immutable"
}, Je = {};
function T(t) {
  var e;
  return (e = t.copy) !== null && e !== void 0 ? e : t.original;
}
function Y(t) {
  return !!b(t);
}
function b(t) {
  return typeof t != "object" ? null : t == null ? void 0 : t[kt];
}
function Be(t) {
  var e;
  const n = b(t);
  return n ? (e = n.copy) !== null && e !== void 0 ? e : n.original : t;
}
function N(t, e) {
  if (!t || typeof t != "object")
    return !1;
  let n;
  return Object.getPrototypeOf(t) === Object.prototype || Array.isArray(t) || t instanceof Map || t instanceof Set || !!(e != null && e.mark) && ((n = e.mark(t, I)) === I.immutable || typeof n == "function");
}
function xt(t, e = []) {
  if (Object.hasOwnProperty.call(t, "key")) {
    const n = b(W(t.parent.copy, t.key));
    if (n !== null && (n == null ? void 0 : n.original) !== t.original)
      return null;
    e.push(t.parent.type === 3 ? Array.from(t.parent.setMap.keys()).indexOf(t.key) : t.key);
  }
  if (t.parent)
    return xt(t.parent, e);
  e.reverse();
  try {
    gn(t.copy, e);
  } catch {
    return null;
  }
  return e;
}
function X(t) {
  return Array.isArray(t) ? 1 : t instanceof Map ? 2 : t instanceof Set ? 3 : 0;
}
function W(t, e) {
  return X(t) === 2 ? t.get(e) : t[e];
}
function ce(t, e, n) {
  X(t) === 2 ? t.set(e, n) : t[e] = n;
}
function Pe(t, e) {
  const n = b(t);
  return (n ? T(n) : t)[e];
}
function Q(t, e) {
  return t === e ? t !== 0 || 1 / t === 1 / e : t !== t && e !== e;
}
function je(t) {
  if (t)
    for (; t.finalities.revoke.length > 0; )
      t.finalities.revoke.pop()();
}
function H(t, e) {
  return e ? t : [""].concat(t).map((n) => {
    const r = `${n}`;
    return r.indexOf("/") === -1 && r.indexOf("~") === -1 ? r : r.replace(/~/g, "~0").replace(/\//g, "~1");
  }).join("/");
}
function gn(t, e) {
  for (let n = 0; n < e.length - 1; n += 1) {
    const r = e[n];
    if (t = W(X(t) === 3 ? Array.from(t) : t, r), typeof t != "object")
      throw new Error(`Cannot resolve patch at '${e.join("/")}'.`);
  }
  return t;
}
function mn(t) {
  const e = Object.create(Object.getPrototypeOf(t));
  return Reflect.ownKeys(t).forEach((n) => {
    let r = Reflect.getOwnPropertyDescriptor(t, n);
    if (r.enumerable && r.configurable && r.writable) {
      e[n] = t[n];
      return;
    }
    r.writable || (r.writable = !0, r.configurable = !0), (r.get || r.set) && (r = {
      configurable: !0,
      writable: !0,
      enumerable: r.enumerable,
      value: t[n]
    }), Reflect.defineProperty(e, n, r);
  }), e;
}
const On = Object.prototype.propertyIsEnumerable;
function Ut(t, e) {
  let n;
  if (Array.isArray(t))
    return Array.prototype.concat.call(t);
  if (t instanceof Set)
    return new Set(t.values());
  if (t instanceof Map)
    return new Map(t);
  if (e != null && e.mark && (n = e.mark(t, I), n !== void 0) && n !== I.mutable) {
    if (n === I.immutable)
      return mn(t);
    if (typeof n == "function") {
      if (e.enablePatches || e.enableAutoFreeze)
        throw new Error("You can't use mark and patches or auto freeze together.");
      return n();
    }
    throw new Error(`Unsupported mark result: ${n}`);
  } else if (typeof t == "object" && Object.getPrototypeOf(t) === Object.prototype) {
    const r = {};
    return Object.keys(t).forEach((s) => {
      r[s] = t[s];
    }), Object.getOwnPropertySymbols(t).forEach((s) => {
      On.call(t, s) && (r[s] = t[s]);
    }), r;
  } else
    throw new Error("Please check mark() to ensure that it is a stable marker draftable function.");
}
function C(t) {
  t.copy || (t.copy = Ut(t.original, t.options));
}
function ie(t) {
  if (!N(t))
    return Be(t);
  if (Array.isArray(t))
    return t.map(ie);
  if (t instanceof Map)
    return new Map(Array.from(t.entries()).map(([n, r]) => [n, ie(r)]));
  if (t instanceof Set)
    return new Set(Array.from(t).map(ie));
  const e = Object.create(Object.getPrototypeOf(t));
  for (const n in t)
    e[n] = ie(t[n]);
  return e;
}
function ye(t) {
  return Y(t) ? ie(t) : t;
}
function z(t) {
  var e;
  t.assignedMap = (e = t.assignedMap) !== null && e !== void 0 ? e : /* @__PURE__ */ new Map(), t.operated || (t.operated = !0, t.parent && z(t.parent));
}
function dt() {
  throw new Error("Cannot modify frozen object");
}
function te(t, e, n, r, s) {
  {
    n = n ?? /* @__PURE__ */ new WeakMap(), r = r ?? [], s = s ?? [];
    const o = n.has(t) ? n.get(t) : t;
    if (r.length > 0) {
      const a = r.indexOf(o);
      if (o && typeof o == "object" && a !== -1)
        throw r[0] === o ? new Error("Forbids circular reference") : new Error(`Forbids circular reference: ~/${s.slice(0, a).map((c, u) => {
          if (typeof c == "symbol")
            return `[${c.toString()}]`;
          const f = r[u];
          return typeof c == "object" && (f instanceof Map || f instanceof Set) ? Array.from(f.keys()).indexOf(c) : c;
        }).join("/")}`);
      r.push(o), s.push(e);
    } else
      r.push(o);
  }
  if (Object.isFrozen(t) || Y(t)) {
    r.pop(), s.pop();
    return;
  }
  switch (X(t)) {
    case 2:
      for (const [a, c] of t)
        te(a, a, n, r, s), te(c, a, n, r, s);
      t.set = t.clear = t.delete = dt;
      break;
    case 3:
      for (const a of t)
        te(a, a, n, r, s);
      t.add = t.clear = t.delete = dt;
      break;
    case 1:
      Object.freeze(t);
      let o = 0;
      for (const a of t)
        te(a, o, n, r, s), o += 1;
      break;
    default:
      Object.freeze(t), Object.keys(t).forEach((a) => {
        const c = t[a];
        te(c, a, n, r, s);
      });
  }
  r.pop(), s.pop();
}
function _e(t, e) {
  return t instanceof Map ? t.has(e) : Object.prototype.hasOwnProperty.call(t, e);
}
function ht(t, e) {
  if (e in t) {
    let n = Reflect.getPrototypeOf(t);
    for (; n; ) {
      const r = Reflect.getOwnPropertyDescriptor(n, e);
      if (r)
        return r;
      n = Reflect.getPrototypeOf(n);
    }
  }
}
function Ke(t, e) {
  const n = X(t);
  if (n === 0)
    Reflect.ownKeys(t).forEach((r) => {
      e(r, t[r], t);
    });
  else if (n === 1) {
    let r = 0;
    for (const s of t)
      e(r, s, t), r += 1;
  } else
    t.forEach((r, s) => e(s, r, t));
}
function $t(t, e, n) {
  if (Y(t) || !N(t, n) || e.has(t) || Object.isFrozen(t))
    return;
  const r = t instanceof Set, s = r ? /* @__PURE__ */ new Map() : void 0;
  if (e.add(t), Ke(t, (i, o) => {
    var a;
    if (Y(o)) {
      const c = b(o);
      C(c);
      const u = !((a = c.assignedMap) === null || a === void 0) && a.size || c.operated ? c.copy : c.original;
      ce(r ? s : t, i, u);
    } else
      $t(o, e, n);
  }), s) {
    const i = t, o = Array.from(i);
    i.clear(), o.forEach((a) => {
      i.add(s.has(a) ? s.get(a) : a);
    });
  }
}
function Sn(t, e) {
  const n = t.type === 3 ? t.setMap : t.copy;
  t.finalities.revoke.length > 1 && t.assignedMap.get(e) && n && $t(W(n, e), t.finalities.handledSet, t.options);
}
function Te(t) {
  t.type === 3 && t.copy && (t.copy.clear(), t.setMap.forEach((e) => {
    t.copy.add(Be(e));
  }));
}
function Ie(t, e, n, r) {
  if (t.operated && t.assignedMap && t.assignedMap.size > 0 && !t.finalized) {
    if (n && r) {
      const i = xt(t);
      i && e(t, i, n, r);
    }
    t.finalized = !0;
  }
}
function He(t, e, n, r) {
  const s = b(n);
  s && (s.callbacks || (s.callbacks = []), s.callbacks.push((i, o) => {
    var a;
    const c = t.type === 3 ? t.setMap : t.copy;
    if (Q(W(c, e), n)) {
      let u = s.original;
      s.copy && (u = s.copy), Te(t), Ie(t, r, i, o), t.options.enableAutoFreeze && (t.options.updatedValues = (a = t.options.updatedValues) !== null && a !== void 0 ? a : /* @__PURE__ */ new WeakMap(), t.options.updatedValues.set(u, s.original)), ce(c, e, u);
    }
  }), t.options.enableAutoFreeze && s.finalities !== t.finalities && (t.options.enableAutoFreeze = !1)), N(n, t.options) && t.finalities.draft.push(() => {
    const i = t.type === 3 ? t.setMap : t.copy;
    Q(W(i, e), n) && Sn(t, e);
  });
}
function vn(t, e, n, r, s) {
  let { original: i, assignedMap: o, options: a } = t, c = t.copy;
  c.length < i.length && ([i, c] = [c, i], [n, r] = [r, n]);
  for (let u = 0; u < i.length; u += 1)
    if (o.get(u.toString()) && c[u] !== i[u]) {
      const f = e.concat([u]), l = H(f, s);
      n.push({
        op: v.Replace,
        path: l,
        // If it is a draft, it needs to be deep cloned, and it may also be non-draft.
        value: ye(c[u])
      }), r.push({
        op: v.Replace,
        path: l,
        // If it is a draft, it needs to be deep cloned, and it may also be non-draft.
        value: ye(i[u])
      });
    }
  for (let u = i.length; u < c.length; u += 1) {
    const f = e.concat([u]), l = H(f, s);
    n.push({
      op: v.Add,
      path: l,
      // If it is a draft, it needs to be deep cloned, and it may also be non-draft.
      value: ye(c[u])
    });
  }
  if (i.length < c.length) {
    const { arrayLengthAssignment: u = !0 } = a.enablePatches;
    if (u) {
      const f = e.concat(["length"]), l = H(f, s);
      r.push({
        op: v.Replace,
        path: l,
        value: i.length
      });
    } else
      for (let f = c.length; i.length < f; f -= 1) {
        const l = e.concat([f - 1]), d = H(l, s);
        r.push({
          op: v.Remove,
          path: d
        });
      }
  }
}
function Mn({ original: t, copy: e, assignedMap: n }, r, s, i, o) {
  n.forEach((a, c) => {
    const u = W(t, c), f = ye(W(e, c)), l = a ? _e(t, c) ? v.Replace : v.Add : v.Remove;
    if (Q(u, f) && l === v.Replace)
      return;
    const d = r.concat(c), h = H(d, o);
    s.push(l === v.Remove ? { op: l, path: h } : { op: l, path: h, value: f }), i.push(l === v.Add ? { op: v.Remove, path: h } : l === v.Remove ? { op: v.Add, path: h, value: u } : { op: v.Replace, path: h, value: u });
  });
}
function An({ original: t, copy: e }, n, r, s, i) {
  let o = 0;
  t.forEach((a) => {
    if (!e.has(a)) {
      const c = n.concat([o]), u = H(c, i);
      r.push({
        op: v.Remove,
        path: u,
        value: a
      }), s.unshift({
        op: v.Add,
        path: u,
        value: a
      });
    }
    o += 1;
  }), o = 0, e.forEach((a) => {
    if (!t.has(a)) {
      const c = n.concat([o]), u = H(c, i);
      r.push({
        op: v.Add,
        path: u,
        value: a
      }), s.unshift({
        op: v.Remove,
        path: u,
        value: a
      });
    }
    o += 1;
  });
}
function ae(t, e, n, r) {
  const { pathAsArray: s = !0 } = t.options.enablePatches;
  switch (t.type) {
    case 0:
    case 2:
      return Mn(t, e, n, r, s);
    case 1:
      return vn(t, e, n, r, s);
    case 3:
      return An(t, e, n, r, s);
  }
}
let Pn = !1;
const we = (t, e, n = !1) => {
  if (typeof t == "object" && t !== null && (!N(t, e) || n) && !Pn)
    throw new Error("Strict mode: Mutable data cannot be accessed directly, please use 'unsafe(callback)' wrap.");
}, ke = {
  get size() {
    return T(b(this)).size;
  },
  has(t) {
    return T(b(this)).has(t);
  },
  set(t, e) {
    const n = b(this), r = T(n);
    return (!r.has(t) || !Q(r.get(t), e)) && (C(n), z(n), n.assignedMap.set(t, !0), n.copy.set(t, e), He(n, t, e, ae)), this;
  },
  delete(t) {
    if (!this.has(t))
      return !1;
    const e = b(this);
    return C(e), z(e), e.original.has(t) ? e.assignedMap.set(t, !1) : e.assignedMap.delete(t), e.copy.delete(t), !0;
  },
  clear() {
    const t = b(this);
    if (this.size) {
      C(t), z(t), t.assignedMap = /* @__PURE__ */ new Map();
      for (const [e] of t.original)
        t.assignedMap.set(e, !1);
      t.copy.clear();
    }
  },
  forEach(t, e) {
    const n = b(this);
    T(n).forEach((r, s) => {
      t.call(e, this.get(s), s, this);
    });
  },
  get(t) {
    var e, n;
    const r = b(this), s = T(r).get(t), i = ((n = (e = r.options).mark) === null || n === void 0 ? void 0 : n.call(e, s, I)) === I.mutable;
    if (r.options.strict && we(s, r.options, i), i || r.finalized || !N(s, r.options) || s !== r.original.get(t))
      return s;
    const o = Je.createDraft({
      original: s,
      parentDraft: r,
      key: t,
      finalities: r.finalities,
      options: r.options
    });
    return C(r), r.copy.set(t, o), o;
  },
  keys() {
    return T(b(this)).keys();
  },
  values() {
    const t = this.keys();
    return {
      [pe]: () => this.values(),
      next: () => {
        const e = t.next();
        return e.done ? e : {
          done: !1,
          value: this.get(e.value)
        };
      }
    };
  },
  entries() {
    const t = this.keys();
    return {
      [pe]: () => this.entries(),
      next: () => {
        const e = t.next();
        if (e.done)
          return e;
        const n = this.get(e.value);
        return {
          done: !1,
          value: [e.value, n]
        };
      }
    };
  },
  [pe]() {
    return this.entries();
  }
}, En = Reflect.ownKeys(ke), pt = (t, e, { isValuesIterator: n }) => () => {
  var r, s;
  const i = e.next();
  if (i.done)
    return i;
  const o = i.value;
  let a = t.setMap.get(o);
  const c = b(a), u = ((s = (r = t.options).mark) === null || s === void 0 ? void 0 : s.call(r, a, I)) === I.mutable;
  if (t.options.strict && we(o, t.options, u), !u && !c && N(o, t.options) && !t.finalized && t.original.has(o)) {
    const f = Je.createDraft({
      original: o,
      parentDraft: t,
      key: o,
      finalities: t.finalities,
      options: t.options
    });
    t.setMap.set(o, f), a = f;
  } else c && (a = c.proxy);
  return {
    done: !1,
    value: n ? a : [a, a]
  };
}, xe = {
  get size() {
    return b(this).setMap.size;
  },
  has(t) {
    const e = b(this);
    if (e.setMap.has(t))
      return !0;
    C(e);
    const n = b(t);
    return !!(n && e.setMap.has(n.original));
  },
  add(t) {
    const e = b(this);
    return this.has(t) || (C(e), z(e), e.assignedMap.set(t, !0), e.setMap.set(t, t), He(e, t, t, ae)), this;
  },
  delete(t) {
    if (!this.has(t))
      return !1;
    const e = b(this);
    C(e), z(e);
    const n = b(t);
    return n && e.setMap.has(n.original) ? (e.assignedMap.set(n.original, !1), e.setMap.delete(n.original)) : (!n && e.setMap.has(t) ? e.assignedMap.set(t, !1) : e.assignedMap.delete(t), e.setMap.delete(t));
  },
  clear() {
    if (!this.size)
      return;
    const t = b(this);
    C(t), z(t);
    for (const e of t.original)
      t.assignedMap.set(e, !1);
    t.setMap.clear();
  },
  values() {
    const t = b(this);
    C(t);
    const e = t.setMap.keys();
    return {
      [Symbol.iterator]: () => this.values(),
      next: pt(t, e, { isValuesIterator: !0 })
    };
  },
  entries() {
    const t = b(this);
    C(t);
    const e = t.setMap.keys();
    return {
      [Symbol.iterator]: () => this.entries(),
      next: pt(t, e, {
        isValuesIterator: !1
      })
    };
  },
  keys() {
    return this.values();
  },
  [pe]() {
    return this.values();
  },
  forEach(t, e) {
    const n = this.values();
    let r = n.next();
    for (; !r.done; )
      t.call(e, r.value, r.value, this), r = n.next();
  }
}, Cn = Reflect.ownKeys(xe), Dt = /* @__PURE__ */ new WeakSet(), qt = {
  get(t, e, n) {
    var r, s;
    const i = (r = t.copy) === null || r === void 0 ? void 0 : r[e];
    if (i && Dt.has(i))
      return i;
    if (e === kt)
      return t;
    let o;
    if (t.options.mark) {
      const u = e === "size" && (t.original instanceof Map || t.original instanceof Set) ? Reflect.get(t.original, e) : Reflect.get(t.original, e, n);
      if (o = t.options.mark(u, I), o === I.mutable)
        return t.options.strict && we(u, t.options, !0), u;
    }
    const a = T(t);
    if (a instanceof Map && En.includes(e)) {
      if (e === "size")
        return Object.getOwnPropertyDescriptor(ke, "size").get.call(t.proxy);
      const u = ke[e];
      if (u)
        return u.bind(t.proxy);
    }
    if (a instanceof Set && Cn.includes(e)) {
      if (e === "size")
        return Object.getOwnPropertyDescriptor(xe, "size").get.call(t.proxy);
      const u = xe[e];
      if (u)
        return u.bind(t.proxy);
    }
    if (!_e(a, e)) {
      const u = ht(a, e);
      return u ? "value" in u ? u.value : (
        // !case: support for getter
        (s = u.get) === null || s === void 0 ? void 0 : s.call(t.proxy)
      ) : void 0;
    }
    const c = a[e];
    if (t.options.strict && we(c, t.options), t.finalized || !N(c, t.options))
      return c;
    if (c === Pe(t.original, e)) {
      if (C(t), t.copy[e] = Ge({
        original: t.original[e],
        parentDraft: t,
        key: t.type === 1 ? Number(e) : e,
        finalities: t.finalities,
        options: t.options
      }), typeof o == "function") {
        const u = b(t.copy[e]);
        return C(u), z(u), u.copy;
      }
      return t.copy[e];
    }
    return c;
  },
  set(t, e, n) {
    var r;
    if (t.type === 3 || t.type === 2)
      throw new Error("Map/Set draft does not support any property assignment.");
    let s;
    if (t.type === 1 && e !== "length" && !(Number.isInteger(s = Number(e)) && s >= 0 && (e === 0 || s === 0 || String(s) === String(e))))
      throw new Error("Only supports setting array indices and the 'length' property.");
    const i = ht(T(t), e);
    if (i != null && i.set)
      return i.set.call(t.proxy, n), !0;
    const o = Pe(T(t), e), a = b(o);
    return a && Q(a.original, n) ? (t.copy[e] = n, t.assignedMap = (r = t.assignedMap) !== null && r !== void 0 ? r : /* @__PURE__ */ new Map(), t.assignedMap.set(e, !1), !0) : (Q(n, o) && (n !== void 0 || _e(t.original, e)) || (C(t), z(t), _e(t.original, e) && Q(n, t.original[e]) ? t.assignedMap.delete(e) : t.assignedMap.set(e, !0), t.copy[e] = n, He(t, e, n, ae)), !0);
  },
  has(t, e) {
    return e in T(t);
  },
  ownKeys(t) {
    return Reflect.ownKeys(T(t));
  },
  getOwnPropertyDescriptor(t, e) {
    const n = T(t), r = Reflect.getOwnPropertyDescriptor(n, e);
    return r && {
      writable: !0,
      configurable: t.type !== 1 || e !== "length",
      enumerable: r.enumerable,
      value: n[e]
    };
  },
  getPrototypeOf(t) {
    return Reflect.getPrototypeOf(t.original);
  },
  setPrototypeOf() {
    throw new Error("Cannot call 'setPrototypeOf()' on drafts");
  },
  defineProperty() {
    throw new Error("Cannot call 'defineProperty()' on drafts");
  },
  deleteProperty(t, e) {
    var n;
    return t.type === 1 ? qt.set.call(this, t, e, void 0, t.proxy) : (Pe(t.original, e) !== void 0 || e in t.original ? (C(t), z(t), t.assignedMap.set(e, !1)) : (t.assignedMap = (n = t.assignedMap) !== null && n !== void 0 ? n : /* @__PURE__ */ new Map(), t.assignedMap.delete(e)), t.copy && delete t.copy[e], !0);
  }
};
function Ge(t) {
  const { original: e, parentDraft: n, key: r, finalities: s, options: i } = t, o = X(e), a = {
    type: o,
    finalized: !1,
    parent: n,
    original: e,
    copy: null,
    proxy: null,
    finalities: s,
    options: i,
    // Mapping of draft Set items to their corresponding draft values.
    setMap: o === 3 ? new Map(e.entries()) : void 0
  };
  (r || "key" in t) && (a.key = r);
  const { proxy: c, revoke: u } = Proxy.revocable(o === 1 ? Object.assign([], a) : a, qt);
  if (s.revoke.push(u), Dt.add(c), a.proxy = c, n) {
    const f = n;
    f.finalities.draft.push((l, d) => {
      var h, y;
      const _ = b(c);
      let w = f.type === 3 ? f.setMap : f.copy;
      const g = W(w, r), O = b(g);
      if (O) {
        let S = O.original;
        O.operated && (S = Be(g)), Te(O), Ie(O, ae, l, d), f.options.enableAutoFreeze && (f.options.updatedValues = (h = f.options.updatedValues) !== null && h !== void 0 ? h : /* @__PURE__ */ new WeakMap(), f.options.updatedValues.set(S, O.original)), ce(w, r, S);
      }
      (y = _.callbacks) === null || y === void 0 || y.forEach((S) => {
        S(l, d);
      });
    });
  } else {
    const f = b(c);
    f.finalities.draft.push((l, d) => {
      Te(f), Ie(f, ae, l, d);
    });
  }
  return c;
}
Je.createDraft = Ge;
function Rn(t, e, n, r, s) {
  var i;
  const o = b(t), a = (i = o == null ? void 0 : o.original) !== null && i !== void 0 ? i : t, c = !!e.length;
  if (o != null && o.operated)
    for (; o.finalities.draft.length > 0; )
      o.finalities.draft.pop()(n, r);
  const u = c ? e[0] : o ? o.operated ? o.copy : o.original : t;
  return o && je(o), s && te(u, u, o == null ? void 0 : o.options.updatedValues), [
    u,
    n && c ? [{ op: v.Replace, path: [], value: e[0] }] : n,
    r && c ? [{ op: v.Replace, path: [], value: a }] : r
  ];
}
function jn(t, e) {
  var n;
  const r = {
    draft: [],
    revoke: [],
    handledSet: /* @__PURE__ */ new WeakSet()
  };
  let s, i;
  e.enablePatches && (s = [], i = []);
  const a = ((n = e.mark) === null || n === void 0 ? void 0 : n.call(e, t, I)) === I.mutable || !N(t, e) ? t : Ge({
    original: t,
    parentDraft: null,
    finalities: r,
    options: e
  });
  return [
    a,
    (c = []) => {
      const [u, f, l] = Rn(a, c, s, i, e.enableAutoFreeze);
      return e.enablePatches ? [u, f, l] : u;
    }
  ];
}
function Ue(t) {
  const { rootDraft: e, value: n, useRawReturn: r = !1, isRoot: s = !0 } = t;
  Ke(n, (i, o, a) => {
    const c = b(o);
    if (c && e && c.finalities === e.finalities) {
      t.isContainDraft = !0;
      const u = c.original;
      if (a instanceof Set) {
        const f = Array.from(a);
        a.clear(), f.forEach((l) => a.add(i === l ? u : l));
      } else
        ce(a, i, u);
    } else typeof o == "object" && o !== null && (t.value = o, t.isRoot = !1, Ue(t));
  }), s && (t.isContainDraft || console.warn("The return value does not contain any draft, please use 'rawReturn()' to wrap the return value to improve performance."), r && console.warn("The return value contains drafts, please don't use 'rawReturn()' to wrap the return value."));
}
function Nt(t) {
  const e = b(t);
  if (!N(t, e == null ? void 0 : e.options))
    return t;
  const n = X(t);
  if (e && !e.operated)
    return e.original;
  let r;
  function s() {
    r = n === 2 ? new Map(t) : n === 3 ? Array.from(e.setMap.values()) : Ut(t, e == null ? void 0 : e.options);
  }
  if (e) {
    e.finalized = !0;
    try {
      s();
    } finally {
      e.finalized = !1;
    }
  } else
    r = t;
  return Ke(r, (i, o) => {
    if (e && Q(W(e.original, i), o))
      return;
    const a = Nt(o);
    a !== o && (r === t && s(), ce(r, i, a));
  }), n === 3 ? new Set(r) : r;
}
function yt(t) {
  if (!Y(t))
    throw new Error(`current() is only used for Draft, parameter: ${t}`);
  return Nt(t);
}
const Tn = (t) => function e(n, r, s) {
  var i, o, a;
  if (typeof n == "function" && typeof r != "function")
    return function(m, ...L) {
      return e(m, (x) => n.call(this, x, ...L), r);
    };
  const c = n, u = r;
  let f = s;
  if (typeof r != "function" && (f = r), f !== void 0 && Object.prototype.toString.call(f) !== "[object Object]")
    throw new Error(`Invalid options: ${f}, 'options' should be an object.`);
  f = Object.assign(Object.assign({}, t), f);
  const l = Y(c) ? yt(c) : c, d = Array.isArray(f.mark) ? (m, L) => {
    for (const x of f.mark) {
      if (typeof x != "function")
        throw new Error(`Invalid mark: ${x}, 'mark' should be a function.`);
      const B = x(m, L);
      if (B)
        return B;
    }
  } : f.mark, h = (i = f.enablePatches) !== null && i !== void 0 ? i : !1, y = (o = f.strict) !== null && o !== void 0 ? o : !1, w = {
    enableAutoFreeze: (a = f.enableAutoFreeze) !== null && a !== void 0 ? a : !1,
    mark: d,
    strict: y,
    enablePatches: h
  };
  if (!N(l, w) && typeof l == "object" && l !== null)
    throw new Error("Invalid base state: create() only supports plain objects, arrays, Set, Map or using mark() to mark the state as immutable.");
  const [g, O] = jn(l, w);
  if (typeof r != "function") {
    if (!N(l, w))
      throw new Error("Invalid base state: create() only supports plain objects, arrays, Set, Map or using mark() to mark the state as immutable.");
    return [g, O];
  }
  let S;
  try {
    S = u(g);
  } catch (m) {
    throw je(b(g)), m;
  }
  const k = (m) => {
    const L = b(g);
    if (!Y(m)) {
      if (m !== void 0 && !Q(m, g) && (L != null && L.operated))
        throw new Error("Either the value is returned as a new non-draft value, or only the draft is modified without returning any value.");
      const B = m == null ? void 0 : m[wn];
      if (B) {
        const Me = B[0];
        return w.strict && typeof m == "object" && m !== null && Ue({
          rootDraft: L,
          value: m,
          useRawReturn: !0
        }), O([Me]);
      }
      if (m !== void 0)
        return typeof m == "object" && m !== null && Ue({ rootDraft: L, value: m }), O([m]);
    }
    if (m === g || m === void 0)
      return O([]);
    const x = b(m);
    if (w === x.options) {
      if (x.operated)
        throw new Error("Cannot return a modified child draft.");
      return O([yt(m)]);
    }
    return O([m]);
  };
  return S instanceof Promise ? S.then(k, (m) => {
    throw je(b(g)), m;
  }) : k(S);
}, In = Tn();
Object.prototype.constructor.toString();
function Lt(t, e) {
  const n = Object.keys(t), r = Object.keys(e);
  return n.length === r.length && Object.keys(t).every((s) => e.hasOwnProperty(s));
}
function _t(t, e) {
  return Object.keys(t).length === Object.keys(e).length && Object.keys(t).every(
    (n) => e.hasOwnProperty(n) && t[n] === e[n]
  );
}
function be(t, e) {
  return typeof t != "object" || typeof e != "object" || t === null || e === null ? t === e : Lt(t, e) ? Object.keys(t).every(
    (n) => be(t[n], e[n])
  ) : !1;
}
function zt(t, e) {
  if (!ne(t) || !ne(e))
    return e;
  const n = {};
  for (const r of Object.keys(t))
    e[r] !== null && (n[r] = t[r]);
  for (const r of Object.keys(e)) {
    if (e[r] === null) continue;
    const s = ne(t[r]) && ne(e[r]);
    n[r] = s ? zt(t[r], e[r]) : e[r];
  }
  return n;
}
function Ft(t, e, n) {
  if (!ne(t))
    return t;
  const r = {};
  for (const [s, i] of Object.entries(t))
    r[s] = ne(i) ? Ft(i, e, n) : i === e ? n : i;
  return r;
}
function ne(t) {
  return typeof t == "object" && t !== null && !Array.isArray(t);
}
function kn(t) {
  return t.cardinality === "one";
}
function Ye(t) {
  return t["value-type"] === "ref";
}
function Vt(t) {
  return t["value-type"] === "blob";
}
function ge(t, e) {
  return t[e];
}
function Xe(t, e) {
  return e.reduce((n, r) => n && n.get(r), t);
}
function q(t, e) {
  if (e.length === 0) throw new Error("path must have at least one element");
  if (e.length === 1) {
    t.delete(e[0]);
    return;
  }
  const [n, ...r] = e;
  t.has(n) && q(t.get(n), r);
}
function R(t, e, n) {
  if (e.length === 0) throw new Error("path must have at least one element");
  if (e.length === 1) {
    t.set(e[0], n);
    return;
  }
  const [r, ...s] = e;
  let i = t.get(r);
  i || (i = /* @__PURE__ */ new Map(), t.set(r, i)), R(i, s, n);
}
function Qt(t, e) {
  const n = /* @__PURE__ */ new Map(), r = /* @__PURE__ */ new Map(), s = /* @__PURE__ */ new Map();
  for (const i of e) {
    const [o, a, c, u] = i, f = ge(t, a);
    if (!f) {
      console.warn("no such attr", o, t);
      continue;
    }
    Ye(f) && R(s, [c, a, o], i), R(n, [o, a, c], i), R(r, [a, o, c], i);
  }
  return { eav: n, aev: r, vae: s };
}
function Wt(t) {
  const e = /* @__PURE__ */ new Map(), n = /* @__PURE__ */ new Map(), r = /* @__PURE__ */ new Map(), s = /* @__PURE__ */ new Map();
  for (const i of Object.values(t)) {
    const o = i["forward-identity"], [a, c, u] = o, f = i["reverse-identity"];
    if (R(r, [c, u], i), Vt(i) && R(e, [c, u], i), i["primary?"] && R(n, [c], i), f) {
      const [l, d, h] = f;
      R(s, [d, h], i);
    }
  }
  return { blobAttrs: e, primaryKeys: n, forwardIdents: r, revIdents: s };
}
function xn(t) {
  return {
    __type: t.__type,
    attrs: t.attrs,
    triples: $(t.eav, 3),
    cardinalityInference: t.cardinalityInference,
    linkIndex: t.linkIndex
  };
}
function Un(t) {
  return $e(
    t.attrs,
    t.triples,
    t.cardinalityInference,
    t.linkIndex
  );
}
function Ze(t) {
  t.attrIndexes = Wt(t.attrs);
}
function $e(t, e, n, r) {
  const s = Qt(t, e);
  return s.attrs = t, s.attrIndexes = Wt(t), s.cardinalityInference = n, s.linkIndex = r, s.__type = "store", s;
}
function me(t, e) {
  var s, i;
  let n;
  if (Array.isArray(e[0])) {
    const [o, a] = e[0], c = t.aev.get(o);
    if (!c)
      return null;
    n = (s = $(c, 2).find((f) => f[2] === a)) == null ? void 0 : s[0];
  } else
    n = e[0];
  if (!n)
    return null;
  const r = e[2];
  if (Array.isArray(r) && r.length === 2 && t.aev.get(r[0])) {
    const [o, a] = r, c = t.aev.get(o);
    if (!c)
      return null;
    const f = (i = $(c, 2).find((_) => _[2] === a)) == null ? void 0 : i[0];
    if (!f)
      return null;
    const [l, d, h, ...y] = e;
    return [n, d, f, ...y];
  } else {
    const [o, ...a] = e;
    return [n, ...a];
  }
}
function $n(t, e) {
  const n = me(t, e);
  if (!n)
    return;
  const [r, s, i] = n, o = ge(t.attrs, s);
  o && (q(t.eav, [r, s, i]), q(t.aev, [s, r, i]), Ye(o) && q(t.vae, [i, s, r]));
}
let Dn = 0;
function Jt(t, e, n) {
  const [r, s, i] = n;
  let o;
  const a = Xe(t.ea, [r, s, i]);
  return a && (o = a[3]), o || Date.now() * 10 + Dn++;
}
function qn(t, e) {
  const n = me(t, e);
  if (!n)
    return;
  const [r, s, i] = n, o = ge(t.attrs, s);
  if (!o)
    return;
  const a = Xe(t.eav, [r, s, i]), c = (a == null ? void 0 : a[3]) ?? Jt(t, o, n), u = [r, s, i, c];
  kn(o) ? (R(t.eav, [r, s], /* @__PURE__ */ new Map([[i, u]])), R(t.aev, [s, r], /* @__PURE__ */ new Map([[i, u]]))) : (R(t.eav, [r, s, i], u), R(t.aev, [s, r, i], u)), Ye(o) && R(t.vae, [i, s, r], u);
}
function Nn(t, e) {
  var d;
  const n = me(t, e);
  if (!n)
    return;
  const [r, s, i] = n, o = ge(t.attrs, s);
  if (!o) return;
  if (!Vt(o))
    throw new Error("merge operation is not supported for links");
  const a = Xe(t.eav, [r, s]);
  if (!a) return;
  const c = (d = a.values().next()) == null ? void 0 : d.value;
  if (!c) return;
  const u = c[2], f = zt(u, i), l = [
    r,
    s,
    f,
    Jt(t, o, c)
  ];
  R(t.eav, [r, s], /* @__PURE__ */ new Map([[f, l]]));
}
function Ln(t, e) {
  var c, u;
  const [n, r] = e, s = me(t, [n]);
  if (!s)
    return;
  const [i] = s, o = t.eav.get(i);
  if (o) {
    for (const f of o.keys()) {
      const l = t.attrs[f];
      // Fall back to deleting everything if we've rehydrated tx-steps from
      // the store that didn't set `etype` in deleteEntity
      (!r || // If we don't know about the attr, let's just get rid of it
      !l || // Make sure it matches the etype
      ((c = l["forward-identity"]) == null ? void 0 : c[1]) === r) && (q(t.aev, [f, i]), q(t.eav, [i, f]));
    }
    o.size === 0 && q(t.eav, [i]);
  }
  const a = t.vae.get(i) && $(t.vae.get(i), 2);
  a && a.forEach((f) => {
    var _;
    const [l, d, h] = f, y = t.attrs[d];
    (!r || !y || ((_ = y["reverse-identity"]) == null ? void 0 : _[1]) === r) && (q(t.eav, [l, d, h]), q(t.aev, [d, l, h]), q(t.vae, [h, d, l]));
  }), ((u = t.vae.get(i)) == null ? void 0 : u.size) === 0 && q(t.vae, [i]);
}
function Bt(t, e) {
  const n = Qt(t.attrs, e);
  Object.keys(n).forEach((r) => {
    t[r] = n[r];
  });
}
function zn(t, [e]) {
  t.attrs[e.id] = e, Ze(t);
}
function Kt(t) {
  return $(t.eav, 3);
}
function Fn(t, [e]) {
  if (!t.attrs[e]) return;
  const n = Kt(t).filter(([r, s]) => s !== e);
  delete t.attrs[e], Ze(t), Bt(t, n);
}
function Vn(t, [e]) {
  const n = t.attrs[e.id];
  n && (t.attrs[e.id] = { ...n, ...e }, Ze(t), Bt(t, Kt(t)));
}
function Qn(t, e) {
  const [n, ...r] = e;
  switch (n) {
    case "add-triple":
      qn(t, r);
      break;
    case "deep-merge-triple":
      Nn(t, r);
      break;
    case "retract-triple":
      $n(t, r);
      break;
    case "delete-entity":
      Ln(t, r);
      break;
    case "add-attr":
      zn(t, r);
      break;
    case "delete-attr":
      Fn(t, r);
      break;
    case "update-attr":
      Vn(t, r);
      break;
    default:
      throw new Error(`unhandled transaction action: ${n}`);
  }
}
function $(t, e, n = []) {
  if (!t || e === 0)
    return n;
  if (e === 1) {
    for (const r of t.values())
      n.push(r);
    return n;
  }
  for (const r of t.values())
    $(r, e - 1, n);
  return n;
}
function fe(t, e, n) {
  var i, o;
  const r = [];
  if (n != null && n.hasOwnProperty("$not")) {
    for (const a of e.keys())
      n.$not !== a && r.push(e.get(a));
    return r;
  }
  if (n != null && n.hasOwnProperty("$isNull")) {
    const { attrId: a, isNull: c, reverse: u } = n.$isNull;
    if (u)
      for (const f of e.keys()) {
        const l = t.vae.get(f), d = !l || ((i = l.get(a)) == null ? void 0 : i.get(null)) || !l.get(a);
        (c ? d : !d) && r.push(e.get(f));
      }
    else {
      const f = t.aev.get(a);
      for (const l of e.keys()) {
        const d = !f || ((o = f.get(l)) == null ? void 0 : o.get(null)) || !f.get(l);
        (c ? d : !d) && r.push(e.get(l));
      }
    }
    return r;
  }
  if (n != null && n.$comparator)
    return $(e, 1).filter(n.$op);
  const s = n.in || n.$in || [n];
  for (const a of s) {
    const c = e.get(a);
    c && r.push(c);
  }
  return r;
}
function Wn(t, e, n) {
  let r = "";
  return t !== void 0 && (r += "e"), e !== void 0 && (r += "a"), n !== void 0 && (r += "v"), r;
}
function Jn(t, [e, n, r]) {
  var i, o;
  switch (Wn(e, n, r)) {
    case "e": {
      const a = t.eav.get(e);
      return $(a, 2);
    }
    case "ea": {
      const a = (i = t.eav.get(e)) == null ? void 0 : i.get(n);
      return $(a, 1);
    }
    case "eav": {
      const a = (o = t.eav.get(e)) == null ? void 0 : o.get(n);
      return a ? fe(t, a, r) : [];
    }
    case "ev": {
      const a = t.eav.get(e);
      if (!a)
        return [];
      const c = [];
      for (const u of a.values())
        c.push(...fe(t, u, r));
      return c;
    }
    case "a": {
      const a = t.aev.get(n);
      return $(a, 2);
    }
    case "av": {
      const a = t.aev.get(n);
      if (!a)
        return [];
      const c = [];
      for (const u of a.values())
        c.push(...fe(t, u, r));
      return c;
    }
    case "v": {
      const a = [];
      for (const c of t.eav.values())
        for (const u of c.values())
          a.push(...fe(t, u, r));
    }
    default:
      return $(t.eav, 3);
  }
}
function Bn(t, e, n) {
  var i;
  const r = t.attrIndexes.blobAttrs.get(e), s = {};
  for (const [o, a] of r.entries()) {
    const c = (i = t.eav.get(n)) == null ? void 0 : i.get(a.id), u = $(c, 1);
    for (const f of u)
      s[o] = f[2];
  }
  return s;
}
function De(t, e, n) {
  var r;
  return (r = t.attrIndexes.forwardIdents.get(e)) == null ? void 0 : r.get(n);
}
function Ht(t, e, n) {
  var r;
  return (r = t.attrIndexes.revIdents.get(e)) == null ? void 0 : r.get(n);
}
function Kn(t, e) {
  var r;
  const n = t.attrIndexes.primaryKeys.get(e);
  return n || ((r = t.attrIndexes.forwardIdents.get(e)) == null ? void 0 : r.get("id"));
}
function Hn(t, e) {
  return In(t, (n) => {
    e.forEach((r) => {
      Qn(n, r);
    });
  });
}
function Gn(t) {
  return typeof t == "string" && t.startsWith("?");
}
function Yn(t, e, n) {
  if (n.hasOwnProperty(t)) {
    const r = n[t];
    return Gt(r, e, n);
  }
  return { ...n, [t]: e };
}
function bt(t, e, n) {
  return t === e ? n : null;
}
function Xn(t) {
  switch (typeof t) {
    case "string":
      return t.startsWith("?") ? Yn : bt;
    default:
      return bt;
  }
}
const Zn = [
  "in",
  "$in",
  "$not",
  "$isNull",
  "$comparator"
  // covers all of $gt, $lt, etc.
];
function er(t) {
  for (const e of Zn)
    if (t.hasOwnProperty(e))
      return !0;
  return !1;
}
function Gt(t, e, n) {
  return n ? typeof t == "object" ? er(t) ? n : null : Xn(t)(t, e, n) : null;
}
function tr(t, e, n) {
  return t.reduce((r, s, i) => {
    const o = e[i];
    return Gt(s, o, r);
  }, n);
}
function nr(t, e, n) {
  return ir(t, e, n).map((r) => tr(e, r, n)).filter((r) => r);
}
function rr(t, e, n) {
  return e.or ? e.or.patterns.flatMap((r) => qe(t, r, n)) : e.and ? e.and.patterns.reduce((r, s) => qe(t, s, r), n) : n.flatMap((r) => nr(t, e, r));
}
function qe(t, e, n = [{}]) {
  return e.reduce((r, s) => rr(t, s, r), n);
}
function et(t, e) {
  return Array.isArray(e) ? e.map((n) => et(t, n)) : Gn(e) ? t[e] : e;
}
function sr(t, { find: e, where: n }) {
  return qe(t, n).map((s) => et(s, e));
}
function ir(t, e, n) {
  return Jn(t, et(n, e));
}
let le;
const or = new Uint8Array(16);
function ar() {
  if (!le && (le = typeof crypto < "u" && crypto.getRandomValues && crypto.getRandomValues.bind(crypto), !le))
    throw new Error("crypto.getRandomValues() not supported. See https://github.com/uuidjs/uuid#getrandomvalues-not-supported");
  return le(or);
}
const A = [];
for (let t = 0; t < 256; ++t)
  A.push((t + 256).toString(16).slice(1));
function cr(t, e = 0) {
  return A[t[e + 0]] + A[t[e + 1]] + A[t[e + 2]] + A[t[e + 3]] + "-" + A[t[e + 4]] + A[t[e + 5]] + "-" + A[t[e + 6]] + A[t[e + 7]] + "-" + A[t[e + 8]] + A[t[e + 9]] + "-" + A[t[e + 10]] + A[t[e + 11]] + A[t[e + 12]] + A[t[e + 13]] + A[t[e + 14]] + A[t[e + 15]];
}
const ur = typeof crypto < "u" && crypto.randomUUID && crypto.randomUUID.bind(crypto), wt = {
  randomUUID: ur
};
function fr(t, e, n) {
  if (wt.randomUUID && !e && !t)
    return wt.randomUUID();
  t = t || {};
  const r = t.random || (t.rng || ar)();
  return r[6] = r[6] & 15 | 64, r[8] = r[8] & 63 | 128, cr(r);
}
function gt(t) {
  const e = t.replace(/-/g, ""), n = [];
  for (let r = 0; r < e.length; r += 2)
    n.push(parseInt(e.substring(r, r + 2), 16));
  return n;
}
function lr(t, e) {
  for (let n = 0; n < t.length; n++) {
    if (t[n] < e[n]) return -1;
    if (t[n] > e[n]) return 1;
  }
  return 0;
}
function mt(t, e) {
  return lr(gt(t), gt(e));
}
function P() {
  return fr();
}
let dr = 0;
function oe(t) {
  return Oe(`_${t}`, dr++);
}
function Oe(t, e) {
  return `?${t}-${e}`;
}
class re extends Error {
  constructor(e) {
    super(e), this.name = "AttrNotFoundError";
  }
}
function Yt(t, e) {
  const n = Kn(t, e);
  if (!n)
    throw new re(`Could not find id attr for ${e}`);
  return n;
}
function Ot(t, e, n, r) {
  return [hr(t, e, n, r)];
}
function hr(t, e, n, r) {
  return [
    t(n, r),
    Yt(e, n).id,
    t(n, r),
    t("time", r)
  ];
}
function pr(t, e, n) {
  return t.map((r) => r === e ? n : r);
}
function Xt(t, e, n, r, s) {
  const i = De(e, n, s), o = Ht(e, n, s), a = i || o;
  if (!a)
    throw new re(`Could not find attr for ${[n, s]}`);
  if (a["value-type"] !== "ref")
    throw new Error(`Attr ${a.id} is not a ref`);
  const [c, u] = a["forward-identity"], [f, l] = a["reverse-identity"], d = r + 1, h = i ? [
    t(u, r),
    a.id,
    t(l, d),
    oe("time")
  ] : [
    t(u, d),
    a.id,
    t(l, r),
    oe("time")
  ];
  return [i ? l : u, d, h, a, !!i];
}
function yr(t, e) {
  if (typeof e != "object" || e.hasOwnProperty("$in") || e.hasOwnProperty("in"))
    return e;
  const n = t["checked-data-type"] === "date";
  return e.hasOwnProperty("$gt") ? {
    $comparator: !0,
    $op: n ? function(s) {
      return new Date(s[2]) > new Date(e.$gt);
    } : function(s) {
      return s[2] > e.$gt;
    }
  } : e.hasOwnProperty("$gte") ? {
    $comparator: !0,
    $op: n ? function(s) {
      return new Date(s[2]) >= new Date(e.$gte);
    } : function(s) {
      return s[2] >= e.$gte;
    }
  } : e.hasOwnProperty("$lt") ? {
    $comparator: !0,
    $op: n ? function(s) {
      return new Date(s[2]) < new Date(e.$lt);
    } : function(s) {
      return s[2] < e.$lt;
    }
  } : e.hasOwnProperty("$lte") ? {
    $comparator: !0,
    $op: n ? function(s) {
      return new Date(s[2]) <= new Date(e.$lte);
    } : function(s) {
      return s[2] <= e.$lte;
    }
  } : e;
}
function _r(t, e, n, r, s, i) {
  const o = De(e, n, s), a = Ht(e, n, s), c = o || a;
  if (!c)
    throw new re(
      `No attr for etype = ${n} label = ${s}`
    );
  if (i != null && i.hasOwnProperty("$isNull")) {
    const u = De(e, n, "id");
    if (!u)
      throw new re(
        `No attr for etype = ${n} label = id`
      );
    return [
      t(n, r),
      u.id,
      { $isNull: { attrId: c.id, isNull: i.$isNull, reverse: !o } },
      oe("time")
    ];
  }
  return o ? [
    t(n, r),
    c.id,
    yr(c, i),
    oe("time")
  ] : [i, c.id, t(n, r), oe("time")];
}
function br(t, e, n, r, s) {
  const [i, o, a] = s.reduce(
    (c, u) => {
      const [f, l, d] = c, [h, y, _] = Xt(
        t,
        e,
        f,
        l,
        u
      );
      return [h, y, [...d, _]];
    },
    [n, r, []]
  );
  return [i, o, a];
}
function Ne(t, e, n, r, s, i) {
  const o = s.slice(0, s.length - 1), a = s[s.length - 1], [c, u, f] = br(
    t,
    e,
    n,
    r,
    o
  ), l = _r(
    t,
    e,
    c,
    u,
    a,
    i
  );
  return f.concat([l]);
}
function wr(t, e) {
  return e ? [e].concat(t) : t;
}
function gr([t, e]) {
  return t === "or" && Array.isArray(e);
}
function mr([t, e]) {
  return t === "and" && Array.isArray(e);
}
function Or(t, e, n) {
  return (r, s) => r == e ? t(r, s) : `${t(r, s)}-${n}`;
}
function St(t, e, n, r, s, i) {
  const o = i.map((c, u) => {
    const f = Or(t, r, u);
    return Zt(f, n, r, s, c);
  }), a = t(r, s);
  return { [e]: { patterns: o, joinSym: a } };
}
function Sr(t) {
  const e = [];
  for (let n = 1; n <= t.length; n++)
    e.push(t.slice(0, n));
  return e;
}
function vt(t, e, n, r, s) {
  return Sr(s).map(
    (i) => Ne(t, e, n, r, i, { $isNull: !0 })
  );
}
function Zt(t, e, n, r, s) {
  return Object.entries(s).flatMap(([i, o]) => {
    if (gr([i, o]))
      return St(t, "or", e, n, r, o);
    if (mr([i, o]))
      return St(t, "and", e, n, r, o);
    const a = i.split(".");
    if (o != null && o.hasOwnProperty("$not")) {
      const c = Ne(t, e, n, r, a, o), u = vt(
        t,
        e,
        n,
        r,
        a
      );
      return [
        {
          or: {
            patterns: [c, ...u],
            joinSym: t(n, r)
          }
        }
      ];
    }
    return o != null && o.hasOwnProperty("$isNull") && o.$isNull === !0 && a.length > 1 ? [
      {
        or: {
          patterns: vt(
            t,
            e,
            n,
            r,
            a
          ),
          joinSym: t(n, r)
        }
      }
    ] : Ne(t, e, n, r, a, o);
  });
}
function vr(t, e, n, r) {
  const s = Oe;
  return r ? Zt(s, t, e, n, r).concat(Ot(s, t, e, n)) : Ot(s, t, e, n);
}
function Mr(t, e, n) {
  return [t(e, n), t("time", n)];
}
function Ar(t, e, n, r, s, i) {
  const [o, a, c, u, f] = Xt(
    t,
    e,
    n,
    r,
    s
  ), l = pr(c, t(n, r), i);
  return [o, a, l, u, f];
}
function Pr(t, e, { etype: n, level: r, form: s }, i) {
  const o = Object.keys(s).filter((a) => a !== "$");
  return o.length ? Object.entries(i).map(function([c, u]) {
    return o.map(function(d) {
      var y, _, w;
      const h = !!(e.cardinalityInference && ((w = (_ = (y = e.linkIndex) == null ? void 0 : y[n]) == null ? void 0 : _[d]) != null && w.isSingular));
      try {
        const [g, O, S] = Ar(
          t,
          e,
          n,
          r,
          d,
          c
        ), k = en(e, {
          etype: g,
          level: O,
          form: s[d],
          join: S
        }), m = h ? k[0] : k;
        return { [d]: m };
      } catch (g) {
        if (g instanceof re)
          return { [d]: h ? void 0 : [] };
        throw g;
      }
    }).reduce(function(d, h) {
      return { ...d, ...h };
    }, u);
  }) : Object.values(i);
}
function Mt(t, e) {
  switch (t) {
    case "asc":
      switch (e) {
        case "number":
          return (n, r) => n < r;
        case "uuid":
          return (n, r) => mt(n, r) === -1;
      }
    case "desc":
      switch (e) {
        case "number":
          return (n, r) => n > r;
        case "uuid":
          return (n, r) => mt(n, r) === 1;
      }
  }
}
function Er(t, e, [n, r, s, i]) {
  return Mt(e, "number")(i, t[3]) || i === t[3] && Mt(e, "uuid")(n, t[0]);
}
function Cr(t, e, n, r, s) {
  const i = Yt(t, e).id, o = sr(t, s).sort(function([f, l], [d, h]) {
    return n === "desc" ? h - l : l - h;
  });
  let a = {};
  const c = r == null ? void 0 : r["start-cursor"];
  for (const [u, f] of o) {
    if (c && i === c[1] && Er(c, n, [u, i, u, f]))
      continue;
    const l = Bn(t, e, u);
    l && (a[u] = l);
  }
  return a;
}
function Rr(t) {
  var n;
  const e = (n = t.$) == null ? void 0 : n.order;
  return e && e[Object.keys(e)[0]] || "asc";
}
function jr(t, { etype: e, level: n, form: r, join: s, pageInfo: i }) {
  var h, y, _, w, g, O, S;
  const o = ((h = r.$) == null ? void 0 : h.limit) || ((y = r.$) == null ? void 0 : y.first) || ((_ = r.$) == null ? void 0 : _.last), a = (w = r.$) == null ? void 0 : w.offset, c = (g = r.$) == null ? void 0 : g.before, u = (O = r.$) == null ? void 0 : O.after;
  if ((a || c || u) && (!i || !i["start-cursor"]))
    return [];
  const f = wr(vr(t, e, n, (S = r.$) == null ? void 0 : S.where), s), l = Mr(Oe, e, n), d = Cr(
    t,
    e,
    Rr(r),
    i,
    { where: f, find: l }
  );
  if (o != null) {
    const k = Object.entries(d);
    return k.length <= o ? d : Object.fromEntries(k.slice(0, o));
  }
  return d;
}
function Tr(t, e) {
  try {
    return jr(t, e);
  } catch (n) {
    if (n instanceof re)
      return {};
    throw n;
  }
}
function en(t, e) {
  const n = Tr(t, e);
  return Pr(Oe, t, e, n);
}
function Ir(t) {
  const e = {};
  for (const [n, r] of Object.entries(t))
    e[n] = {
      startCursor: r["start-cursor"],
      endCursor: r["end-cursor"],
      hasNextPage: r["has-next-page?"],
      hasPreviousPage: r["has-previous-page?"]
    };
  return e;
}
function kr({ store: t, pageInfo: e, aggregate: n }, r) {
  const i = { data: Object.keys(r).reduce(function(a, c) {
    return n != null && n[c] || (a[c] = en(t, {
      etype: c,
      form: r[c],
      level: 0,
      pageInfo: e == null ? void 0 : e[c]
    })), a;
  }, {}) };
  return e && (i.pageInfo = Ir(e)), n && (i.aggregate = n), i;
}
function Le(t, e, n) {
  return new Proxy({}, {
    get: (r, s) => s === "__ops" ? n : (i) => Le(t, e, [
      ...n,
      [s, t, e, i]
    ])
  });
}
function Vs(t, e) {
  return `lookup__${t}__${JSON.stringify(e)}`;
}
function ze(t) {
  return t.startsWith("lookup__");
}
function tn(t) {
  const [e, n, ...r] = t.split("__");
  return [n, JSON.parse(r.join("__"))];
}
function xr(t) {
  return new Proxy(
    {},
    {
      get(e, n) {
        return ze(n) ? Le(t, tn(n), []) : Le(t, n, []);
      }
    }
  );
}
function tt() {
  return new Proxy(
    {},
    {
      get(t, e) {
        return xr(e);
      }
    }
  );
}
const Qs = tt();
function Ur(t) {
  return t.__ops;
}
function $r(t, e) {
  const { attrIdMap: n, refSwapAttrIds: r } = t, s = [];
  for (const o of e) {
    const a = n[o];
    if (a)
      s.push(a);
    else if (Array.isArray(o) && o.length == 2 && n[o[0]]) {
      const [c, u] = o;
      s.push([n[c], u]);
    } else
      s.push(o);
  }
  const [i] = e;
  if ((i === "add-triple" || i === "retract-triple") && r.has(e[2])) {
    const o = s[1];
    s[1] = s[3], s[3] = o;
  }
  return s;
}
function j(t, e, n) {
  return Object.values(t).find((r) => {
    const [s, i, o] = r["forward-identity"];
    return i === e && o === n;
  });
}
function G(t, e, n) {
  return Object.values(t).find((r) => {
    const s = r["reverse-identity"];
    if (!s) return !1;
    const [i, o, a] = s;
    return o === e && a === n;
  });
}
function Dr(t) {
  if (Array.isArray(t))
    return t;
  const e = Object.entries(t);
  if (e.length !== 1)
    throw new Error(
      "lookup must be an object with a single unique attr and value."
    );
  return e[0];
}
function Fe(t, e, n) {
  return n.indexOf(".") !== -1 && // attr names can have `.` in them, so use the attr we find with a `.`
  // before assuming it's a ref lookup.
  !j(t, e, n);
}
function Ve(t) {
  const [e, n, ...r] = t.split(".");
  if (r.length > 0 || n !== "id")
    throw new Error(`${t} is not a valid lookup attribute.`);
  return e;
}
function qr(t, e, n) {
  if (!Fe(t, e, n))
    return j(t, e, n);
  const r = Ve(n), s = j(t, e, r) || G(t, e, r);
  if (s && s["value-type"] !== "ref")
    throw new Error(`${n} does not reference a valid link attribute.`);
  return s;
}
function Qe(t) {
  return typeof t == "string" && !ze(t) ? null : typeof t == "string" && ze(t) ? tn(t) : Dr(t);
}
function D(t, e, n) {
  const r = Qe(n);
  if (r === null)
    return n;
  const [s, i] = r, o = qr(t, e, s);
  if (!o || !o["unique?"])
    throw new Error(`${s} is not a unique attribute.`);
  return [o.id, i];
}
function Nr(t, [e, n, r]) {
  return Object.entries(r).flatMap(([i, o]) => {
    const a = Array.isArray(o) ? o : [o], c = j(t, e, i), u = G(t, e, i);
    return a.map((f) => c ? [
      "add-triple",
      D(t, e, n),
      c.id,
      D(t, c["reverse-identity"][1], f)
    ] : [
      "add-triple",
      D(t, u["forward-identity"][1], f),
      u.id,
      D(t, e, n)
    ]);
  });
}
function Lr(t, [e, n, r]) {
  return Object.entries(r).flatMap(([i, o]) => {
    const a = Array.isArray(o) ? o : [o], c = j(t, e, i), u = G(t, e, i);
    return a.map((f) => c ? [
      "retract-triple",
      D(t, e, n),
      c.id,
      D(t, c["reverse-identity"][1], f)
    ] : [
      "retract-triple",
      D(t, u["forward-identity"][1], f),
      u.id,
      D(t, e, n)
    ]);
  });
}
function zr(t, [e, n, r]) {
  const s = D(t, e, n);
  return [["id", D(t, e, n)]].concat(Object.entries(r)).map(([o, a]) => {
    const c = j(t, e, o);
    return ["add-triple", s, c.id, a];
  });
}
function Fr(t, [e, n]) {
  return [["delete-entity", D(t, e, n), e]];
}
function Vr(t, [e, n, r]) {
  const s = D(t, e, n), i = Object.entries(r).map(([a, c]) => {
    const u = j(t, e, a), f = Ft(c, void 0, null);
    return ["deep-merge-triple", s, u.id, f];
  });
  return [[
    "add-triple",
    s,
    j(t, e, "id").id,
    s
  ]].concat(i);
}
function Qr(t) {
  const [e, n, r, s] = t;
  if (!s)
    return t;
  const i = { ...s };
  return delete i.id, [e, n, r, i];
}
function Wr(t, e) {
  const [n, ...r] = Qr(e);
  switch (n) {
    case "merge":
      return Vr(t, r);
    case "update":
      return zr(t, r);
    case "link":
      return Nr(t, r);
    case "unlink":
      return Lr(t, r);
    case "delete":
      return Fr(t, r);
    default:
      throw new Error(`unsupported action ${n}`);
  }
}
function Jr(t) {
  switch (t) {
    case "string":
    case "date":
    case "boolean":
    case "number":
      return t;
    default:
      return;
  }
}
function Br(t, e, n) {
  var a, c;
  const r = (c = (a = t.entities[e]) == null ? void 0 : a.attrs) == null ? void 0 : c[n];
  if (n === "id") return null;
  if (!r)
    throw new Error(`${e}.${n} does not exist in your schema`);
  const { unique: s, indexed: i } = r == null ? void 0 : r.config, o = Jr(r == null ? void 0 : r.valueType);
  return {
    "index?": i,
    "unique?": s,
    "checked-data-type": o
  };
}
function Ee(t, e, n, r) {
  const s = t ? Br(t, e, n) : null, i = P(), a = [P(), e, n];
  return {
    id: i,
    "forward-identity": a,
    "value-type": "blob",
    cardinality: "one",
    "unique?": !1,
    "index?": !1,
    isUnsynced: !0,
    ...s || {},
    ...r || {}
  };
}
function Kr(t, e, n) {
  return Object.values(t.links).find((s) => s.forward.on === e && s.forward.label === n || s.reverse.on === e && s.reverse.label === n);
}
function Hr(t, e, n) {
  const r = Kr(t, e, n);
  if (!r)
    throw new Error(`Couldn't find the link ${e}.${n} in your schema`);
  const { forward: s, reverse: i } = r;
  return {
    "forward-identity": [P(), s.on, s.label],
    "reverse-identity": [P(), i.on, i.label],
    cardinality: s.has === "one" ? "one" : "many",
    "unique?": i.has === "one"
  };
}
function At(t, e, n, r) {
  const s = t ? Hr(t, e, n) : null, i = P(), o = [P(), e, n], a = [P(), n, e];
  return {
    id: i,
    "forward-identity": o,
    "reverse-identity": a,
    "value-type": "ref",
    cardinality: "many",
    "unique?": !1,
    "index?": !1,
    isUnsynced: !0,
    ...s || {},
    ...r || {}
  };
}
const Gr = /* @__PURE__ */ new Set(["update", "merge", "link", "unlink"]), Yr = /* @__PURE__ */ new Set(["link", "unlink"]), Xr = /* @__PURE__ */ new Set(["update", "merge"]), Zr = /* @__PURE__ */ new Set([
  "link",
  "unlink",
  "update",
  "merge",
  "delete"
]), We = { "unique?": !0, "index?": !0 }, es = { ...We, cardinality: "one" };
function ts(t) {
  const e = [], [n, r, s, i] = t;
  if (!Zr.has(n))
    return e;
  const o = Qe(s);
  if (o && e.push({ etype: r, lookupPair: o }), n === "link")
    for (const [a, c] of Object.entries(i)) {
      const u = Array.isArray(c) ? c : [c];
      for (const f of u) {
        const l = Qe(f);
        l && e.push({
          etype: r,
          lookupPair: l,
          linkLabel: a
        });
      }
    }
  return e;
}
function ns({ attrs: t, schema: e }, n) {
  var u, f;
  const [r, s, i] = [/* @__PURE__ */ new Set(), { ...t }, []];
  function o(l) {
    s[l.id] = l, i.push(["add-attr", l]), r.add(l.id);
  }
  function a(l) {
    l != null && l.isUnsynced && !r.has(l.id) && (i.push(["add-attr", l]), r.add(l.id));
  }
  function c(l, d) {
    const h = j(s, l, d), y = G(s, l, d);
    a(h), a(y), !h && !y && o(At(e, l, d, es));
  }
  for (const l of n)
    for (const { etype: d, lookupPair: h, linkLabel: y } of ts(l)) {
      const _ = h[0];
      if (y) {
        c(d, y);
        const w = j(s, d, y), g = G(s, d, y);
        a(w), a(g);
        const O = ((u = w == null ? void 0 : w["reverse-identity"]) == null ? void 0 : u[1]) || ((f = g == null ? void 0 : g["forward-identity"]) == null ? void 0 : f[1]) || y;
        if (Fe(s, O, _))
          c(O, Ve(_));
        else {
          const S = j(s, O, _);
          S || o(
            Ee(e, O, _, We)
          ), a(S);
        }
      } else if (Fe(s, d, _))
        c(d, Ve(_));
      else {
        const w = j(s, d, _);
        w || o(Ee(e, d, _, We)), a(w);
      }
    }
  for (const l of n) {
    const [d, h, y, _] = l;
    if (Gr.has(d)) {
      const w = Object.keys(_);
      w.push("id");
      for (const g of w) {
        const O = j(s, h, g);
        if (a(O), Xr.has(d) && (O || o(
          Ee(
            e,
            h,
            g,
            g === "id" ? { "unique?": !0 } : null
          )
        )), Yr.has(d)) {
          const S = G(s, h, g);
          !O && !S && o(At(e, h, g)), a(S);
        }
      }
    }
  }
  return [s, i];
}
function rs(t, e) {
  const r = (Array.isArray(e) ? e : [e]).flatMap((a) => Ur(a)), [s, i] = ns(t, r), o = r.flatMap((a) => Wr(s, a));
  return [...i, ...o];
}
class nt {
  constructor(e) {
    this.dbName = e, this._storeName = "kv", this._dbPromise = this._init();
  }
  _init() {
    return new Promise((e, n) => {
      const r = indexedDB.open(this.dbName, 1);
      r.onerror = (s) => {
        n(s);
      }, r.onsuccess = (s) => {
        e(s.target.result);
      }, r.onupgradeneeded = (s) => {
        s.target.result.createObjectStore(this._storeName);
      };
    });
  }
  async getItem(e) {
    const n = await this._dbPromise;
    return new Promise((r, s) => {
      const a = n.transaction([this._storeName], "readonly").objectStore(this._storeName).get(e);
      a.onerror = (c) => {
        s(c);
      }, a.onsuccess = (c) => {
        a.result ? r(a.result) : r(null);
      };
    });
  }
  async setItem(e, n) {
    const r = await this._dbPromise;
    return new Promise((s, i) => {
      const c = r.transaction([this._storeName], "readwrite").objectStore(this._storeName).put(n, e);
      c.onerror = (u) => {
        i(u);
      }, c.onsuccess = (u) => {
        s();
      };
    });
  }
}
class rt {
  static async getIsOnline() {
    return navigator.onLine;
  }
  static listen(e) {
    const n = () => {
      e(!0);
    }, r = () => {
      e(!1);
    };
    return addEventListener("online", n), addEventListener("offline", r), () => {
      removeEventListener("online", n), removeEventListener("offline", r);
    };
  }
}
async function J(t, e) {
  const n = await fetch(t, e), r = await n.json();
  return n.status === 200 ? Promise.resolve(r) : Promise.reject({ status: n.status, body: r });
}
function ss({ apiURI: t, appId: e, email: n }) {
  return J(`${t}/runtime/auth/send_magic_code`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ "app-id": e, email: n })
  });
}
async function is({ apiURI: t, appId: e, email: n, code: r }) {
  return await J(`${t}/runtime/auth/verify_magic_code`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ "app-id": e, email: n, code: r })
  });
}
async function os({ apiURI: t, appId: e, refreshToken: n }) {
  return await J(`${t}/runtime/auth/verify_refresh_token`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({
      "app-id": e,
      "refresh-token": n
    })
  });
}
async function Pt({
  apiURI: t,
  appId: e,
  code: n,
  codeVerifier: r
}) {
  return await J(`${t}/runtime/oauth/token`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({
      app_id: e,
      code: n,
      code_verifier: r
    })
  });
}
async function as({
  apiURI: t,
  appId: e,
  nonce: n,
  idToken: r,
  clientName: s,
  refreshToken: i
}) {
  return await J(`${t}/runtime/oauth/id_token`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({
      app_id: e,
      nonce: n,
      id_token: r,
      client_name: s,
      refresh_token: i
    })
  });
}
async function cs({ apiURI: t, appId: e, refreshToken: n }) {
  return await J(`${t}/runtime/signout`, {
    method: "POST",
    headers: {
      "content-type": "application/json"
    },
    body: JSON.stringify({
      app_id: e,
      refresh_token: n
    })
  });
}
async function us({
  apiURI: t,
  appId: e,
  fileName: n,
  refreshToken: r,
  metadata: s = {}
}) {
  const { data: i } = await J(`${t}/storage/signed-upload-url`, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      authorization: `Bearer ${r}`
    },
    body: JSON.stringify({
      app_id: e,
      filename: n
    })
  });
  return i;
}
async function fs(t, e) {
  return (await fetch(t, {
    method: "PUT",
    body: e,
    headers: {
      "Content-Type": e.type
    }
  })).ok;
}
async function ls({
  apiURI: t,
  appId: e,
  path: n,
  refreshToken: r
}) {
  const { data: s } = await J(
    `${t}/storage/signed-download-url?app_id=${e}&filename=${encodeURIComponent(n)}`,
    {
      method: "GET",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${r}`
      }
    }
  );
  return s;
}
async function ds({
  apiURI: t,
  appId: e,
  path: n,
  refreshToken: r
}) {
  const { data: s } = await J(
    `${t}/storage/files?app_id=${e}&filename=${encodeURIComponent(n)}`,
    {
      method: "DELETE",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${r}`
      }
    }
  );
  return s;
}
function Et(t, e) {
  if (!e) return t;
  const n = {};
  return e.forEach((r) => {
    n[r] = t[r];
  }), n;
}
function hs(t, e, n) {
  const r = {
    peers: {}
  };
  if (e && "user" in e ? e.user : !0) {
    const i = Et(t.user ?? {}, e == null ? void 0 : e.keys);
    r.user = { ...i, peerId: n };
  }
  for (const i of Object.keys(t.peers ?? {})) {
    const o = (e == null ? void 0 : e.peers) === void 0, a = Array.isArray(e == null ? void 0 : e.peers) && (e == null ? void 0 : e.peers.includes(i));
    if (o || a) {
      const c = Et(t.peers[i], e == null ? void 0 : e.keys);
      r.peers[i] = { ...c, peerId: i };
    }
  }
  return r;
}
function ps(t, e) {
  if (t.isLoading !== e.isLoading || t.error !== e.error || (t.user || e.user) && (!t.user || !e.user || !_t(t.user, e.user)) || !Lt(t.peers, e.peers)) return !0;
  for (const r of Object.keys(t.peers))
    if (!_t(t.peers[r], e.peers[r])) return !0;
  return !1;
}
class Ct {
  constructor() {
    p(this, "promise");
    p(this, "_resolve");
    p(this, "_reject");
    this.promise = new Promise((e, n) => {
      this._resolve = e, this._reject = n;
    });
  }
  resolve(e) {
    this._resolve(e);
  }
  reject(e) {
    this._reject(e);
  }
}
class Rt {
  constructor(e, n, r, s, i = (c) => JSON.stringify(c), o = (c) => JSON.parse(c), a = 100) {
    this._persister = e, this._key = n, this._onMerge = s, this._loadedCbs = [], this._isLoading = !0, this.currentValue = r, this.toJSON = i, this.fromJSON = o, this._saveThrottleMs = a, this._pendingSaveCbs = [], this._version = 0, this._load();
  }
  async _load() {
    const e = this.fromJSON(await this._persister.getItem(this._key));
    this._isLoading = !1, this._onMerge(e, this.currentValue);
    for (const n of this._loadedCbs)
      n();
  }
  async waitForLoaded() {
    if (!this._isLoading)
      return;
    await new Promise((n) => {
      this._loadedCbs.push(n);
    });
  }
  isLoading() {
    return this._isLoading;
  }
  version() {
    return this._version;
  }
  async waitForSync() {
    if (!this._nextSave)
      return;
    await new Promise((n) => {
      this._pendingSaveCbs.push(n);
    });
  }
  _writeToStorage() {
    this._persister.setItem(this._key, this.toJSON(this.currentValue));
    for (const e of this._pendingSaveCbs)
      e();
    this._pendingSaveCbs.length = 0;
  }
  async flush() {
    this._nextSave && (clearTimeout(this._nextSave), this._writeToStorage());
  }
  _enqueuePersist(e) {
    if (this._nextSave) {
      e && this._pendingSaveCbs.push(e);
      return;
    }
    this._nextSave = setTimeout(() => {
      this._nextSave = null, this._writeToStorage();
    }, this._saveThrottleMs);
  }
  set(e, n) {
    this._version++, this.currentValue = e(this.currentValue), this._isLoading ? this._loadedCbs.push(() => this._enqueuePersist(n)) : this._enqueuePersist(n);
  }
}
function nn(t, e = []) {
  t.forEach((n) => {
    const { data: r } = n, { "datalog-result": s } = r, { "join-rows": i } = s;
    for (const o of i)
      for (const a of o)
        e.push(a);
    nn(n["child-nodes"], e);
  });
}
function jt(t) {
  const e = [];
  return nn(t, e), e;
}
function ys(t) {
  return Object.values(t.links).reduce((e, n) => {
    var r, s;
    return e[r = n.forward.on] ?? (e[r] = {}), e[n.forward.on][n.forward.label] = {
      isForward: !0,
      isSingular: n.forward.has === "one",
      link: n
    }, e[s = n.reverse.on] ?? (e[s] = {}), e[n.reverse.on][n.reverse.label] = {
      isForward: !1,
      isSingular: n.reverse.has === "one",
      link: n
    }, e;
  }, {});
}
const _s = "v0.15.9", K = {
  CONNECTING: "connecting",
  OPENED: "opened",
  AUTHENTICATED: "authenticated",
  CLOSED: "closed",
  ERRORED: "errored"
}, bs = 3e4, ws = 0, Tt = 1, gs = {
  apiURI: "https://api.instantdb.com",
  websocketURI: "wss://api.instantdb.com/runtime/session"
}, Ce = "_instant_oauth_redirect", Re = "currentUser";
let ms = 0;
function Os(t) {
  const e = new WebSocket(t);
  return e._id = ms++, e;
}
function Ss() {
  return typeof window < "u" || typeof chrome < "u";
}
const It = {
  "set-presence": !0,
  "set-presence-ok": !0,
  "refresh-presence": !0
};
function vs(t) {
  var n;
  const e = JSON.parse(t);
  for (const r in e) {
    const s = e[r];
    (n = s == null ? void 0 : s.result) != null && n.store && (s.result.store = Un(s.result.store));
  }
  return e;
}
function Ms(t) {
  var n;
  const e = {};
  for (const r in t) {
    const s = t[r], i = { ...s };
    (n = s.result) != null && n.store && (i.result = {
      ...s.result,
      store: xn(s.result.store)
    }), e[r] = i;
  }
  return JSON.stringify(e);
}
class rn {
  constructor(e, n = nt, r = rt, s) {
    p(this, "attrs");
    p(this, "_isOnline", !0);
    p(this, "_isShutdown", !1);
    p(this, "status", K.CONNECTING);
    /** @type {PersistedObject} */
    p(this, "querySubs");
    /** @type {PersistedObject} */
    p(this, "pendingMutations");
    /** @type {Record<string, Array<{ q: any, cb: (data: any) => any }>>} */
    p(this, "queryCbs", {});
    /** @type {Record<string, Array<{ q: any, eventId: string, dfd: Deferred }>>} */
    p(this, "queryOnceDfds", {});
    p(this, "authCbs", []);
    p(this, "attrsCbs", []);
    p(this, "mutationErrorCbs", []);
    p(this, "config");
    p(this, "_persister");
    p(this, "mutationDeferredStore", /* @__PURE__ */ new Map());
    p(this, "_reconnectTimeoutId", null);
    p(this, "_reconnectTimeoutMs", 0);
    p(this, "_ws");
    p(this, "_localIdPromises", {});
    p(this, "_errorMessage", null);
    /** @type {Promise<null | {error: {message: string}}>}**/
    p(this, "_oauthCallbackResponse", null);
    /** @type {null | import('./utils/linkIndex').LinkIndex}} */
    p(this, "_linkIndex", null);
    /** @type BroadcastChannel | undefined */
    p(this, "_broadcastChannel");
    /** @type {Record<string, {isConnected: boolean; error: any}>} */
    p(this, "_rooms", {});
    /** @type {Record<string, boolean>} */
    p(this, "_roomsPendingLeave", {});
    p(this, "_presence", {});
    p(this, "_broadcastQueue", []);
    p(this, "_broadcastSubs", {});
    p(this, "_currentUserCached", { isLoading: !0, error: void 0, user: void 0 });
    p(this, "_beforeUnloadCbs", []);
    p(this, "_dataForQueryCache", {});
    /**
     *  merge querySubs from storage and in memory. Has the following side
     *  effects:
     *  - We notify all queryCbs because results may been added during merge
     */
    p(this, "_onMergeQuerySubs", (e, n) => {
      const r = e || {}, s = { ...n };
      Object.entries(n).forEach(([o, a]) => {
        var f;
        const c = (f = r == null ? void 0 : r[o]) == null ? void 0 : f.result, u = a.result;
        c && !u && (s[o].result = c);
      }), Object.keys(r).filter((o) => !n[o]).slice(0, 10).forEach((o) => {
        s[o] = r[o];
      }), this.querySubs.set((o) => s), this.loadedNotifyAll();
    });
    /**
     * merge pendingMutations from storage and in memory. Has a side effect of
     * sending mutations that were stored but not acked
     */
    p(this, "_onMergePendingMutations", (e, n) => {
      const r = new Map([...e.entries(), ...n.entries()]);
      this.pendingMutations.set((i) => r), this.loadedNotifyAll(), this._rewriteMutations(
        this.attrs,
        e
      ).forEach((i, o) => {
        !n.has(o) && !i["tx-id"] && this._sendMutation(o, i);
      });
    });
    // ---------------------------
    // Queries
    p(this, "getPreviousResult", (e) => {
      const n = U(e);
      return this.dataForQuery(n);
    });
    /** Re-run instaql and call all callbacks with new data */
    p(this, "notifyOne", (e) => {
      var i;
      const n = this.queryCbs[e] ?? [], r = (i = this._dataForQueryCache[e]) == null ? void 0 : i.data, s = this.dataForQuery(e);
      s && (be(s, r) || n.forEach((o) => o.cb(s)));
    });
    p(this, "notifyOneQueryOnce", (e) => {
      const n = this.queryOnceDfds[e] ?? [], r = this.dataForQuery(e);
      n.forEach((s) => {
        this._completeQueryOnce(s.q, e, s.dfd), s.dfd.resolve(r);
      });
    });
    p(this, "notifyQueryError", (e, n) => {
      (this.queryCbs[e] || []).forEach((s) => s.cb({ error: n }));
    });
    /** Applies transactions locally and sends transact message to server */
    p(this, "pushTx", (e) => {
      try {
        const n = rs(
          { attrs: this.optimisticAttrs(), schema: this.config.schema },
          e
        );
        return this.pushOps(n);
      } catch (n) {
        return this.pushOps([], n);
      }
    });
    /**
     * @param {*} txSteps
     * @param {*} [error]
     * @returns
     */
    p(this, "pushOps", (e, n) => {
      const r = P(), s = {
        op: "transact",
        "tx-steps": e,
        error: n
      };
      this.pendingMutations.set((o) => (o.set(r, s), o));
      const i = new Ct();
      return this.mutationDeferredStore.set(r, i), this._sendMutation(r, s), this.notifyAll(), i.promise;
    });
    p(this, "_wsOnOpen", (e) => {
      const n = e.target;
      if (this._ws !== n) {
        E.info(
          "[socket][open]",
          n._id,
          "skip; this is no longer the current ws"
        );
        return;
      }
      E.info("[socket][open]", this._ws._id), this._setStatus(K.OPENED), this.getCurrentUser().then((r) => {
        var s;
        this._trySend(P(), {
          op: "init",
          "app-id": this.config.appId,
          "refresh-token": (s = r.user) == null ? void 0 : s.refresh_token,
          versions: this.versions,
          // If an admin token is provided for an app, we will
          // skip all permission checks. This is an advanced feature,
          // to let users write internal tools
          // This option is not exposed in `Config`, as it's
          // not ready for prme time
          "__admin-token": this.config.__adminToken
        });
      });
    });
    p(this, "_wsOnMessage", (e) => {
      const n = e.target, r = JSON.parse(e.data.toString());
      if (this._ws !== n) {
        E.info(
          "[socket][message]",
          n._id,
          r,
          "skip; this is no longer the current ws"
        );
        return;
      }
      this._handleReceive(n._id, JSON.parse(e.data.toString()));
    });
    p(this, "_wsOnError", (e) => {
      const n = e.target;
      if (this._ws !== n) {
        E.info(
          "[socket][error]",
          n._id,
          "skip; this is no longer the current ws"
        );
        return;
      }
      E.error("[socket][error]", n._id, e);
    });
    p(this, "_wsOnClose", (e) => {
      const n = e.target;
      if (this._ws !== n) {
        E.info(
          "[socket][close]",
          n._id,
          "skip; this is no longer the current ws"
        );
        return;
      }
      this._setStatus(K.CLOSED);
      for (const r of Object.values(this._rooms))
        r.isConnected = !1;
      if (this._isShutdown) {
        E.info(
          "[socket][close]",
          n._id,
          "Reactor has been shut down and will not reconnect"
        );
        return;
      }
      E.info(
        "[socket][close]",
        n._id,
        "schedule reconnect, ms =",
        this._reconnectTimeoutMs
      ), setTimeout(() => {
        if (this._reconnectTimeoutMs = Math.min(
          this._reconnectTimeoutMs + 1e3,
          1e4
        ), !this._isOnline) {
          E.info(
            "[socket][close]",
            n._id,
            "we are offline, no need to start socket"
          );
          return;
        }
        this._startSocket();
      }, this._reconnectTimeoutMs);
    });
    this.config = { ...gs, ...e }, this.versions = { ...s || {}, "@instantdb/core": _s }, this.config.schema && (this._linkIndex = ys(this.config.schema)), Ss() && (typeof BroadcastChannel == "function" && (this._broadcastChannel = new BroadcastChannel("@instantdb"), this._broadcastChannel.addEventListener("message", async (i) => {
      var o;
      if (((o = i.data) == null ? void 0 : o.type) === "auth") {
        const a = await this.getCurrentUser();
        this.updateUser(a.user);
      }
    })), this._oauthCallbackResponse = this._oauthLoginInit(), this._initStorage(n), this.getCurrentUser(), r.getIsOnline().then((i) => {
      this._isOnline = i, this._startSocket(), r.listen((o) => {
        o !== this._isOnline && (E.info("[network] online =", o), this._isOnline = o, this._isOnline && this._startSocket());
      });
    }), typeof addEventListener < "u" && (this._beforeUnload = this._beforeUnload.bind(this), addEventListener("beforeunload", this._beforeUnload)));
  }
  _initStorage(e) {
    this._persister = new e(`instant_${this.config.appId}_5`), this.querySubs = new Rt(
      this._persister,
      "querySubs",
      {},
      this._onMergeQuerySubs,
      Ms,
      vs
    ), this.pendingMutations = new Rt(
      this._persister,
      "pendingMutations",
      /* @__PURE__ */ new Map(),
      this._onMergePendingMutations,
      (n) => JSON.stringify([...n.entries()]),
      (n) => new Map(JSON.parse(n))
    ), this._beforeUnloadCbs.push(() => {
      this.pendingMutations.flush(), this.querySubs.flush();
    });
  }
  _beforeUnload() {
    for (const e of this._beforeUnloadCbs)
      e();
  }
  /**
   * @param {'enqueued' | 'pending' | 'synced' | 'timeout' |  'error' } status
   * @param string clientId
   * @param {{message?: string, hint?: string, error?: Error}} [errDetails]
   */
  _finishTransaction(e, n, r) {
    const s = this.mutationDeferredStore.get(n);
    this.mutationDeferredStore.delete(n);
    const i = e !== "error" && e !== "timeout";
    !s && !i && console.error("Mutation failed", { status: e, clientId: n, ...r }), s && (i ? s.resolve({ status: e, clientId: n }) : s.reject({ status: e, clientId: n, ...r }));
  }
  _setStatus(e, n) {
    this.status = e, this._errorMessage = n;
  }
  /**
   * On refresh we clear out pending mutations that we know have been applied
   * by the server and thus those mutations are applied in the instaql result
   * returned by the server
   */
  _cleanPendingMutations(e) {
    this.pendingMutations.set((n) => {
      const r = new Map(n);
      return [...n.entries()].forEach(([s, i]) => {
        i["tx-id"] <= e && r.delete(s);
      }), r;
    });
  }
  _flushEnqueuedRoomData(e) {
    var s, i;
    const n = (i = (s = this._presence[e]) == null ? void 0 : s.result) == null ? void 0 : i.user, r = this._broadcastQueue[e];
    if (this._broadcastQueue[e] = [], n && this._trySetPresence(e, n), r)
      for (const o of r) {
        const { topic: a, roomType: c, data: u } = o;
        this._tryBroadcast(e, c, a, u);
      }
  }
  _handleReceive(e, n) {
    var s, i, o, a;
    const r = !!this.config.schema && ("cardinalityInference" in this.config ? !!this.config.cardinalityInference : !0);
    switch (It[n.op] || E.info("[receive]", e, n.op, n), n.op) {
      case "init-ok":
        this._setStatus(K.AUTHENTICATED), this._reconnectTimeoutMs = 0, this._setAttrs(n.attrs), this._flushPendingMessages(), this._sessionId = n["session-id"];
        for (const M of Object.keys(this._rooms))
          this._tryJoinRoom(M);
        break;
      case "add-query-exists":
        this.notifyOneQueryOnce(U(n.q));
        break;
      case "add-query-ok":
        const { q: c, result: u, "processed-tx-id": f } = n;
        this._cleanPendingMutations(f);
        const l = U(c), d = (i = (s = u == null ? void 0 : u[0]) == null ? void 0 : s.data) == null ? void 0 : i["page-info"], h = (a = (o = u == null ? void 0 : u[0]) == null ? void 0 : o.data) == null ? void 0 : a.aggregate, y = jt(u), _ = $e(
          this.attrs,
          y,
          r,
          this._linkIndex
        );
        this.querySubs.set((M) => (M[l].result = { store: _, pageInfo: d, aggregate: h }, M)), this.notifyOne(l), this.notifyOneQueryOnce(l);
        break;
      case "refresh-ok":
        const { computations: w, attrs: g, "processed-tx-id": O } = n;
        this._cleanPendingMutations(O), this._setAttrs(g);
        const S = w.map((M) => {
          var ct, ut, ft, lt;
          const ee = M["instaql-query"], V = M["instaql-result"], Ae = U(ee), ue = jt(V), hn = $e(
            this.attrs,
            ue,
            r,
            this._linkIndex
          ), pn = (ut = (ct = V == null ? void 0 : V[0]) == null ? void 0 : ct.data) == null ? void 0 : ut["page-info"], yn = (lt = (ft = V == null ? void 0 : V[0]) == null ? void 0 : ft.data) == null ? void 0 : lt.aggregate;
          return { hash: Ae, store: hn, pageInfo: pn, aggregate: yn };
        });
        S.forEach(({ hash: M, store: ee, pageInfo: V, aggregate: Ae }) => {
          this.querySubs.set((ue) => (ue[M].result = { store: ee, pageInfo: V, aggregate: Ae }, ue));
        }), S.forEach(({ hash: M }) => {
          this.notifyOne(M);
        });
        break;
      case "transact-ok":
        const { "client-event-id": k, "tx-id": m } = n, x = this._rewriteMutations(
          this.attrs,
          this.pendingMutations.currentValue
        ).get(k);
        if (!x)
          break;
        const B = { ...x, "tx-id": m };
        this.pendingMutations.set((M) => (M.set(k, B), M)), this._finishTransaction("synced", k);
        const Me = x["tx-steps"].filter(([M, ...ee]) => M === "add-attr").map(([M, ee]) => ee).concat(Object.values(this.attrs));
        this._setAttrs(Me);
        break;
      case "refresh-presence":
        const st = n["room-id"];
        this._setPresencePeers(st, n.data), this._notifyPresenceSubs(st);
        break;
      case "server-broadcast":
        const ln = n["room-id"], dn = n.topic;
        this._notifyBroadcastSubs(ln, dn, n);
        break;
      case "join-room-ok":
        const Z = n["room-id"], it = this._rooms[Z];
        if (!it) {
          this._roomsPendingLeave[Z] && (this._tryLeaveRoom(Z), delete this._roomsPendingLeave[Z]);
          break;
        }
        it.isConnected = !0, this._notifyPresenceSubs(Z), this._flushEnqueuedRoomData(Z);
        break;
      case "join-room-error":
        const ot = n["room-id"], at = this._rooms[ot];
        at && (at.error = n.error), this._notifyPresenceSubs(ot);
        break;
      case "error":
        this._handleReceiveError(n);
        break;
    }
  }
  /**
   * @param {'timeout' | 'error'} status
   * @param {string} eventId
   * @param {{message?: string, hint?: string, error?: Error}} errDetails
   */
  _handleMutationError(e, n, r) {
    const s = this.pendingMutations.currentValue.get(n);
    s && (e !== "timeout" || !s["tx-id"]) && (this.pendingMutations.set((i) => (i.delete(n), i)), this.notifyAll(), this.notifyAttrsSubs(), this.notifyMutationErrorSubs(r), this._finishTransaction(e, n, r));
  }
  _handleReceiveError(e) {
    var c, u, f;
    const n = e["client-event-id"], r = this.pendingMutations.currentValue.get(n), s = {
      message: e.message || "Uh-oh, something went wrong. Ping Joe & Stopa."
    };
    if (e.hint && (s.hint = e.hint), r) {
      const l = {
        message: e.message,
        hint: e.hint
      };
      this._handleMutationError("error", n, l);
      return;
    }
    const i = e.q || ((c = e["original-event"]) == null ? void 0 : c.q);
    if (i) {
      const l = U(i);
      this.querySubs.set((d) => (delete d[l], d)), this.notifyQueryError(U(i), s), this.notifyQueryOnceError(l, n, s);
      return;
    }
    if (((u = e["original-event"]) == null ? void 0 : u.op) === "init") {
      if (e.type === "record-not-found" && ((f = e.hint) == null ? void 0 : f["record-type"]) === "app-user") {
        this.changeCurrentUser(null);
        return;
      }
      this._setStatus(K.ERRORED, s), this.notifyAll();
      return;
    }
    const a = { ...e };
    delete a.message, delete a.hint, console.error(e.message, a), e.hint && console.error(
      `This error comes with some debugging information. Here it is: 
`,
      e.hint
    );
  }
  notifyQueryOnceError(e, n, r) {
    var i;
    const s = (i = this.queryOnceDfds[e]) == null ? void 0 : i.find((o) => o.eventId === n);
    s && s.dfd.reject(r);
  }
  _setAttrs(e) {
    this.attrs = e.reduce((n, r) => (n[r.id] = r, n), {}), this.notifyAttrsSubs();
  }
  _startQuerySub(e, n) {
    const r = P();
    return this.querySubs.set((s) => (s[n] = s[n] || { q: e, result: null, eventId: r }, s)), this._trySendAuthed(r, { op: "add-query", q: e }), r;
  }
  /**
   *  When a user subscribes to a query the following side effects occur:
   *
   *  - We update querySubs to include the new query
   *  - We update queryCbs to include the new cb
   *  - If we already have a result for the query we call cb immediately
   *  - We send the server an `add-query` message
   *
   *  Returns an unsubscribe function
   */
  subscribeQuery(e, n) {
    const r = U(e), s = this.getPreviousResult(e);
    return s && n(s), this.queryCbs[r] = this.queryCbs[r] ?? [], this.queryCbs[r].push({ q: e, cb: n }), this._startQuerySub(e, r), () => {
      this._unsubQuery(e, r, n);
    };
  }
  queryOnce(e) {
    const n = new Ct();
    if (!this._isOnline)
      return n.reject(
        new Error("We can't run `queryOnce`, because the device is offline.")
      ), n.promise;
    if (!this.querySubs)
      return n.reject(
        new Error(
          "We can't run `queryOnce` on the backend. Use adminAPI.query instead: https://www.instantdb.com/docs/backend#query"
        )
      ), n.promise;
    const r = U(e), s = this._startQuerySub(e, r);
    return this.queryOnceDfds[r] = this.queryOnceDfds[r] ?? [], this.queryOnceDfds[r].push({ q: e, dfd: n, eventId: s }), setTimeout(
      () => n.reject(new Error("Query timed out")),
      bs
    ), n.promise;
  }
  _completeQueryOnce(e, n, r) {
    this.queryOnceDfds[n] && (this.queryOnceDfds[n] = this.queryOnceDfds[n].filter(
      (s) => s.dfd !== r
    ), this._cleanupQuery(e, n));
  }
  _unsubQuery(e, n, r) {
    this.queryCbs[n] && (this.queryCbs[n] = this.queryCbs[n].filter((s) => s.cb !== r), this._cleanupQuery(e, n));
  }
  _cleanupQuery(e, n) {
    var s, i;
    (s = this.queryCbs[n]) != null && s.length || (i = this.queryOnceDfds[n]) != null && i.length || (delete this.queryCbs[n], delete this.queryOnceDfds[n], this._trySendAuthed(P(), { op: "remove-query", q: e }));
  }
  // When we `pushTx`, it's possible that we don't yet have `this.attrs`
  // This means that `tx-steps` in `pendingMutations` will include `add-attr`
  // commands for attrs that already exist.
  //
  // This will also affect `add-triple` and `retract-triple` which
  // reference attr-ids that do not match the server.
  //
  // We fix this by rewriting `tx-steps` in each `pendingMutation`.
  // We remove `add-attr` commands for attrs that already exist.
  // We update `add-triple` and `retract-triple` commands to use the
  // server attr-ids.
  _rewriteMutations(e, n) {
    if (!e) return n;
    const r = (c) => {
      const [u, f, l] = c["forward-identity"];
      return j(e, f, l);
    }, s = (c) => {
      const [u, f, l] = c["forward-identity"];
      return G(e, f, l);
    }, i = { attrIdMap: {}, refSwapAttrIds: /* @__PURE__ */ new Set() }, o = (c) => {
      const u = [];
      for (const f of c) {
        const [l] = f;
        if (l === "add-attr") {
          const [h, y] = f, _ = r(y);
          if (_) {
            i.attrIdMap[y.id] = _.id;
            continue;
          }
          if (y["value-type"] === "ref") {
            const w = s(y);
            if (w) {
              i.attrIdMap[y.id] = w.id, i.refSwapAttrIds.add(y.id);
              continue;
            }
          }
        }
        const d = $r(i, f);
        u.push(d);
      }
      return u;
    }, a = /* @__PURE__ */ new Map();
    for (const [c, u] of n.entries())
      a.set(c, { ...u, "tx-steps": o(u["tx-steps"]) });
    return a;
  }
  // ---------------------------
  // Transact
  optimisticAttrs() {
    var o;
    const e = [
      ...this.pendingMutations.currentValue.values()
    ].flatMap((a) => a["tx-steps"]), n = new Set(
      e.filter(([a, c]) => a === "delete-attr").map(([a, c]) => c)
    ), r = [];
    for (const [a, c] of e)
      if (a === "add-attr")
        r.push(c);
      else if (a === "update-attr" && c.id && ((o = this.attrs) != null && o[c.id])) {
        const u = { ...this.attrs[c.id], ...c };
        r.push(u);
      }
    const s = [
      ...Object.values(this.attrs || {}),
      ...r
    ].filter((a) => !n.has(a.id));
    return Object.fromEntries(
      s.map((a) => [a.id, a])
    );
  }
  /** Runs instaql on a query and a store */
  dataForQuery(e) {
    const n = this._errorMessage;
    if (n)
      return { error: n };
    if (!this.querySubs || !this.pendingMutations) return;
    const r = this.querySubs.version(), s = this.querySubs.currentValue, i = this.pendingMutations.version(), o = this.pendingMutations.currentValue, { q: a, result: c } = s[e] || {};
    if (!c) return;
    const u = this._dataForQueryCache[e];
    if (u && r === u.querySubVersion && i === u.pendingMutationsVersion)
      return u.data;
    const { store: f, pageInfo: l, aggregate: d } = c, y = [...this._rewriteMutations(f.attrs, o).values()].flatMap((g) => g["tx-steps"]), _ = Hn(f, y), w = kr({ store: _, pageInfo: l, aggregate: d }, a);
    return this._dataForQueryCache[e] = {
      querySubVersion: r,
      pendingMutationsVersion: i,
      data: w
    }, w;
  }
  /** Re-compute all subscriptions */
  notifyAll() {
    Object.keys(this.queryCbs).forEach((e) => {
      this.notifyOne(e);
    });
  }
  loadedNotifyAll() {
    this.pendingMutations.isLoading() || this.querySubs.isLoading() || this.notifyAll();
  }
  shutdown() {
    this._isShutdown = !0, this._ws.close();
  }
  /**
   * Sends mutation to server and schedules a timeout to cancel it if
   * we don't hear back in time.
   * Note: If we're offline we don't schedule a timeout, we'll schedule it
   * later once we're back online and send the mutation again
   *
   */
  _sendMutation(e, n) {
    if (n.error) {
      this._handleMutationError("error", e, {
        error: n.error,
        message: n.error.message
      });
      return;
    }
    if (this.status !== K.AUTHENTICATED) {
      this._finishTransaction("enqueued", e);
      return;
    }
    const r = Math.max(
      5e3,
      this.pendingMutations.currentValue.size * 5e3
    );
    this._isOnline ? (this._trySend(e, n), setTimeout(() => {
      this._finishTransaction("pending", e);
    }, 3e3), setTimeout(() => {
      this._isOnline && this._handleMutationError("timeout", e, {
        message: "transaction timed out"
      });
    }, r)) : this._finishTransaction("enqueued", e);
  }
  // ---------------------------
  // Websocket
  /** Send messages we accumulated while we were connecting */
  _flushPendingMessages() {
    Object.keys(this.queryCbs).map((s) => this.querySubs.currentValue[s]).filter((s) => s).forEach(({ eventId: s, q: i }) => {
      this._trySendAuthed(s, { op: "add-query", q: i });
    }), Object.values(this.queryOnceDfds).flat().forEach(({ eventId: s, q: i }) => {
      this._trySendAuthed(s, { op: "add-query", q: i });
    }), this._rewriteMutations(
      this.attrs,
      this.pendingMutations.currentValue
    ).forEach((s, i) => {
      s["tx-id"] || this._sendMutation(i, s);
    });
  }
  _trySendAuthed(...e) {
    this.status === K.AUTHENTICATED && this._trySend(...e);
  }
  _trySend(e, n, r) {
    this._ws.readyState === Tt && (It[n.op] || E.info("[send]", this._ws._id, n.op, n), this._ws.send(JSON.stringify({ "client-event-id": e, ...n })));
  }
  _startSocket() {
    if (this._ws && this._ws.readyState == ws) {
      E.info(
        "[socket][start]",
        this._ws._id,
        "maintained as current ws, we were still in a connecting state"
      );
      return;
    }
    const e = this._ws;
    this._ws = Os(
      `${this.config.websocketURI}?app_id=${this.config.appId}`
    ), this._ws.onopen = this._wsOnOpen, this._ws.onmessage = this._wsOnMessage, this._ws.onclose = this._wsOnClose, this._ws.onerror = this._wsOnError, E.info("[socket][start]", this._ws._id), (e == null ? void 0 : e.readyState) === Tt && (E.info(
      "[socket][start]",
      this._ws._id,
      "close previous ws id = ",
      e._id
    ), e.close());
  }
  /**
   * Given a key, returns a stable local id, unique to this device and app.
   *
   * This can be useful if you want to create guest ids for example.
   *
   * Note: If the user deletes their local storage, this id will change.
   *
   * We use this._localIdPromises to ensure that we only generate a local
   * id once, even if multiple callers call this function concurrently.
   */
  async getLocalId(e) {
    const n = `localToken_${e}`, r = await this._persister.getItem(n);
    if (r) return r;
    if (this._localIdPromises[n])
      return this._localIdPromises[n];
    const s = P();
    return this._localIdPromises[n] = this._persister.setItem(n, s).then(() => s), this._localIdPromises[n];
  }
  // ----
  // Auth
  _replaceUrlAfterOAuth() {
    if (typeof URL > "u")
      return;
    const e = new URL(window.location.href);
    if (e.searchParams.get(Ce)) {
      const n = e.toString();
      e.searchParams.delete(Ce), e.searchParams.delete("code"), e.searchParams.delete("error");
      const r = e.pathname + (e.searchParams.size ? "?" + e.searchParams : "") + e.hash;
      if (history.replaceState(history.state, "", r), // @ts-ignore (waiting for ts support)
      typeof navigation == "object" && // @ts-ignore (waiting for ts support)
      typeof navigation.addEventListener == "function" && // @ts-ignore (waiting for ts support)
      typeof navigation.removeEventListener == "function") {
        let s = !1;
        const i = (o) => {
          var a;
          s || (s = !0, navigation.removeEventListener("navigate", i), !o.userInitiated && o.navigationType === "replace" && ((a = o.destination) == null ? void 0 : a.url) === n && history.replaceState(history.state, "", r));
        };
        navigation.addEventListener("navigate", i);
      }
    }
  }
  /**
   *
   * @returns Promise<null | {error: {message: string}}>
   */
  async _oauthLoginInit() {
    var s, i, o, a;
    if (typeof window > "u" || typeof window.location > "u" || typeof URLSearchParams > "u")
      return null;
    const e = new URLSearchParams(window.location.search);
    if (!e.get(Ce))
      return null;
    const n = e.get("error");
    if (n)
      return this._replaceUrlAfterOAuth(), { error: { message: n } };
    const r = e.get("code");
    if (!r)
      return null;
    this._replaceUrlAfterOAuth();
    try {
      const { user: c } = await Pt({
        apiURI: this.config.apiURI,
        appId: this.config.appId,
        code: r
      });
      return this.setCurrentUser(c), null;
    } catch (c) {
      return ((s = c == null ? void 0 : c.body) == null ? void 0 : s.type) === "record-not-found" && ((o = (i = c == null ? void 0 : c.body) == null ? void 0 : i.hint) == null ? void 0 : o["record-type"]) === "app-oauth-code" && await this._hasCurrentUser() ? null : { error: { message: ((a = c == null ? void 0 : c.body) == null ? void 0 : a.message) || "Error logging in." } };
    }
  }
  async _waitForOAuthCallbackResponse() {
    return await this._oauthCallbackResponse;
  }
  __subscribeMutationErrors(e) {
    return this.mutationErrorCbs.push(e), () => {
      this.mutationErrorCbs = this.mutationErrorCbs.filter((n) => n !== e);
    };
  }
  subscribeAuth(e) {
    this.authCbs.push(e);
    const n = this._currentUserCached;
    n.isLoading || e(this._currentUserCached);
    let r = !1;
    return this.getCurrentUser().then((s) => {
      r || be(s, n) || e(s);
    }), () => {
      r = !0, this.authCbs = this.authCbs.filter((s) => s !== e);
    };
  }
  subscribeAttrs(e) {
    return this.attrsCbs.push(e), this.attrs && e(this.attrs), () => {
      this.attrsCbs = this.attrsCbs.filter((n) => n !== e);
    };
  }
  notifyAuthSubs(e) {
    this.authCbs.forEach((n) => n(e));
  }
  notifyMutationErrorSubs(e) {
    this.mutationErrorCbs.forEach((n) => n(e));
  }
  notifyAttrsSubs() {
    if (!this.attrs) return;
    const e = this.optimisticAttrs();
    this.attrsCbs.forEach((n) => n(e));
  }
  async setCurrentUser(e) {
    await this._persister.setItem(Re, JSON.stringify(e));
  }
  getCurrentUserCached() {
    return this._currentUserCached;
  }
  async getCurrentUser() {
    const e = await this._waitForOAuthCallbackResponse();
    if (e != null && e.error) {
      const s = { error: e.error, user: void 0 };
      return this._currentUserCached = { isLoading: !1, ...s }, s;
    }
    const n = await this._persister.getItem(Re), r = { user: JSON.parse(n), error: void 0 };
    return this._currentUserCached = {
      isLoading: !1,
      ...r
    }, r;
  }
  async _hasCurrentUser() {
    const e = await this._persister.getItem(Re);
    return JSON.parse(e) != null;
  }
  async changeCurrentUser(e) {
    var r;
    const { user: n } = await this.getCurrentUser();
    if (!be(n, e)) {
      await this.setCurrentUser(e), this.updateUser(e);
      try {
        (r = this._broadcastChannel) == null || r.postMessage({ type: "auth" });
      } catch (s) {
        console.error("Error posting message to broadcast channel", s);
      }
    }
  }
  updateUser(e) {
    const n = { error: void 0, user: e };
    this._currentUserCached = { isLoading: !1, ...n }, this._dataForQueryCache = {}, this.querySubs.set((r) => (Object.keys(r).forEach((s) => {
      delete r[s].result;
    }), r)), this._reconnectTimeoutMs = 0, this._ws.close(), this._oauthCallbackResponse = null, this.notifyAuthSubs(n);
  }
  sendMagicCode({ email: e }) {
    return ss({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      email: e
    });
  }
  async signInWithMagicCode({ email: e, code: n }) {
    const r = await is({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      email: e,
      code: n
    });
    return await this.changeCurrentUser(r.user), r;
  }
  async signInWithCustomToken(e) {
    const n = await os({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      refreshToken: e
    });
    return await this.changeCurrentUser(n.user), n;
  }
  async signOut() {
    var r;
    const e = await this.getCurrentUser(), n = (r = e == null ? void 0 : e.user) == null ? void 0 : r.refresh_token;
    if (n)
      try {
        await cs({
          apiURI: this.config.apiURI,
          appId: this.config.appId,
          refreshToken: n
        });
      } catch {
      }
    await this.changeCurrentUser(null);
  }
  /**
   * Creates an OAuth authorization URL.
   * @param {Object} params - The parameters to create the authorization URL.
   * @param {string} params.clientName - The name of the client requesting authorization.
   * @param {string} params.redirectURL - The URL to redirect users to after authorization.
   * @returns {string} The created authorization URL.
   */
  createAuthorizationURL({ clientName: e, redirectURL: n }) {
    const { apiURI: r, appId: s } = this.config;
    return `${r}/runtime/oauth/start?app_id=${s}&client_name=${e}&redirect_uri=${n}`;
  }
  async exchangeCodeForToken({ code: e, codeVerifier: n }) {
    const r = await Pt({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      code: e,
      codeVerifier: n
    });
    return await this.changeCurrentUser(r.user), r;
  }
  issuerURI() {
    const { apiURI: e, appId: n } = this.config;
    return `${e}/runtime/${n}`;
  }
  /**
   * @param {Object} params
   * @param {string} params.clientName - The name of the client requesting authorization.
   * @param {string} params.idToken - The id_token from the external service
   * @param {string | null | undefined} [params.nonce] - The nonce used when requesting the id_token from the external service
   */
  async signInWithIdToken({ idToken: e, clientName: n, nonce: r }) {
    var a;
    const s = await this.getCurrentUser(), i = (a = s == null ? void 0 : s.user) == null ? void 0 : a.refresh_token, o = await as({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      idToken: e,
      clientName: n,
      nonce: r,
      refreshToken: i
    });
    return await this.changeCurrentUser(o.user), o;
  }
  // --------
  // Rooms
  joinRoom(e) {
    return this._rooms[e] || (this._rooms[e] = {
      isConnected: !1,
      error: void 0
    }), this._presence[e] = this._presence[e] || {}, this._tryJoinRoom(e), () => {
      this._cleanupRoom(e);
    };
  }
  _cleanupRoom(e) {
    var n, r, s;
    if (!((r = (n = this._presence[e]) == null ? void 0 : n.handlers) != null && r.length) && !Object.keys(this._broadcastSubs[e] ?? {}).length) {
      const i = (s = this._rooms[e]) == null ? void 0 : s.isConnected;
      delete this._rooms[e], delete this._presence[e], delete this._broadcastSubs[e], i ? this._tryLeaveRoom(e) : this._roomsPendingLeave[e] = !0;
    }
  }
  // --------
  // Presence
  // TODO: look into typing again
  getPresence(e, n, r = {}) {
    const s = this._rooms[n], i = this._presence[n];
    return !s || !i || !i.result ? null : {
      ...hs(i.result, r, this._sessionId),
      isLoading: !s.isConnected,
      error: s.error
    };
  }
  // TODO: look into typing again
  publishPresence(e, n, r) {
    const s = this._rooms[n], i = this._presence[n];
    if (!s || !i)
      return;
    i.result = i.result || {};
    const o = {
      ...i.result.user,
      ...r
    };
    i.result.user = o, s.isConnected && (this._trySetPresence(n, o), this._notifyPresenceSubs(n));
  }
  _trySetPresence(e, n) {
    this._trySendAuthed(P(), {
      op: "set-presence",
      "room-id": e,
      data: n
    });
  }
  _tryJoinRoom(e) {
    this._trySendAuthed(P(), { op: "join-room", "room-id": e }), delete this._roomsPendingLeave[e];
  }
  _tryLeaveRoom(e) {
    this._trySendAuthed(P(), { op: "leave-room", "room-id": e });
  }
  // TODO: look into typing again
  subscribePresence(e, n, r, s) {
    const i = this.joinRoom(n), o = { ...r, roomId: n, cb: s, prev: null };
    return this._presence[n] = this._presence[n] || {}, this._presence[n].handlers = this._presence[n].handlers || [], this._presence[n].handlers.push(o), this._notifyPresenceSub(n, o), () => {
      var a, c;
      this._presence[n].handlers = ((c = (a = this._presence[n]) == null ? void 0 : a.handlers) == null ? void 0 : c.filter((u) => u !== o)) ?? [], i();
    };
  }
  _notifyPresenceSubs(e) {
    var n, r;
    (r = (n = this._presence[e]) == null ? void 0 : n.handlers) == null || r.forEach((s) => {
      this._notifyPresenceSub(e, s);
    });
  }
  _notifyPresenceSub(e, n) {
    const r = this.getPresence("", e, n);
    r && (n.prev && !ps(r, n.prev) || (n.prev = r, n.cb(r)));
  }
  _setPresencePeers(e, n) {
    const r = { ...n };
    delete r[this._sessionId];
    const s = Object.fromEntries(
      Object.entries(r).map(([i, o]) => [i, o.data])
    );
    this._presence[e] = this._presence[e] || {}, this._presence[e].result = this._presence[e].result || {}, this._presence[e].result.peers = s;
  }
  // --------
  // Broadcast
  publishTopic({ roomType: e, roomId: n, topic: r, data: s }) {
    const i = this._rooms[n];
    if (i) {
      if (!i.isConnected) {
        this._broadcastQueue[n] = this._broadcastQueue[n] ?? [], this._broadcastQueue[n].push({ topic: r, roomType: e, data: s });
        return;
      }
      this._tryBroadcast(n, e, r, s);
    }
  }
  _tryBroadcast(e, n, r, s) {
    this._trySendAuthed(P(), {
      op: "client-broadcast",
      "room-id": e,
      roomType: n,
      topic: r,
      data: s
    });
  }
  subscribeTopic(e, n, r) {
    const s = this.joinRoom(e);
    return this._broadcastSubs[e] = this._broadcastSubs[e] || {}, this._broadcastSubs[e][n] = this._broadcastSubs[e][n] || [], this._broadcastSubs[e][n].push(r), this._presence[e] = this._presence[e] || {}, () => {
      this._broadcastSubs[e][n] = this._broadcastSubs[e][n].filter((i) => i !== r), this._broadcastSubs[e][n].length || delete this._broadcastSubs[e][n], s();
    };
  }
  _notifyBroadcastSubs(e, n, r) {
    var s, i, o;
    (o = (i = (s = this._broadcastSubs) == null ? void 0 : s[e]) == null ? void 0 : i[n]) == null || o.forEach((a) => {
      var f, l, d, h, y, _;
      const c = (f = r.data) == null ? void 0 : f.data, u = r.data["peer-id"] === this._sessionId ? (d = (l = this._presence[e]) == null ? void 0 : l.result) == null ? void 0 : d.user : (_ = (y = (h = this._presence[e]) == null ? void 0 : h.result) == null ? void 0 : y.peers) == null ? void 0 : _[r.data["peer-id"]];
      return a(c, u);
    });
  }
  // --------
  // Storage
  async upload(e, n) {
    var c;
    const r = await this.getCurrentUser(), s = (c = r == null ? void 0 : r.user) == null ? void 0 : c.refresh_token, i = e || n.name, o = await us({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      fileName: i,
      refreshToken: s
    });
    return await fs(o, n);
  }
  async getDownloadUrl(e) {
    var i;
    const n = await this.getCurrentUser(), r = (i = n == null ? void 0 : n.user) == null ? void 0 : i.refresh_token;
    return await ls({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      path: e,
      refreshToken: r
    });
  }
  async deleteFile(e) {
    var i;
    const n = await this.getCurrentUser(), r = (i = n == null ? void 0 : n.user) == null ? void 0 : i.refresh_token;
    return await ds({
      apiURI: this.config.apiURI,
      appId: this.config.appId,
      path: e,
      refreshToken: r
    });
  }
}
class F {
  constructor(e, n, r = { indexed: !1, unique: !1 }) {
    this.valueType = e, this.required = n, this.config = r;
  }
  optional() {
    return new F(this.valueType, !1);
  }
  unique() {
    return new F(
      this.valueType,
      this.required,
      {
        ...this.config,
        unique: !0
      }
    );
  }
  indexed() {
    return new F(
      this.valueType,
      this.required,
      {
        ...this.config,
        indexed: !0
      }
    );
  }
  // clientValidate(clientValidator: (value: ValueType) => boolean) {
  //   return new DataAttrDef(this.valueType, this.required, {
  //     ...this.config,
  //     clientValidator,
  //   });
  // }
}
class Se {
  constructor(e, n) {
    this.attrs = e, this.links = n;
  }
  asType() {
    return new Se(this.attrs, this.links);
  }
}
class ve {
  constructor(e, n, r) {
    this.entities = e, this.links = n, this.rooms = r;
  }
  withRoomSchema() {
    return new ve(
      this.entities,
      this.links,
      {}
    );
  }
}
function As(t, e) {
  return new ve(
    sn(t, e),
    // (XXX): LinksDef<any> stems from TypeScript’s inability to reconcile the
    // type EntitiesWithLinks<EntitiesWithoutLinks, Links> with
    // EntitiesWithoutLinks. TypeScript is strict about ensuring that types are
    // correctly aligned and does not allow for substituting a type that might
    // be broader or have additional properties.
    e,
    void 0
  );
}
function Ps(t) {
  return new Se(t, {});
}
function Es() {
  return new F("string", !0);
}
function Cs() {
  return new F("number", !0);
}
function Rs() {
  return new F("boolean", !0);
}
function js() {
  return new F("date", !0);
}
function Ts() {
  return new F("json", !0);
}
function Is() {
  return new F("json", !0);
}
function sn(t, e) {
  var s, i, o, a;
  const n = { fwd: {}, rev: {} };
  for (const c of Object.values(e))
    (s = n.fwd)[i = c.forward.on] || (s[i] = {}), (o = n.rev)[a = c.reverse.on] || (o[a] = {}), n.fwd[c.forward.on][c.forward.label] = {
      entityName: c.reverse.on,
      cardinality: c.forward.has
    }, n.rev[c.reverse.on][c.reverse.label] = {
      entityName: c.forward.on,
      cardinality: c.reverse.has
    };
  return Object.fromEntries(
    Object.entries(t).map(([c, u]) => [
      c,
      new Se(u.attrs, {
        ...n.fwd[c],
        ...n.rev[c]
      })
    ])
  );
}
function ks({
  entities: t,
  links: e,
  rooms: n
}) {
  return new ve(
    sn(t, e),
    // (XXX): LinksDef<any> stems from TypeScript’s inability to reconcile the
    // type EntitiesWithLinks<EntitiesWithoutLinks, Links> with
    // EntitiesWithoutLinks. TypeScript is strict about ensuring that types are
    // correctly aligned and does not allow for substituting a type that might
    // be broader or have additional properties.
    e,
    n
  );
}
const Ws = {
  // constructs
  graph: As,
  schema: ks,
  entity: Ps,
  // value types
  string: Es,
  number: Cs,
  boolean: Rs,
  date: js,
  json: Ts,
  any: Is
};
let de;
function on(t) {
  de == null || de.dispose();
  const e = Ds(), n = $s(o), r = Us(xs(t));
  function s(u) {
    var f;
    u.source === r.element.contentWindow && ((f = u.data) == null ? void 0 : f.type) === "close" && e.isVisible() && o();
  }
  function i(u) {
    const f = u.shiftKey && u.ctrlKey && u.key === "0", l = u.key === "Escape" || u.key === "Esc";
    (f || l && e.isVisible()) && o();
  }
  function o() {
    e.isVisible() ? e.element.style.display = "none" : (e.element.style.display = "block", e.element.contains(r.element) || e.element.appendChild(r.element));
  }
  function a() {
    e.element.remove(), n.element.remove(), removeEventListener("keydown", i), removeEventListener("message", s);
  }
  function c() {
    document.body.appendChild(e.element), document.body.appendChild(n.element), addEventListener("keydown", i), addEventListener("message", s), de = {
      dispose: a
    };
  }
  return c();
}
function xs(t) {
  return `${window.DEV_DEVTOOL ? "http://localhost:3000" : "https://instantdb.com"}/_devtool?appId=${t}`;
}
function Us(t) {
  const e = document.createElement("iframe");
  return e.src = t, e.className = "instant-devtool-iframe", Object.assign(e.style, {
    width: "100%",
    height: "100%",
    borderRadius: "4px",
    backgroundColor: "white",
    border: "none"
  }), { element: e };
}
function $s(t) {
  const e = `
    <svg width="32" height="32" viewBox="0 0 512 512" fill="none" xmlns="http://www.w3.org/2000/svg">
      <rect width="512" height="512" fill="black"/>
      <rect x="97.0973" y="91.3297" width="140" height="330" fill="white"/>
    </svg>
  `, n = document.createElement("button");
  return n.innerHTML = e, n.className = "instant-devtool-toggler", Object.assign(n.style, {
    // pos
    position: "fixed",
    bottom: "24px",
    left: "24px",
    height: "32px",
    width: "32px",
    // layout
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    // look
    padding: "0",
    margin: "0",
    border: "none",
    cursor: "pointer"
  }), n.addEventListener("click", t), { element: n };
}
function Ds() {
  const t = document.createElement("div");
  Object.assign(t.style, {
    position: "fixed",
    bottom: "24px",
    right: "24px",
    left: "60px",
    top: "72px",
    display: "block",
    borderRadius: "4px",
    border: "1px #ccc solid",
    boxShadow: "0px 0px 8px #00000044",
    backgroundColor: "#eee",
    zIndex: "999990"
  }), t.style.display = "none", t.className = "instant-devtool-container";
  function e() {
    return t.style.display !== "none";
  }
  return { element: t, isVisible: e };
}
const an = !0, cn = {
  apiURI: "https://api.instantdb.com",
  websocketURI: "wss://api.instantdb.com/runtime/session"
};
function qs() {
  return globalThis.__instantDbStore = globalThis.__instantDbStore ?? {}, globalThis.__instantDbStore;
}
const se = qs();
function Js(t, e, n) {
  return Ns(t, e, n);
}
function Ns(t, e, n, r) {
  const s = se[t.appId];
  if (s)
    return s;
  const i = new rn(
    {
      ...cn,
      ...t
    },
    e || nt,
    n || rt,
    r
  ), o = new Ls(
    i
  );
  return se[t.appId] = o, typeof window < "u" && typeof window.location < "u" && // show widget by default?
  ("devtool" in t ? t.devtool : an) && // only run on localhost (dev env)
  window.location.hostname === "localhost" && // used by dash and other internal consumers
  !globalThis._nodevtool && on(t.appId), o;
}
class Ls {
  constructor(e) {
    this.tx = tt(), this._reactor = e, this.auth = new un(this._reactor), this.storage = new fn(this._reactor);
  }
  /**
   * Use this to write data! You can create, update, delete, and link objects
   *
   * @see https://instantdb.com/docs/instaml
   *
   * @example
   *   // Create a new object in the `goals` namespace
   *   const goalId = id();
   *   db.transact(tx.goals[goalId].update({title: "Get fit"}))
   *
   *   // Update the title
   *   db.transact(tx.goals[goalId].update({title: "Get super fit"}))
   *
   *   // Delete it
   *   db.transact(tx.goals[goalId].delete())
   *
   *   // Or create an association:
   *   todoId = id();
   *   db.transact([
   *    tx.todos[todoId].update({ title: 'Go on a run' }),
   *    tx.goals[goalId].link({todos: todoId}),
   *  ])
   */
  transact(e) {
    return this._reactor.pushTx(e);
  }
  getLocalId(e) {
    return this._reactor.getLocalId(e);
  }
  /**
   * Use this to query your data!
   *
   * @see https://instantdb.com/docs/instaql
   *
   * @example
   *  // listen to all goals
   *  db.subscribeQuery({ goals: {} }, (resp) => {
   *    console.log(resp.data.goals)
   *  })
   *
   *  // goals where the title is "Get Fit"
   *  db.subscribeQuery(
   *    { goals: { $: { where: { title: "Get Fit" } } } },
   *    (resp) => {
   *      console.log(resp.data.goals)
   *    }
   *  )
   *
   *  // all goals, _alongside_ their todos
   *  db.subscribeQuery({ goals: { todos: {} } }, (resp) => {
   *    console.log(resp.data.goals)
   *  });
   */
  subscribeQuery(e, n) {
    return this._reactor.subscribeQuery(e, n);
  }
  /**
   * Listen for the logged in state. This is useful
   * for deciding when to show a login screen.
   *
   * @see https://instantdb.com/docs/auth
   * @example
   *   const unsub = db.subscribeAuth((auth) => {
   *     if (auth.user) {
   *     console.log('logged in as', auth.user.email)
   *    } else {
   *      console.log('logged out')
   *    }
   *  })
   */
  subscribeAuth(e) {
    return this._reactor.subscribeAuth(e);
  }
  /**
   * Join a room to publish and subscribe to topics and presence.
   *
   * @see https://instantdb.com/docs/presence-and-topics
   * @example
   * // init
   * const db = init();
   * const room = db.joinRoom(roomType, roomId);
   * // usage
   * const unsubscribeTopic = room.subscribeTopic("foo", console.log);
   * const unsubscribePresence = room.subscribePresence({}, console.log);
   * room.publishTopic("hello", { message: "hello world!" });
   * room.publishPresence({ name: "joe" });
   * // later
   * unsubscribePresence();
   * unsubscribeTopic();
   * room.leaveRoom();
   */
  joinRoom(e = "_defaultRoomType", n = "_defaultRoomId") {
    return {
      leaveRoom: this._reactor.joinRoom(n),
      subscribeTopic: (s, i) => this._reactor.subscribeTopic(n, s, i),
      subscribePresence: (s, i) => this._reactor.subscribePresence(e, n, s, i),
      publishTopic: (s, i) => this._reactor.publishTopic({ roomType: e, roomId: n, topic: s, data: i }),
      publishPresence: (s) => this._reactor.publishPresence(e, n, s),
      getPresence: (s) => this._reactor.getPresence(e, n, s)
    };
  }
  shutdown() {
    delete se[this._reactor.config.appId], this._reactor.shutdown();
  }
  /**
   * Use this for one-off queries.
   * Returns local data if available, otherwise fetches from the server.
   * Because we want to avoid stale data, this method will throw an error
   * if the user is offline or there is no active connection to the server.
   *
   * @see https://instantdb.com/docs/instaql
   *
   * @example
   *
   *  const resp = await db.queryOnce({ goals: {} });
   *  console.log(resp.data.goals)
   */
  queryOnce(e) {
    return this._reactor.queryOnce(e);
  }
}
class un {
  constructor(e) {
    this.db = e, this.sendMagicCode = (n) => this.db.sendMagicCode(n), this.signInWithMagicCode = (n) => this.db.signInWithMagicCode(n), this.signInWithToken = (n) => this.db.signInWithCustomToken(n), this.createAuthorizationURL = (n) => this.db.createAuthorizationURL(n), this.signInWithIdToken = (n) => this.db.signInWithIdToken(n), this.exchangeOAuthCode = (n) => this.db.exchangeCodeForToken(n), this.issuerURI = () => this.db.issuerURI(), this.signOut = () => this.db.signOut();
  }
}
class fn {
  constructor(e) {
    this.db = e, this.upload = (n, r) => this.db.upload(n, r), this.put = this.upload, this.getDownloadUrl = (n) => this.db.getDownloadUrl(n), this.delete = (n) => this.db.deleteFile(n);
  }
}
function Bs(t) {
  return JSON.parse(JSON.stringify(t));
}
class zs {
  constructor(e) {
    this.tx = tt(), this._reactor = e, this.auth = new un(this._reactor), this.storage = new fn(this._reactor);
  }
  /**
   * Use this to write data! You can create, update, delete, and link objects
   *
   * @see https://instantdb.com/docs/instaml
   *
   * @example
   *   // Create a new object in the `goals` namespace
   *   const goalId = id();
   *   db.transact(tx.goals[goalId].update({title: "Get fit"}))
   *
   *   // Update the title
   *   db.transact(tx.goals[goalId].update({title: "Get super fit"}))
   *
   *   // Delete it
   *   db.transact(tx.goals[goalId].delete())
   *
   *   // Or create an association:
   *   todoId = id();
   *   db.transact([
   *    tx.todos[todoId].update({ title: 'Go on a run' }),
   *    tx.goals[goalId].link({todos: todoId}),
   *  ])
   */
  transact(e) {
    return this._reactor.pushTx(e);
  }
  getLocalId(e) {
    return this._reactor.getLocalId(e);
  }
  /**
   * Use this to query your data!
   *
   * @see https://instantdb.com/docs/instaql
   *
   * @example
   *  // listen to all goals
   *  db.subscribeQuery({ goals: {} }, (resp) => {
   *    console.log(resp.data.goals)
   *  })
   *
   *  // goals where the title is "Get Fit"
   *  db.subscribeQuery(
   *    { goals: { $: { where: { title: "Get Fit" } } } },
   *    (resp) => {
   *      console.log(resp.data.goals)
   *    }
   *  )
   *
   *  // all goals, _alongside_ their todos
   *  db.subscribeQuery({ goals: { todos: {} } }, (resp) => {
   *    console.log(resp.data.goals)
   *  });
   */
  subscribeQuery(e, n) {
    return this._reactor.subscribeQuery(e, n);
  }
  /**
   * Listen for the logged in state. This is useful
   * for deciding when to show a login screen.
   *
   * @see https://instantdb.com/docs/auth
   * @example
   *   const unsub = db.subscribeAuth((auth) => {
   *     if (auth.user) {
   *     console.log('logged in as', auth.user.email)
   *    } else {
   *      console.log('logged out')
   *    }
   *  })
   */
  subscribeAuth(e) {
    return this._reactor.subscribeAuth(e);
  }
  /**
   * Join a room to publish and subscribe to topics and presence.
   *
   * @see https://instantdb.com/docs/presence-and-topics
   * @example
   * // init
   * const db = init();
   * const room = db.joinRoom(roomType, roomId);
   * // usage
   * const unsubscribeTopic = room.subscribeTopic("foo", console.log);
   * const unsubscribePresence = room.subscribePresence({}, console.log);
   * room.publishTopic("hello", { message: "hello world!" });
   * room.publishPresence({ name: "joe" });
   * // later
   * unsubscribePresence();
   * unsubscribeTopic();
   * room.leaveRoom();
   */
  joinRoom(e = "_defaultRoomType", n = "_defaultRoomId") {
    return {
      leaveRoom: this._reactor.joinRoom(n),
      subscribeTopic: (s, i) => this._reactor.subscribeTopic(n, s, i),
      subscribePresence: (s, i) => this._reactor.subscribePresence(e, n, s, i),
      publishTopic: (s, i) => this._reactor.publishTopic({ roomType: e, roomId: n, topic: s, data: i }),
      publishPresence: (s) => this._reactor.publishPresence(e, n, s),
      getPresence: (s) => this._reactor.getPresence(e, n, s)
    };
  }
  shutdown() {
    delete se[this._reactor.config.appId], this._reactor.shutdown();
  }
  /**
   * Use this for one-off queries.
   * Returns local data if available, otherwise fetches from the server.
   * Because we want to avoid stale data, this method will throw an error
   * if the user is offline or there is no active connection to the server.
   *
   * @see https://instantdb.com/docs/instaql
   *
   * @example
   *
   *  const resp = await db.queryOnce({ goals: {} });
   *  console.log(resp.data.goals)
   */
  queryOnce(e) {
    return this._reactor.queryOnce(e);
  }
}
function Ks(t, e, n) {
  const r = se[t.appId];
  if (r)
    return r;
  const s = new rn(
    {
      ...cn,
      ...t,
      cardinalityInference: !!t.schema
    },
    e || nt,
    n || rt
  ), i = new zs(s);
  return se[t.appId] = i, typeof window < "u" && typeof window.location < "u" && // show widget by default?
  ("devtool" in t ? t.devtool : an) && // only run on localhost (dev env)
  window.location.hostname === "localhost" && // used by dash and other internal consumers
  !globalThis._nodevtool && on(t.appId), i;
}
export {
  un as Auth,
  nt as IndexedDBStorage,
  Ls as InstantClient,
  zs as InstantCoreDatabase,
  fn as Storage,
  rt as WindowNetworkListener,
  Ns as _init_internal,
  Bs as coerceQuery,
  Ur as getOps,
  Ws as i,
  P as id,
  Js as init,
  Ks as init_experimental,
  Vs as lookup,
  Qs as tx,
  tt as txInit,
  _s as version,
  U as weakHash
};
