// @instantdb/core/dist/schemaTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstantGraph = exports.InstantSchemaDef = exports.EntityDef = exports.LinkAttrDef = exports.DataAttrDef = void 0;
class DataAttrDef {
    constructor(valueType, required, config = { indexed: false, unique: false }) {
        this.valueType = valueType;
        this.required = required;
        this.config = config;
    }
    optional() {
        return new DataAttrDef(this.valueType, false);
    }
    unique() {
        return new DataAttrDef(this.valueType, this.required, Object.assign(Object.assign({}, this.config), { unique: true }));
    }
    indexed() {
        return new DataAttrDef(this.valueType, this.required, Object.assign(Object.assign({}, this.config), { indexed: true }));
    }
}
exports.DataAttrDef = DataAttrDef;
class LinkAttrDef {
    constructor(cardinality, entityName) {
        this.cardinality = cardinality;
        this.entityName = entityName;
    }
}
exports.LinkAttrDef = LinkAttrDef;
class EntityDef {
    constructor(attrs, links) {
        this.attrs = attrs;
        this.links = links;
    }
    asType() {
        return new EntityDef(this.attrs, this.links);
    }
}
exports.EntityDef = EntityDef;
class InstantSchemaDef {
    constructor(entities, links, rooms) {
        this.entities = entities;
        this.links = links;
        this.rooms = rooms;
    }
    withRoomSchema() {
        return new InstantSchemaDef(this.entities, this.links, {});
    }
}
exports.InstantSchemaDef = InstantSchemaDef;
class InstantGraph {
    constructor(entities, links) {
        this.entities = entities;
        this.links = links;
    }
    withRoomSchema() {
        return new InstantGraph(this.entities, this.links);
    }
}
exports.InstantGraph = InstantGraph;
